---
title: "Initial ARTs Results"
author: "Sarah P. Flanagan"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: default
  html_document: default
header-includes: #allows you to add in your own Latex packages
- \usepackage{float} #use the 'float' package
always_allow_html: yes
---

The purpose of this model is to understand how genetic architectures of alternative reproductive tactics impact their maintenance in populations. I'm using an individual-based simulation model with different selection scenarios, types of alternative tactics, and genetic architectures (genome-wide additive genetic variance, supergenes, expression networks) To test the model and make sure that everything has been implemented correctly, I'm first testing to ensure that the model produces the expected results without the genetic architectures.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H")
knitr::opts_knit$set(root.dir='../results/')
```
```{r, echo=FALSE}
library(knitr)
library(RColorBrewer)
library(scales)
read_chunk("../R/002_expectationTests.R")
cols<-c(courter="#33a02c",parent="#1f78b4",cpref="#b2df8a",ppref="#a6cee3")
cols2<-c(CP='#a1dab4',NCP='#41b6c4',CNP='#2c7fb8',NCNP='#253494')
```


<!-- Based on previous theoretical work, the following expectations exist: -->
<!-- ```{r expecations, echo=FALSE} -->
<!-- expectations<-data.frame(Scenario=c("Courtship trait", "Parental Trait","Courtship and Parental Traits", -->
<!--                                     "Courtship Trait + Neg Freq Dep Sel", "Parental Trait + Neg Freq Dep Sel", -->
<!--                                     "Parental Trait and courtship Trait + Neg Freq Dep Sel"), -->
<!--                          Expected_Outcome=c("Both alts maintained if costs = benefits","Both alts maintained if costs = benefits", -->
<!--                                             "Coevolution of both alts, and both alts maintained if costs = benefits", -->
<!--                                             "both alts maintained", "?","?")) -->
<!-- kable(expectations, caption = "Expected outcomes") -->
<!-- ``` -->

<!-- One caveat to comparing to previous work is that in my model the parental and courtship traits have a trade-off, and I haven't found a model incorporated tradeoffs and negative frequency dependent selection. -->

<!-- But we can look at what happens with the runs that I've done *(note that these are individual runs, not averages of multiple runs)*. Let's start with the ones that do not have a genetic architecture. -->

## Overview of the model 

Males can be courters or not-courters and parents or not-parents. When the model is run with both traits, this results in four possible morphs: courter/parent, courter/not-parent, not-courter/parent, and not-courter/not-parent. In each generation, the population follows the following timeline: 

1. Females choose a nest
2. Males fertilize eggs
3. Nests survive or die
4. Viability selection on progeny
5. Stochastic survival to adulthood

```{r diagram}
library(DiagrammeR)
# mermaid("
#   graph LR
#     A[Females Choose a Nest]-->B[Males Fertilize Eggs]
#     B[Males Fertilize Eggs]-->C[Nest Survival]
#     C[Nest Survival]-->D[Viability Selection on Progeny]
#     D[Viability Selection on Progeny]-->E[Stochastic Survival to Adulthood]
#     E-->A
#   fontsize 2
#   ")

nodes<-create_node_df(
  n=5,
  nodes = LETTERS[1:5],
  label = c("Females Choose a Nest","Males Fertilize Eggs","Nest Survival",
            "Viability Selection on Progeny","Stochastic Survival to Adulthood"),
  shape = "rectangle", fontsize=7, type="lower",width=1.4,height=0.35,
  x=c(0,0.75,0.6,-1,-0.9),y=c(0.9,0,-0.9,-0.75,0.25)
)

edges<-create_edge_df(
  from = c(1,2,3,4,5),
  to = c(2,3,4,5,1)
)

graph <-
  create_graph(
    nodes_df = nodes,
    edges_df = edges
  )

render_graph(graph)
```


### 1. Choosing a nest

A female samples 50 males and chooses a male to nest with based on his courtship trait. If there are no courtship traits in the model, she chooses based on the male's parental trait. If she does not encouter an acceptable male, she does not nest. If she encounters multiple equally-acceptable males, she randomly selects one of them.  

### 2. Fertilization

Once a female decides to nest, up to three males can fertilize the nest. Courters and parental males can contribute more sperm than non-courter and non-parental males: $r_{courter}=r_{parent}=8$ and $r_{non-courter}=r_{non-parent}=4$. A courter/non-parent has $r_{non-parent}$ and a non-courter/parent has $r_{parent}$. The male with whom the female is nesting gets $r_{parent}/\Sigma{n_{sperm}}$ and additional sneaker males (up to 2) get $(r_{non-parent}*0.5/\Sigma{n_{sperm}})$, where $\Sigma{n_{sperm}}$ is the total number of sperm contributed by all of the males, weighted by the sperm competition factor (0.5 is the default for all males except the nesting male). So, when a female mates with one courter and two sneakers, $\Sigma{n_{sperm}}$ = $r_{courter}$ + 2$*$(0.5$*$$r_{sneaker}$), where $r_{courter}$ = 8 and $r_{sneaker}$ = 4, therefore $\Sigma{n_{sperm}}$ = `r sum(8,rep((0.5*4),2))`. 

That being said, every time a male mates he uses his sperm, so after one mating where a courter fertilizes 50% of the female's 4 eggs, he only has 6 sperm for his next mating.

### 3. Nest Survival

Before the babies can survive, the nest has to survive. This step is only relevant when parental traits are in the model - if only the courtship trait is specified, then all progeny in the nest survive at this point. When males have the parental trait, if the female has given eggs to a non-parental male (because she chose based on courtship traits), then the nest has a 10% chance of surviving. If the female has given eggs to a parental male, the nest has a 90% chance of surviving. 

### 4. Viability selection and 5. Stochastic survival

Then the offspring experience viability selection. Courters and parental males are disfavored in viability selection, with a survival probability of `r exp(-0.5/(2*50))`. If an individual is both a courter and a parental male, the survival probability is `r exp(-0.5/(2*50))*exp(-0.5/(2*50))`. Non-courters and non-parental males have survival probabilities of 1. Once viability selection has been imposed, individuals die or survive randomly, and the next generation gets a chance to mate.

### Evaluating equilibrium

After 10000 generations, I evaluate whether the population is at equilibrium. Each generation I track the change in frequency of the courter and parent traits, and I calculate the variance in the change in frequency over those initial 10000 generations. I declare an equilibrium has been reached if the last change in frequency of both traits is less than the variance in changes in frequency. If the last change in frequency is larger than the variance, then I run the model for an additional 2000 generations or until an equilibrium is reached, whichever comes first. 

<!-- ## No genetic basis to traits -->

<!-- In these cases, males are randomly assigned a trait value from a normal distribution with a standard deviation of 0.5. This trait value is compared to the population-wide threshold to determine whether the individual is a courter or a sneaker. The threshold is set in the initial generation of the model as the mean of the male trait values at that point. In these iterations, the threshold is static.  -->

<!-- To choose a nest, females prefer "better" males (courting and/or parental males). If none of the males a female encounters are a courter (or a parental when courtship traits are absent), she does not nest.  -->

<!-- In these models, courters and parental males are disfavored in viability selection, with a survival probability of `r exp(-0.5/(2*50))`. If an individual is both a courter and a parental male, the survival probability is `r exp(-0.5/(2*50))*exp(-0.5/(2*50))`. Non-courters and non-parental males have survival probabilities of 1. -->

<!-- ### Males have courtship traits -->

<!-- Females choose nests based on whether the male is a courter or not. Because parental care is not incorporated in this model, all nests survive. -->

<!-- ```{r CourterConditionalData, echo=FALSE} -->
<!-- ``` -->
<!-- ```{r plot_courter_conditional, echo=FALSE} -->
<!-- plot(cc.sum$CourterFreq,lwd=2,xlab="Generations",ylab="Frequency of Courting Males", -->
<!--      bty="L",type="l",ylim = c(0,1),col=cols["courter"]) -->
<!-- ``` -->


<!-- ### Males have parental traits -->

<!-- Females choose nests based on whether males are parentals or not, and only mate with parental males. Parental males provide care that allows nests to have a 90% chance of survival.  -->

<!-- ```{r parental_traits_conditional, echo=FALSE} -->
<!-- pc.sum<-read.delim("parent-conditional_summary.txt") -->
<!-- plot(pc.sum$ParentFreq,lwd=2,xlab="Generations",ylab="Frequency of Parental Males", -->
<!--      bty="L",type="l",ylim = c(0,1),col=cols["parent"]) -->
<!-- ``` -->


<!-- ### Males have both courtship and parental traits -->

<!-- Females choose nests based on males' courtship morph (nest with courting males only), and then the survival of the nest depends on whether the courting male is also a parental male. If the chosen male is a parental male, the nest has a 90% chance of survival. Otherwise, it only has a 10% chance. -->

<!-- ```{r parentalCourtship_conditional, echo=FALSE} -->
<!-- pcc.sum<-read.delim("parent-courter-conditional_summary.txt") -->
<!-- plot(pcc.sum$ParentFreq,lwd=2,xlab="Generations",ylab="Frequency of Parental Males", -->
<!--      bty="u",type="l",ylim = c(0,1),col=alpha(cols["parent"],0.75)) -->
<!-- points(pcc.sum$CourterFreq,lwd=2,type="l",col=alpha(cols["courter"],0.75)) -->
<!-- mtext("Frequency of Courtship Males",4) -->
<!-- ``` -->

<!-- ### Courtship + Negative Frequency Dependent Selection -->

<!-- In this case, the female preference is based on the frequency of the courtship morphs in the population - females prefer the *less frequent* morph. Everything else is the same as above; males are assigned a trait value at random (it's not inherited), all nests survive since parental care is not incorporated in this model, and courters and parental males are disfavored in viability selection. -->

<!-- ```{r courtship_nfds,echo=FALSE} -->
<!-- cn<-read.delim("courter-conditional_nfds_summary.txt") -->
<!-- plot(cn$CourterFreq,lwd=2,xlab="Generations",ylab="Frequency of Courtship Males", -->
<!--      bty="l",type="l",ylim=c(0,1),col=alpha(cols["courter"],0.75)) -->
<!-- ``` -->

<!-- ### Parental Males + Negative Frequency Dependent Selection -->

<!-- In this case, the female preference is based on the frequency of the parental morphs in the population - females prefer the *less frequent* morph. Everything else is the same as above; males are assigned a trait value at random (it's not inherited), nests have a 90% chance of survival, and courters and parental males are disfavored in viability selection. -->

<!-- ```{r parent_nfds,echo=FALSE} -->
<!-- pn<-read.delim("parent-conditional_nfds_summary.txt") -->
<!-- plot(pn$ParentFreq,lwd=2,xlab="Generations",ylab="Frequency of Parental Males", -->
<!--      bty="l",type="l",ylim=c(0,1),col=alpha(cols["parent"],0.75)) -->
<!-- ``` -->

<!-- ### Courtship + Parental Males + Negative Frequency Dependent Selection -->

<!-- In this case, the female preference is based on the frequency of the courtship morphs in the population - females prefer the *less frequent* morph. Everything else is the same as above; males are assigned a trait value at random (it's not inherited), parental males' nests survive with a probability of 90% and non-parental males' nests survive with a probability of 10%, and courters and parental males are disfavored in viability selection. -->

<!-- ```{r parent_courtship_nfds,echo=FALSE} -->
<!-- pcn<-read.delim("parent-courter-conditional_nfds_summary.txt") -->
<!-- plot(pcn$CourterFreq,lwd=2,xlab="Generations",ylab="Frequency of Courtship Males", -->
<!--      bty="u",type="l",ylim=c(0,1),col=alpha(cols["courter"],0.75)) -->
<!-- points(pcn$ParentFreq,lwd=2,type="l",col=alpha(cols["parent"],0.75)) -->
<!-- mtext(text = "Frequency of Parental Males",side=4) -->
<!-- ``` -->

## Unlinked additive genetic variance

In these cases, the traits are encoded by a number (50) of loci, whose alleles contribute additively to determine the trait value. These alleles are all freely recombining and are not adhered to any physical genomic location (aka this is a classical quantitative genetics approach). The overall trait value is compared to a population-level threshold (which is static, in these cases), and if the trait value is above the threshold the male takes the parent or courter morph and if it is below he does not. Below, I'm showing the results from 10 replicates of each scenario. 

### Courter trait

Females choose nests based on whether the male is a courter or not, and they all prefer courters all of the time (the female preference does not have a genetic basis and does not evolve). The only way that non-courters produce offspring is through sneaking, but all males can be sneakers (both courters and non-courters). Because parental care is not incorporated in this model, all nests survive.

```{r unlinked_courter,out.extra='',fig.pos="H", fig.cap="Frequency of the courter morph (each color represents a different replicate)"}
plot.courter.reps<-function(pattern,cols,x.lim=c(0,12000)){
  courter.files<-list.files(pattern=pattern)
  #set up colors
  mycols <- colorRampPalette(c("grey",cols["courter"],"black"))(length(courter.files))[2:(1+length(courter.files))]
  #set up an empty plot
  plot(0:1,0:1,bty="L",xlab="Generations",ylab="Frequency of Courting Males",
       type='n',xlim=x.lim,ylim=c(0,1))
  #plot each run's output and save it
  s<-lapply(courter.files,function(file){
    summ<-suppressWarnings(read.delim(file))
    n<-as.numeric(gsub("\\w+.*_(\\d+)_\\w+.*","\\1",file))
    points(summ$Generation,summ$CourterFreq,col=alpha(mycols[n],0.5),
           lwd=2,type="l")
    return(summ)
  })
  text(x=10000,y=1.1,xpd=TRUE,"Equilibrium\nevaluated")
  clip(x1 = 0,x2=12000,y1=-0.1,y2=1.04)
  abline(v=10000,lty=2,lwd=2)
  return(s)
}
s<-plot.courter.reps(pattern="^courter-nogenetics_\\d+_summary.txt",cols)
```

Of the `r length(s)` replicates, `r length(unlist(lapply(s,nrow))[unlist(lapply(s,nrow))==10000])` reached an equilibrium by 10000 generations. 

```{r unlinked_courter_final}
get.courter.freqs<-function(s){
  courter.final.freqs<-do.call(rbind,lapply(s,function(dat){
    last<-dat[nrow(dat),]
    return(last[,c("Generation","CourterFreq","CourterW","NonCourterW")])
  }))
  rownames(courter.final.freqs)<-unlist(lapply(seq(1:nrow(courter.final.freqs)),function(n){paste("Rep",n,sep="")}))
  return(courter.final.freqs)
}
courter.final.freqs<-get.courter.freqs(s)
#aggregate the data
court.ws<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterW"],rep["NonCourterW"]))})
court.freqs<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterFreq"],1-rep["CourterFreq"]))})
court.eq<-apply(courter.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```
```{r courter_barplot,eval=FALSE,echo=FALSE,out.extra='',fig.pos="H",fig.cap="Fitnesses of each morph with frequncies in the bars. E below the bars means the rep reached equilibrium, N means it did"}
#plot the final fitnesses
b<-barplot(court.ws,beside=TRUE,col=c(alpha(cols["courter"],0.75),alpha("darkgrey",0.75)),ylim=c(0,5),
        names.arg = rep("",10),ylab="Average Fitness",border = NA)
text(x=b[1,],y=1,label=court.freqs[1,],srt=90)
text(x=b[2,],y=1,label=court.freqs[2,],srt=90)
text(x=seq(2,30,3),y=-0.15,label=court.eq,xpd=T)
text(x=-1,y=-0.175,label="Equilibrium:",xpd=T)
legend("topright",bty='n',col=c(alpha(cols["courter"],0.75),alpha("darkgrey",0.75)),
       c("Courter","Non-courter"),pch=15)
```


```{r courter_table}
final.freqs<-courter.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of courters in final generation",
      row.names = TRUE)
```



```{r courter_traits,eval=FALSE,echo=FALSE}
#Now let's look at the traits of the individuals in the final generation.
source("../../gwscaR/R/gwscaR_plot.R") #for the gwsca.vioplot function
#get the files
trait.files<-list.files(pattern="^courter-nogenetics_\\d+_traits.txt")
t<-lapply(trait.files,function(file){
  summ<-suppressWarnings(read.delim(file))
  return(summ)
})
par(mfrow=c(5,4),oma=c(2,2,2,2),mar=c(2,2,2,2))
lapply(t,function(traits){
  males<-traits[traits$Sex == "MALE",]
  females<-traits[traits$Sex == "FEMALE",]
  hist(males$CourtTrait,col=cols["courter"],border = cols["courter"],
       xlab="",ylab="",axes=FALSE,main="")
  axis(1,pos=0,xlim=c(0,1),at=seq(0,1,0.2))
  mtext("Courter Trait Value",1,line=1)
  axis(2,las=1,ylim=c(0,500),at=seq(0,500,100))
  mtext("Number of males",2,line=2.5)
  
  gwsca.vioplot(traits$LifetimeRS[traits$Sex=="MALE"],traits$LifetimeRS[traits$Sex=="FEMALE"],
          col=c(cols["courter"],cols["cpref"]),border=c(cols["courter"],cols["cpref"]),ylim=c(0,40),plot.axes = FALSE)
  axis(1,at=c(1,2),labels = c("Courter","Non-courter"),lty=0)
  axis(2,at=seq(0,40,10),las=1)
  mtext("Lifetime Reproductive Success",2,line=2)
  
})

```

### Parental trait

All females nest with parental males, so the only way non-parental males reproduce is through sneaking. Parental males provide care that allows nests to have a 90% chance of survival. The female preference does not have a genetic basis and does not evolve.

```{r parental_nogenetics, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of parent morph (each color represents a different replicate)"}
plot.parent.reps<-function(pattern,cols,x.lim=c(0,12000)){
  parent.files<-list.files(pattern=pattern)
  #set up colors
  mycols <- colorRampPalette(c("grey",cols["parent"],"black"))(length(parent.files))[2:(1+length(parent.files))]
  #set up an empty plot
  plot(0:1,0:1,bty="L",xlab="Generations",ylab="Frequency of Parental Males",
       type='n',xlim=x.lim,ylim=c(0,1))
  #plot each run's output and save it
  s<-lapply(parent.files,function(file){
    summ<-suppressWarnings(read.delim(file))
    n<-as.numeric(gsub("\\w+.*_(\\d+)_\\w+.*","\\1",file))
    points(summ$Generation,summ$ParentFreq,col=alpha(mycols[n],0.5),
           lwd=2,type="l")
    return(summ)
  })
  text(x=10000,y=1.1,xpd=TRUE,"Equilibrium\nevaluated")
  clip(x1 = 0,x2=12000,y1=-0.1,y2=1.04)
  abline(v=10000,lty=2,lwd=2)
  return(s)
}
s<-plot.parent.reps("^parent-nogenetics_\\d+_summary.txt",cols)
```

Of the `r length(s)` replicates, `r length(unlist(lapply(s,nrow))[unlist(lapply(s,nrow))==10000])` reached an equilibrium by 10000 generations. 

```{r unlinked_parent_final}
get.parent.freqs<-function(s){
  parent.final.freqs<-do.call(rbind,lapply(s,function(dat){
    last<-dat[nrow(dat),]
    return(last[,c("Generation","ParentFreq","ParentW","NonParentW")])
  }))
  rownames(parent.final.freqs)<-unlist(lapply(seq(1:nrow(parent.final.freqs)),function(n){paste("Rep",n,sep="")}))
  return(parent.final.freqs)
}
parent.final.freqs<-get.parent.freqs(s)
#aggregate the data
parent.ws<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentW"],rep["NonParentW"]))})
parent.freqs<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentFreq"],1-rep["ParentFreq"]))})
parent.eq<-apply(parent.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```
```{r parent_barplot,eval=FALSE,echo=FALSE,fig.cap="Fitnesses of each morph with frequncies in the bars. E below a set of bars means that replicate resulted in an equilibrium and N means it did not reach an equilibrium."}
#plot the final fitnesses
b<-barplot(parent.ws,beside=TRUE,col=c(alpha(cols["parent"],0.75),alpha("darkgrey",0.75)),ylim=c(0,5),
        names.arg = rep("",10),ylab="Average Fitness",border = NA)
text(x=b[1,],y=1,label=parent.freqs[1,],srt=90)
text(x=b[2,],y=1,label=parent.freqs[2,],srt=90)
text(x=seq(2,30,3),y=-0.15,label=parent.eq,xpd=T)
text(x=-1,y=-0.175,label="Equilibrium:",xpd=T)
legend("topright",bty='n',col=c(alpha(cols["parent"],0.75),alpha("darkgrey",0.75)),
       c("Parent","Non-parent"),pch=15)
```

```{r parent_table}
final.freqs<-parent.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of parents in final generation",
      row.names = TRUE)
```



### Courtship and Parental Traits

Females choose nests based on males' courtship trait (they all only nest with courting males, and the female preference does not have a genetic basis and does not evolve), and then the survival of the nest depends on whether the courting male is also a parental male. If the chosen male is a parental male, the nest has a 90% chance of survival. Otherwise, it only has a 10% chance. Non-courters only reproduce through sneaking.

```{r parentalCourtship_nogenetics, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the two morphs (courter = green, parent = blue)"}
plot.pc.reps<-function(pattern,cols,x.lim=c(0,12000)){
  pc.files<-list.files(pattern=pattern)
  #set up colors
  parent.cols <- colorRampPalette(c("grey",cols["parent"],"black"))(length(pc.files))[2:(1+length(pc.files))]
  courtr.cols <- colorRampPalette(c("grey",cols["courter"],"black"))(length(pc.files))[2:(1+length(pc.files))]
  #set up an empty plot
  par(mar=c(4,4,4,4))
  plot(0:1,0:1,bty="U",xlab="Generations",ylab="Frequency of Courter Males",
       type='n',xlim=x.lim,ylim=c(0,1))
  axis(4)
  mtext("Frequency of Parental Males",4,line=2)
  #plot each run's output and save it
  s<-lapply(pc.files,function(file){
    summ<-suppressWarnings(read.delim(file))
    n<-as.numeric(gsub("\\w+.*_(\\d+)_\\w+.*","\\1",file))
    points(summ$Generation,summ$CourterFreq,col=alpha(courtr.cols[n],0.5),
           lwd=2,type="l")
    points(summ$Generation,summ$ParentFreq,col=alpha(parent.cols[n],0.5),
           lwd=2,type="l")
    return(summ)
  })
  text(x=10000,y=1.1,xpd=TRUE,"Equilibrium\nevaluated")
  clip(x1 = 0,x2=12000,y1=-0.1,y2=1.09)
  abline(v=10000,lty=2,lwd=2)
  return(s)
}
s<-plot.pc.reps("^parent-courter-nogenetics_\\d+_summary.txt",cols)
```

The different runs have different outcomes. 

Let's look at the morph frequencies.


```{r courtParent_morphfreqs,echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the 4 morphs in each rep"}
plot.morphs.reps<-function(s,cols2){
  par(mfrow=c(2,5),mar=c(2,2,2,1),oma=c(2,2,2,2))
  lapply(s,function(summ){
    plot(0:1,0:1,bty="L",xlab="",ylab="",
       type='n',xlim=c(0,12000),ylim=c(0,1))
    points(summ$Generation, summ$FreqNcNp,type="l",col=cols2["NCNP"],lwd=2)
    points(summ$Generation, summ$FreqCNp,type="l",col=cols2["CNP"],lwd=2)
    points(summ$Generation, summ$FreqNcP,type="l",col=cols2["NCP"],lwd=2)
    points(summ$Generation, summ$FreqCP,type="l",col=cols2["CP"],lwd=2)
  })
  mtext("Frequency of Morphs",2,outer=TRUE,cex=0.8)
  mtext("Generations",1,outer=TRUE,cex=0.8)
  par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
  plot(0, 0, type='n', bty='n', xaxt='n', yaxt='n')
  legend("top",c("Courter/Parent","Non-Courter/Parent","Courter/Non-Parent","Non-courter/Non-parent"),
         col=cols2,bty='n',ncol = 4,lty=1,lwd=2)
}
plot.morphs.reps(s,cols2)
```

The second panel in the top row shows a run where the population crashed after 12 generations.

We can look at the final frequencies in a table as well:

```{r unlinked_courterParent_final}
get.morph.freqs<-function(s){
  pc.final.freqs<-do.call(rbind,lapply(s,function(dat){
    last<-dat[nrow(dat),]
    return(last[,11:14])
  }))
  rownames(pc.final.freqs)<-unlist(lapply(seq(1:nrow(pc.final.freqs)),function(n){paste("Rep",n,sep="")}))
  return(pc.final.freqs)
}
pc.final.freqs<-get.morph.freqs(s)
kable(pc.final.freqs,caption = "Frequency of morphs in final generation",
      row.names = TRUE)
```

Multiple morphs are maintained in `r nrow(pc.final.freqs)-length(unlist(apply(pc.final.freqs,1,grep,pattern="1")))` of the `r nrow(pc.final.freqs)` replicates, and those morphs contain either a parent or a courter. However, one of those reps with variation actually crashed (second panel, top row).

```{r courtParent_rs,echo=FALSE,eval=FALSE}
ymax<-round(max(pcng.sum$ParentW,pcng.sum$NonParentW,pcng.sum$CourterW,pcng.sum$NonCourterW)+1)
par(mar=c(4,4,1,4),mfrow=c(2,1))
plot(pcng.sum$ParentW,lwd=2,xlab="Generations",ylab="Parental Reproductive Success",
     bty="u",type="l",ylim = c(0,ymax),col=alpha(cols["parent"],0.75),las=1)
points(pcng.sum$NonParentW,lwd=2,lty=3,type="l",col=alpha(cols["parent"],0.75))
points(pcng.sum$NonParentW,lwd=2,lty=3,type="l",col=alpha("grey",0.75))
plot(pcng.sum$CourterW,xlab="Generations",ylab="Courter Reproductive Success",
     bty="u",lwd=2,type="l",col=alpha(cols["courter"],0.75),ylim = c(0,ymax))
points(pcng.sum$NonCourterW,lwd=2,lty=3,type="l",col=alpha(cols["courter"],0.75))
points(pcng.sum$NonCourterW,lwd=2,lty=3,type="l",col=alpha("grey",0.75))
legend("topright",bty='n',legend=c("Preferred","Not Preferred"),lty=c(1,2),col="darkgrey",lwd=2)
```


```{r courtParent_traitsLandscape,echo=FALSE,eval=FALSE}
pcng.traits<-read.delim("parent-courter-nogenetics_traits.txt")
pcng.traits$Morph<-factor(paste(pcng.traits$Courter,pcng.traits$Parent,sep=","))
rcols<-colorRampPalette(cols)(25)
plot(jitter(pcng.traits$Courter,1.5),jitter(pcng.traits$Parent,1.5),
     cex=((pcng.traits$LifetimeRS/max(pcng.traits$LifetimeRS))+1),xlim=c(-0.05,1.05),ylim=c(-0.05,1.05),
     axes=F,xlab="",ylab="",pch=19,col=alpha(cols2[pcng.traits$Morph],0.75),lwd=2)
text(x=c(0,1),y=-0.04,c("Non-Courter","Courter"),xpd=TRUE)
text(x=-0.1,y=c(0,1),c("Non-Parent","Parent"),xpd=TRUE,srt=90)

```


### Courtship + Negative Frequency Dependent Selection

In this case, the female preference is based on the frequency of the courtship morphs in the population - females prefer the *less frequent* morph. Everything else is the same as above; all nests survive since parental care is not incorporated in this model, and courters and parental males are disfavored in viability selection.

```{r courtship_nogenetics_nfds,echo=FALSE, fig.cap="Frequency of courter morph"}
cnng<-plot.courter.reps(pattern="^courter-nogenetics-nfds_\\d+_summary.txt",cols)
```
```{r courtship_nfds_unlinked_table,echo=FALSE}
cnng.freqs<-get.courter.freqs(cnng)
kable(cnng.freqs,caption = "Courter Frequncies with negative frequency dependent preferences",
      row.names = TRUE)
```

With negative frequency dependent preferences, both courters and non-courters are maintained in the population.

### Parenting + Negative Frequency Dependent Selection

In this case, the female preference is based on the frequency of the parental morphs in the population - females prefer the *less frequent* morph. Everything else is the same as above; nests given to parents have a 90% chance of survival and a 10% chance of survival if given to non-parents, and courters and parental males are disfavored in viability selection.

```{r parent_nogenetics_nfds,echo=FALSE, fig.cap="Frequency of parental morph"}
pnng<-plot.parent.reps(pattern="^parent-nogenetics-nfds_\\d+_summary.txt",cols)
```
```{r parent_nfds_unlinked_table, echo=FALSE}
pnng.freqs<-get.parent.freqs(pnng)
kable(pnng.freqs,caption = "Parent Frequncies with negative frequency dependent preferences",
      row.names = TRUE)
```

When negative frequency dependent preferences are acting on the parental traits, the population crashes. If we zoom in on the actual timeframes until population crashes, it becomes clear that in all cases when population crashes occur the frequency of parental males was high or had just been very high.

```{r replot_parentnfds,echo=FALSE, fig.cap="Frequency of parental morph (black dashed line is at frequency of 0.5)"}
pnng<-plot.parent.reps(pattern="^parent-nogenetics-nfds_\\d+_summary.txt",cols,x.lim=c(0,50))
abline(h=0.5,lty=2,lwd=2)
```

Thus, the population crashes are likely due to the very low reproductive potential of the non-parental males (they can contribute 4 sperm instead of the parental males' 8) and the fact that when parental males are more frequent, females prefer non-parental males, and nests left with non-parental males have a 10% survival rate. This low reproductive output from non-parental males means that the population size shrinks, and because the male trait is heritable, the frequency of the parental morph doesn't dip below 0.5 (or at least not often), so females continue to prefer non-parental males.

### Courtship + Parenting + Negative Frequency Dependent Selection

In this case, the female preference is based on the frequency of the courtship morphs in the population - females prefer the *less frequent* morph. Everything else is the same as above; parental males' nests survive with a probability of 90% and non-parental males' nests survive with a probability of 10%, and courters and parental males are disfavored in viability selection.

```{r unlinked_parent_courtship_nfds,echo=FALSE, fig.cap="Frequency of courter and parental males"}
pcnng<-plot.pc.reps(pattern="^parent-courter-nogenetics-nfds_\\d+_summary.txt",cols)
```
```{r unlinked_pcnfds_plotMorphs,echo=FALSE, fig.cap="Frequency of four morphs per replicate"}
plot.morphs.reps(pcnng,cols2)
```
```{r unlinked_pcnfds_table,echo=FALSE}
pcnng.freqs<-get.morph.freqs(pcnng)
kable(pcnng.freqs,caption="Morph Frequencies with negative frequency dependent preferences",
      row.names=TRUE)
```

When negative frequency dependent preferences are acting on the courtship traits and there are parent traits as well, the population crashes. 

```{r replot_pcnfds,echo=FALSE, fig.cap="Frequency of parental and courter traits (black dashed line is at frequency of 0.5)"}
par(mfrow=c(1,2),oma=c(2,2,1,2),mar=c(4,4,1,1))
pcnng<-plot.parent.reps(pattern="^parent-courter-nogenetics-nfds_\\d+_summary.txt",cols,x.lim=c(0,100))
abline(h=0.5,lty=2,lwd=2)
pcnng<-plot.courter.reps(pattern="^parent-courter-nogenetics-nfds_\\d+_summary.txt",cols,x.lim=c(0,100))
abline(h=0.5,lty=2,lwd=2)

```

It appears that a correlation arises between the two morphs, so that preferences for the courter traits result the same dynamics as seen when females directly prefer parental males (above).

### Courtship + Heritable Female Preferences

Here, the female preference has unlinked additive genetic variance that is inherited. The traits begin as uncorrelated and are not pleiotropic (i.e., they have different genes underlying them). A threshold is set in the first generation that determines the switch point for when females prefer courters or non-courters. This threshold does not change over the generations. All males can be sneakers, but the male who females choose to nest with get a first male advantage.

```{r courtship_pref_nogenetics,echo=FALSE, fig.cap="Courter frequencies"}
cpref<-plot.courter.reps(pattern="^courter-pref-nogenetics_\\d+_summary.txt",cols)
```
```{r courtship_pref_unlinked_table,echo=FALSE}
cpref.freqs<-get.courter.freqs(cpref)
kable(cpref.freqs,caption = "Courter Frequncies with heritable female preferences",
      row.names = TRUE)
```

Evolving female preferences allow variation to be maintained in `r nrow(cpref.freqs[cpref.freqs$CourterFreq<1,])` of `r nrow(cpref.freqs)` replicates.

### Parenting + Heritable Female Preferences

Here, the female preference has unlinked additive genetic variance that is inherited. The traits begin as uncorrelated and are not pleiotropic (i.e., they have different genes underlying them). A threshold is set in the first generation that determines the switch point for when females prefer parents or non-parents. This threshold does not change over the generations. Only non-parental males can be sneakers, and when females choose to mate with non-parents, the nest has a 10% survival rate.

```{r parent_pref_nogenetics,echo=FALSE, fig.cap="Parent frequencies"}
ppref<-plot.parent.reps(pattern="^parent-pref-nogenetics_\\d+_summary.txt",cols)
```
```{r parent_pref_unlinked_table, echo=FALSE}
ppref.freqs<-get.parent.freqs(ppref)
kable(ppref.freqs,caption = "Parent Frequncies with heritable female preferences",
      row.names = TRUE)
```

Evolving female preferences maintain variation, although parental traits can lead to population crashes (in `r nrow(ppref.freqs[ppref.freqs$Generation < 9999,])` of the `r nrow(ppref.freqs)` reps). In suriving populations, variation was maintained in `r nrow(ppref.freqs[ppref.freqs$ParentFreq<1,])` of `r nrow(ppref.freqs)` replicates.

### Courthip and Parenting + Heritable Female Preferences

Here, the female preference has unlinked additive genetic variance that is inherited. The traits begin as uncorrelated and are not pleiotropic (i.e., they have different genes underlying them). A threshold is set in the first generation that determines the switch point for when females prefer courters or non-courters. This threshold does not change over the generations. Only non-parental males are sneakers, and when females choose to mate with non-parents, the nest has a 10% survival rate.

```{r unlinked_parent_courtship_pref,echo=FALSE, fig.cap="Parent and courter frequencies"}
pcpref<-plot.pc.reps(pattern="^parent-courter-pref-nogenetics_\\d+_summary.txt",cols)
```
```{r unlinked_pc_pref_plotMorphs,echo=FALSE, fig.cap="Frequency of four morphs per rep"}
plot.morphs.reps(pcpref,cols2)
```
```{r unlinked_pc_pref_table,echo=FALSE}
pcpref.freqs<-get.morph.freqs(pcpref)
kable(pcpref.freqs,caption="Morph Frequencies with heritable preferences",
      row.names=TRUE)
```

Again, evolving preferences maintain variation in `r length(apply(pcpref.freqs,1,function(r){ length(r[r==0])})[apply(pcpref.freqs,1,function(r){ length(r[r==0])})<3])` of the `r nrow(pcpref.freqs)`. `r length(unlist(lapply(pcpref,nrow))[lapply(pcpref,nrow) < 9999])` of the `r length(pcpref)` populations crashed, though.

<!-- ## Linked additive genetic variance -->

<!-- ### Courter Trait -->

<!-- ```{r CourterTraitFiles,eval=FALSE} -->
<!-- ``` -->
<!-- ```{r PlotCourterTraits,eval=FALSE} -->
<!-- ``` -->
<!-- ```{r PlotCourterAlleles, eval=FALSE} -->
<!-- ``` -->

<!-- ### Parent Trait -->

<!-- ```{r ParentTrait, eval=FALSE} -->
<!-- ``` -->

<!-- ### Frequency-Dependent Preference -->

<!-- ```{r FrequencyDependentPreference, eval=FALSE} -->
<!-- ``` -->

