---
title: "Comparing basic frequency-dependent models without explicit genetic architectures"
author: "Sarah P. Flanagan"
date:  "`r format(Sys.time(), '%d %B, %Y')`"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    template: manuscript.latex
    toc: yes
    toc_depth: 2
spacing: singlespacing
fontsize: 11pt
capsize: normalsize
documentclass: article
footerdate: yes
graphics: yes
csl: evolution.csl
editor_options:
  chunk_output_type: console
---

The purpose of this model is to understand how genetic architectures of alternative reproductive tactics impact their maintenance in populations. I'm using an individual-based simulation model with different selection scenarios, types of alternative tactics, and genetic architectures (genome-wide additive genetic variance vs supergenes). To explore the parameter space, I'm also using a baseline analytical model without the genetic architectures to identify regions of parameter space where we expect multiple morphs to be maintained. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200,fig.height = 7,fig.width = 7)
knitr::opts_knit$set(root.dir='../results/') #change

```
```{r opts, echo = FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "../figs/"
)
```
```{r librarysetup, echo=FALSE,echo=FALSE,message=FALSE}
library(RColorBrewer)
library(scales)
library(sm)
library(kableExtra)
library(plotly)
library(dplyr)
library(vegan)
library(tidyr)
source("../R/freq_functions.R")
source("../R/formatting.R")
source("../morph_predictions/check_freqs.R")
source("../morph_predictions/morph_gens_ns.R")

```

```{r spftools}
if(length(system.file(package='spfTools'))<=1){
  devtools::install_github("https://github.com/spflanagan/spfTools/")
}
library(spfTools)
```


## Baseline analytical model overview

The next generation's frequencies are determined by the frequency of the morphs that survive to adulthood. To investigate equilibrial outcomes, I performed numerical analysis over 100 generations. These results were saved from across a range of initial morph frequencies, reproductive contributions for male morphs ($r_M$), sperm competition coefficients ($c$) and number of sneakers that can fertilize each nest ($num_{sneakers}$). Interactive plots of the diversity of the population at the end of 100 generations can be viewed as [a shiny app](https://spflanagan.shinyapps.io/morph_predictions/). 

```{r getMathModelResults}
morph_results<-readRDS("../morph_predictions/morph_results_Ns.RDS")
morph_results$diversity<-vegan::diversity(round(morph_results[,c("CP","CN","NP","NN")],4))
```


The shiny app shows the results for a variety of initial morph frequencies, but I also wanted to explore the relationships between the key parameters (the sperm competition coefficient, $c$, the relative reproductive allocation, $r$, and the number of sneakers/mates allowed ($num_{sneak}$)). We can see that each one plays an important role in facilitating diversity, with $r$ driving the most restrictive areas of parameter space (Fig. \@ref(fig:rVc), \@ref(fig:rVnsneak)). Nonetheless, $c$ and $num_{sneak}$ interact to create areas of high-diversity space (Fig. \@ref(fig:cVnsneak)), which is why it is important to include variation in all three.

```{r rVc, fig.cap="The diversity of the population for different values of the sperm competition coefficient (c) and the reproductive allocation (r). These results are assuming 4 mates allowed and equal initial frequencies of all four morphs, and are the frequencies after 100 generations."}

morph_equal<-morph_results[morph_results$initial_CP == 0.25 &
                             morph_results$initial_CN == 0.25 &
                             morph_results$initial_NP == 0.25 &
                             morph_results$initial_NN == 0.25 &
                             morph_results$num_sneak == 4 ,]
morph_equal$ident<-paste0(morph_equal$r,"_",morph_equal$c,"_",morph_equal$num_sneak)
data_wide <- tidyr::spread(
    morph_equal[,c("r","c","diversity")],
    r,
    diversity
  )
rownames(data_wide)<-data_wide$c
data_wide<-data_wide[,-1]
  
fig1<-plot_ly(
    x = as.numeric(colnames(data_wide)), 
    y = as.numeric(rownames(data_wide)), 
    z = as.matrix(data_wide),
    colorscale=list(seq(0,1,length.out = 9),
                    c('#ffffd9','#edf8b1','#c7e9b4','#7fcdbb','#41b6c4','#1d91c0','#225ea8','#253494','#081d58')),
    type = "contour"
  )
  # add axis labels
  x<-list(title="r")
  y<-list(title="c")
  fig1 <- fig1 %>% layout(xaxis=x,yaxis=y)
  # add label to contour names
  fig1 <- fig1 %>% colorbar(title = "Diversity of the population")
  
  fig1
  

```

```{r rVnsneak, fig.cap="The diversity of the population for different values of the reproductive allocation (r) and the number of mates allowed. These results are assuming a sperm competition coefficient (c) of 0.5 and equal initial frequencies of all four morphs, and are the frequencies after 100 generations."}
morph_equal<-morph_results[morph_results$initial_CP == 0.25 &
                             morph_results$initial_CN == 0.25 &
                             morph_results$initial_NP == 0.25 &
                             morph_results$initial_NN == 0.25 &
                             morph_results$c == 0.5 ,]

data_wide <- tidyr::spread(
    morph_equal[,c("r","num_sneak","diversity")],
    r,
    diversity
  )
rownames(data_wide)<-data_wide$num_sneak
data_wide<-data_wide[,-1]
  
fig1<-plot_ly(
    x = as.numeric(colnames(data_wide)), 
    y = as.numeric(rownames(data_wide)), 
    z = as.matrix(data_wide),
    colorscale=list(seq(0,1,length.out = 9),
                    c('#ffffd9','#edf8b1','#c7e9b4','#7fcdbb','#41b6c4','#1d91c0','#225ea8','#253494','#081d58')),
    type = "contour"
  )
  # add axis labels
  x<-list(title="r")
  y<-list(title="number of mates allowed")
  fig1 <- fig1 %>% layout(xaxis=x,yaxis=y)
  # add label to contour names
  fig1 <- fig1 %>% colorbar(title = "Diversity of the population")
  
  fig1

```


```{r cVnsneak, fig.cap="The diversity of the population for different values of the sperm competition coefficient (c) and the number of mates allowed. These results are assuming a reproductive allocation (r) of 0.5 and equal initial frequencies of all four morphs, and are the frequencies after 100 generations."}
morph_equal<-morph_results[morph_results$initial_CP == 0.25 &
                             morph_results$initial_CN == 0.25 &
                             morph_results$initial_NP == 0.25 &
                             morph_results$initial_NN == 0.25 &
                             morph_results$r == 0.5 ,]

data_wide <- tidyr::spread(
    morph_equal[,c("c","num_sneak","diversity")],
    c,
    diversity
  )
rownames(data_wide)<-data_wide$num_sneak
data_wide<-data_wide[,-1]
  
fig1<-plot_ly(
    x = as.numeric(colnames(data_wide)), 
    y = as.numeric(rownames(data_wide)), 
    z = as.matrix(data_wide),
    colorscale=list(seq(0,1,length.out = 9),
                    c('#ffffd9','#edf8b1','#c7e9b4','#7fcdbb','#41b6c4','#1d91c0','#225ea8','#253494','#081d58')),
    type = "contour"
  )
  # add axis labels
  x<-list(title="c")
  y<-list(title="number of mates allowed")
  fig1 <- fig1 %>% layout(xaxis=x,yaxis=y)
  # add label to contour names
  fig1 <- fig1 %>% colorbar(title = "Diversity of the population")
  
  fig1

```


Before selecting precise values, I ran the model with equal initial frequencies for all four morphs with 10,000 generations, to ensure that the diversity identified is not lost over time in the mathematical model. 

```{r multiGenContours, fig.cap="Contour plots showing the diversity of the populations given equal initial starting frequencies of all four morphs, the relative reproductive investments (x-axis), sperm competition coefficients (y-axis), and number of sneakers allowed per nest (1 through 5, plots going from top to bottom).", fig.width=4.5, fig.height=7, warning=FALSE,message=FALSE}

results<-readRDS("../morph_predictions/morph_results_10000_equalStart.RDS")
results$diversity<-vegan::diversity(round(results[,c("CP","CN","NP","NN")],4))

contours<-by(results, results$num_sneak, function(sub_calcs){
  
  
  data_wide<-tidyr::spread(
    sub_calcs[,c("r","c","diversity")],
    r,
    diversity
  )
  rownames(data_wide)<-data_wide[,1]
  data_wide<-data_wide[,-1]
  
  fig <- plot_ly(
    x = as.numeric(colnames(data_wide)), 
    y = as.numeric(rownames(data_wide)), 
    z = as.matrix(data_wide), 
    type = "contour"
  )
  # add axis labels
  x<-list(title="r")
  y<-list(title="c")
  fig <- fig %>% layout(xaxis=x,yaxis=y)
  # add label to contour names
  fig <- fig %>% colorbar(title = "Diversity")
  # plot
  return(fig)
})


multi_fig <- subplot(contours, nrows = 5, shareY = TRUE, shareX = TRUE)

multi_fig %>% layout(autosize = F, width = 450, height = 650)


```


## Comparing analytical model to simulation model results

For these comparisons, I chose a low diversity parameter space and a high diversity parameter space. In the simulation model, I explored the effects of mating system (extent of multiple mating by males), so those results are shown here. 

### Low diversity parameter space

The [results from the mathematical model](https://spflanagan.shinyapps.io/morph_predictions/) were inspected to identify parameter regions where multiple morphs were *not* maintained. I focused on parameter sets when morphs had equal initial frequencies, as the simulation model initiates populations with nearly equal frequencies of the morphs. 


Given these results taken together, I selected the low diversity parameters in Table \@ref(tab:lowDivParams) to implement in the simulation model with a single locus underpinning the courter and parent traits. 

```{r lowDivParams}
highDivParams<-data.frame(
  parameter = c("r (in shiny app)",
                "r_CP",
                "r_CN",
                "r_NP",
                "r_NN",
                "c",
                "number of sneakers"),
  value = c( 2, 
             8,
             8,
             4,
             4,
             0.5,
             2)
)
kable(highDivParams, "latex",booktab=TRUE,
      caption="Table of parameters selected to create a low-diversity population. All morphs started at equal frequencies.") %>% kableExtra::kable_styling(latex_options = "HOLD_position")

```

Based on the mathematical model, we expected the courter-parent morph to dominate the population (Table \@ref(tab:lowMathResults)), and that it would dominate the population rapidly (Fig. \@ref(fig:mathOverTimeLow)). The same patterns are observed in the simulation model results (Fig. \@ref(fig:lowDivPlot)).

```{r lowMathResults}
expected<-rbind(round(morph_results[morph_results$initial_CP == 0.25 & 
                      morph_results$initial_CN == 0.25 & 
                      morph_results$initial_NP == 0.25 & 
                      morph_results$initial_NN == 0.25 & 
                      morph_results$r==2 &
                      morph_results$c == 0.5 &
                      morph_results$num_sneak == 2,
                    c("CP","CN","NP","NN")],2),
                round(results[results$r==2 & 
                          results$c == 0.5 & 
                          results$num_sneak==2,
                        1:4],2))
rownames(expected)<-c("100 generations","10,000 generations")
kable(expected,
      "latex",booktab=TRUE,
      caption="Expected frequencies of the four morphs after 100 generations with the chosen high-diversity parameter settings.")  %>% kableExtra::kable_styling(latex_options = "HOLD_position")
```


```{r mathOverTimeLow, fig.cap="The frequencies of the morphs cycle over time in the mathematical model with the low diversity paramter set.", fig.height=4, fig.height=6}

gens<-100
freqs<-c(CP=0.25,CN=0.25,NP=0.25,NN=0.25)
output<-data.frame(matrix(nrow=gens+1,ncol=4))
colnames(output)<-c("CP","CN","NP","NN")
output[1,]<-freqs
for (i in 1:(gens+1)){
  output[i+1,]<-one_gen(freqs=output[i,],rs=c(8,8,4,4),c=0.5, max_sneakers=2)
  if(all(is.na(output[i+1,]))){
    #then the population has crashed
    output[(i+1):(gens+1),]<-c(0,0,0,0)
    break
    
  }
}
plot(output$NN,type='l',ylim=c(0,1), col=cols2["NCNP"],lwd=2,
     xlab="Generation",
     ylab="Morph Frequency",
     bty='L')
lines(output$NP,type='l',col=cols2["NCP"], lty=2,lwd=2)
lines(output$CP,type='l',col=cols2["CP"], lty=3,lwd=2)
lines(output$CN,type='l',col=cols2["CNP"],lty=4,lwd=2)
legend("right",c("courter-parent","noncourter-parent", "courter-nonparent","noncourter-nonparent"),
       col=cols2,lty=c(3,2,4,1),lwd=2,bty='n',cex=1.5)
```


```{r lowDiv}
lowDiv<-get.morph.freqs(plot.pc.reps(pattern="lowDiversity.*summary.txt",path="single_locus",cols,make.plot=FALSE))
```
```{r lowDivPlot, fig.cap="Final generation frequencies based on the single locus model with parameters that are expected to result in no diversity, based on the mathematical model. In 'monogamy' cases, the males were constrained to only accept eggs from one female into their nests, whereas under 'polygyny' males could mate with multiple females. In both cases females mated once. Parameter settings lablled with 'nm' were runs where females did not mate if they could not find a suitable mate (i.e., if no courters were available)."}
var_patts<-unique(gsub(".*lowDiversity_(.*)_\\d_summary.txt_\\d","\\1",rownames(lowDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)

par(mfrow=c(2,length(var_patts)/2),oma=c(3,1,4,1),mar=c(3,1,4,1),xpd=TRUE)
varDat<-lapply(var_patts,function(pattern){
  dat<-lowDiv[grep(paste0(pattern,"_\\d"),rownames(lowDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main=pattern)
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})
par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0,0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=15,xpd = TRUE,ncol=2)

```


### High diversity parameter space

The [results from the mathematical model](https://spflanagan.shinyapps.io/morph_predictions/) were inspected to identify parameter regions where multiple morphs were maintained. I focused on parameter sets when morphs had equal initial frequencies, as the simulation model initiates populations with nearly equal frequencies of the morphs. I selected the high-diversity parameters in Table \@ref(tab:highDivParams) to implement in the simulation model with a single locus underpinning the courter and parent traits. 

```{r highDivParams}
highDivParams<-data.frame(
  parameter = c("r (in shiny app)",
                "r_CP",
                "r_CN",
                "r_NP",
                "r_NN",
                "c",
                "number of sneakers"),
  value = c( 0.75, 
             6,
             6,
             8,
             8,
             0.7,
             2)
)
kable(highDivParams, "latex",booktab=TRUE,
      caption="Table of parameters selected to create a high-diversity population. All morphs started at equal frequencies.")  %>% kableExtra::kable_styling(latex_options = "HOLD_position")

```

Based on the mathematical model, we expected the courter-parent, noncourter-parent, and noncourter-nonparent morphs to be maintained at reasonably high frequencies (Table \@ref(tab:highMathResults)). Interestingly, the polymorphism maintained in the mathematical model seems to be maintained through cyclical shifting of morph frequencies (Fig. \@ref(fig:mathOverTime)).

```{r highMathResults}
expected<-rbind(round(morph_results[morph_results$initial_CP == 0.25 & 
                      morph_results$initial_CN == 0.25 & 
                      morph_results$initial_NP == 0.25 & 
                      morph_results$initial_NN == 0.25 & 
                      round(morph_results$r,1)==0.7 &
                      morph_results$c == 0.75 &
                      morph_results$num_sneak == 3,
                    c("CP","CN","NP","NN")],2),
                round(results[round(results$r,1)==0.7 & 
                          results$c == 0.75 & 
                          results$num_sneak==2,
                        1:4],2))
rownames(expected)<-c("100 generations","10,000 generations")
kable(expected,
      "latex",booktab=TRUE,
      caption="Expected frequencies of the four morphs after 100 generations with the chosen high-diversity parameter settings.")  %>% kableExtra::kable_styling(latex_options = "HOLD_position")
```

```{r mathOverTime, fig.cap="The frequencies of the morphs cycle over time in the mathematical model, with polymorphism being maintained by the dominant morph shifting every few generations.", fig.height=4, fig.width=6}
gens<-100
freqs<-c(CP=0.25,CN=0.25,NP=0.25,NN=0.25)
output<-data.frame(matrix(nrow=gens+1,ncol=4))
colnames(output)<-c("CP","CN","NP","NN")
output[1,]<-freqs
for (i in 1:(gens+1)){
  output[i+1,]<-one_gen(freqs=output[i,],rs=c(0.75*8,0.75*8,8,8),c=0.7, max_sneakers=2)
  if(all(is.na(output[i+1,]))){
    #then the population has crashed
    output[(i+1):(gens+1),]<-c(0,0,0,0)
    break
    
  }
}
plot(output$NN,type='l',ylim=c(0,1), col=cols2["NCNP"],lwd=2,
     xlab="Generation",
     ylab="Morph Frequency",
     bty='L')
lines(output$NP,type='l',col=cols2["NCP"], lty=2,lwd=2)
lines(output$CP,type='l',col=cols2["CP"], lty=3,lwd=2)
lines(output$CN,type='l',col=cols2["CNP"],lty=4,lwd=2)
legend("top",c("courter-parent","noncourter-parent", "courter-nonparent","noncourter-nonparent"),
       col=cols2,lty=c(3,2,4,1),lwd=2,ncol=4,bty='n')
```

The simulation model with the same parameter settings also resulted in polymorphism, but only when males were restricted to accepting eggs from a single female, and when random mating was suppressed (Fig. \@ref(fig:highDivPlot)). Another unexpected result was that few of the simulation runs resulted in the maintenance of all three predicted morphs. 


```{r highDiv}
highDiv<-get.morph.freqs(plot.pc.reps(pattern="highDiversity_.*summary.txt",path="single_locus",cols,make.plot=FALSE))
```
```{r highDivPlot, fig.cap="Final generation frequencies based on the single locus model with parameters that are expected to result in diversity, based on the mathematical model. In 'monogamy' cases, the males were constrained to only accept eggs from one female into their nests. Under 'polygyny' males accepted eggs from multiple females but were still constrained by the number of offspring they could fertilize. Parameter settings lablled with 'nm' were runs where females did not mate if they could not find a suitable mate (i.e., if no courters were available).", fig.height=8,fig.width=8}
var_patts<-unique(gsub(".*highDiversity_(.*)_\\d_summary.txt_\\d","\\1",rownames(highDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)

par(mfrow=c(2,length(var_patts)/2),oma=c(3,1,4,1),mar=c(3,1,4,1),xpd=TRUE)
varDat<-lapply(var_patts,function(pattern){
  dat<-highDiv[grep(paste0(pattern,"_\\d"),rownames(highDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main=pattern)
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})
par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0,0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=15,xpd = TRUE,ncol=2,cex=1.25)

```


### Individual-level results from simulation models

We can understand what is driving these unexpected results from simulation model -- and ensure they are aligned with the mathematical model -- by looking at individual-level patterns of mating success and reproductive success in males and females in the initial generation and in the final generation for one of the replicates with monogamy and polygyny.

```{r}
traits<-read.delim("single_locus/pcu_1locus_highDiversity_polygyny_nm_1_traits.txt")
traits_mono<-read.delim("single_locus/pcu_1locus_highDiversity_monogamy_nm_1_traits.txt")
```

```{r make_trait_hists}
make_trait_hists<-function(traits,which_trait,which_sex,which_gen,which_pop,cols){
  these_traits<-traits[traits$Gen==which_gen & 
                         traits$Pop==which_pop &
                         traits$Sex==which_sex,]
  if(which_sex=="MALE"){
    ylabel<-"Number of males"
  } else{
    ylabel<-"Number of females"
  }
  cp<-these_traits[these_traits$Courter==1 & these_traits$Parent==1,which_trait]
  cn<-these_traits[these_traits$Courter==1 & these_traits$Parent==0,which_trait]
  np<-these_traits[these_traits$Courter==0 & these_traits$Parent==1,which_trait]
  nn<-these_traits[these_traits$Courter==0 & these_traits$Parent==0,which_trait]
  
  num<-sum(c(length(cp)>0,length(cn)>0,length(np)>0,length(nn)>0))
  bks<-seq(-0.5,max(these_traits[,which_trait])+0.5,1)
  if(num>2){
    par(mfrow=c(2,2))
  } else if(num==2){
    par(mfrow=c(1,2))
  } else if(num==1){
    par(mfrow=c(1,1))
  }
  
  if(length(cp)>0){
    hist(cp,
     col = cols["CP"], 
     main="Courter-Parents",
     xlab=which_trait,
     ylab=ylabel,
     breaks=bks)
  }
  if(length(cn)>0){
    hist(cn,
     col = cols["CNP"], 
     main="Courter-Nonparents",
     xlab=which_trait,
     ylab=ylabel,
     breaks=bks)
  }
  if(length(np)>0){
    hist(np,
     col = cols["NCP"], 
     main="Noncourter-Parents",
     xlab=which_trait,
     ylab=ylabel,
     breaks=bks)
  }
  if(length(nn)>0){
    hist(nn,
     col = cols["NCNP"], 
     main="Nonourter-Nonparents",
     xlab=which_trait,
     ylab=ylabel,
     breaks=bks)
  }
}
```



#### Individual level results: monogamy


```{r maleMS0mono, fig.cap="Male mating success in generation 0 for each morph under monogamy (with no mating allowed)."}
make_trait_hists(traits_mono,
                 which_trait="MateFound",
                 which_sex="MALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)

```

```{r maleRS0mono, fig.cap="Male lifetime reproductive success in generation 0 for each morph under monogamy (with no mating allowed)."}
make_trait_hists(traits_mono,
                 which_trait="LifetimeRS",
                 which_sex="MALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)
```

```{r femaleMS0mono, fig.cap="Female mating success in generation 0 for each morph under monogamy (with no mating allowed)."}
make_trait_hists(traits_mono,
                 which_trait="MateFound",
                 which_sex="FEMALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)

```

```{r femaleRS0mono, fig.cap="Female lifetime reproductive success in generation 0 for each morph under monogamy (with no mating allowed)."}
make_trait_hists(traits_mono,
                 which_trait="LifetimeRS",
                 which_sex="FEMALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)
```

By the final generation, only two morphs remain, and as expected courter-parents receive all of the matings and noncourter-nonparents receive no mates (Fig. \@ref(fig:maleMS12000)), but both have net-positive lifetime reproductive success, with noncourter-nonparents having more variable numbers of offspring (Fig. \@ref(fig:maleRS12000)). Females of the two morphs both have similar problems finding mates (Fig. \@ref(fig:femaleMS12000)) but have similar distributions of lifetime reproductive success (Fig. \@ref(fig:femaleRS12000)). Combined, these results suggest that the model is behaving as expected, and in a manner similar to the mathematical model.
 
```{r maleMS12000mono, fig.cap="Male mating success in generation 12000 for each morph under monogamy (with no mating allowed).", fig.height=4}
make_trait_hists(traits_mono,
                 which_trait="MateFound",
                 which_sex="MALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)

```

```{r maleRS12000mono, fig.cap="Male lifetime reproductive success in generation 12000 for each morph under monogamy (with no mating allowed).", fig.height=4}
make_trait_hists(traits_mono,
                 which_trait="LifetimeRS",
                 which_sex="MALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)
```

```{r femaleMS12000mono, fig.cap="Female mating success in generation 12000 for each morph under monogamy (with no mating allowed).", fig.height=4}
make_trait_hists(traits_mono,
                 which_trait="MateFound",
                 which_sex="FEMALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)

```

```{r femaleRS12000mono, fig.cap="Female lifetime reproductive success in generation 12000 for each morph under monogamy (with no mating allowed).", fig.height=4}
make_trait_hists(traits_mono,
                 which_trait="LifetimeRS",
                 which_sex="FEMALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)
```


#### Individual level results: Polygyny

As expected, the non-courters do not mate (Fig. \@ref(fig:maleMS0poly)), and the courter-nonparents do not have any offspring after generation 1 (Fig. \@ref(fig:maleRS0poly)), whereas females have even mating success (Fig. \@ref(fig:femaleMS0poly)) and similar lifetime reproductive success Fig. \@ref(fig:femaleRS0poly)).

```{r maleMS0poly, fig.cap="Male mating success in generation 0 for each morph under polygyny (with no mating allowed)."}
make_trait_hists(traits,
                 which_trait="MateFound",
                 which_sex="MALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)

```

```{r maleRS0poly, fig.cap="Male lifetime reproductive success in generation 0 for each morph under polygyny (with no mating allowed)."}
make_trait_hists(traits,
                 which_trait="LifetimeRS",
                 which_sex="MALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)
```

```{r femaleMS0poly, fig.cap="Female mating success in generation 0 for each morph under polygyny (with no mating allowed)."}
make_trait_hists(traits,
                 which_trait="MateFound",
                 which_sex="FEMALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)

```

```{r femaleRS0poly, fig.cap="Female lifetime reproductive success in generation 0 for each morph under polygyny (with no mating allowed)."}
make_trait_hists(traits,
                 which_trait="LifetimeRS",
                 which_sex="FEMALE",
                 which_gen=0,
                 which_pop=0,
                 cols=cols2)
```

By the final generation, only two morphs remain, and as expected courter-parents receive all of the matings and noncourter-nonparents receive no mates (Fig. \@ref(fig:maleMS12000)), but both have net-positive lifetime reproductive success, with noncourter-nonparents having more variable numbers of offspring (Fig. \@ref(fig:maleRS12000)). Females of the two morphs both have similar problems finding mates (Fig. \@ref(fig:femaleMS12000)) but have similar distributions of lifetime reproductive success (Fig. \@ref(fig:femaleRS12000)). Combined, these results suggest that the model is behaving as expected, and in a manner similar to the mathematical model.
 
```{r maleMS12000, fig.cap="Male mating success in generation 12000 for each morph.", fig.height=4}
make_trait_hists(traits,
                 which_trait="MateFound",
                 which_sex="MALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)

```

```{r maleRS12000, fig.cap="Male lifetime reproductive success in generation 12000 for each morph.", fig.height=4}
make_trait_hists(traits,
                 which_trait="LifetimeRS",
                 which_sex="MALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)
```

```{r femaleMS12000, fig.cap="Female mating success in generation 12000 for each morph.", fig.height=4}
make_trait_hists(traits,
                 which_trait="MateFound",
                 which_sex="FEMALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)

```

```{r femaleRS12000, fig.cap="Female lifetime reproductive success in generation 12000 for each morph.", fig.height=4}
make_trait_hists(traits,
                 which_trait="LifetimeRS",
                 which_sex="FEMALE",
                 which_gen=12000,
                 which_pop=0,
                 cols=cols2)
```



What these results suggest is that under polygyny, the density-dependent nature of the model allows males to escape the costs of courtship, because even though they 'use up' all of their sperm, offspring are created in proportion to the surviving offspring, and those males are able to escape the costs and therefore increase their reproductive success beyond what is possible for males in the mathematical model. If we impose stricter costs on the males ($r=0.25$, or $r_{courter}=2$ and $r_{non-courter}=8$), we can see that multiple morphs are maintained under polygyny (with non-random mating, Fig. \@ref(fig:highDivStrictPlot)).

```{r highDivStrict}
highDiv<-get.morph.freqs(plot.pc.reps(pattern="highDiversityStrict.*summary.txt",path="single_locus",cols,make.plot=FALSE))
```
```{r highDivStrictPlot, fig.cap="Final generation frequencies based on the single locus model with parameters that are expected to result in diversity, based on the mathematical model. In 'monogamy' cases, the males were constrained to only accept eggs from one female into their nests. Under 'polygyny' males accepted eggs from multiple females but were still constrained by the number of offspring they could fertilize. Parameter settings lablled with 'nm' were runs where females did not mate if they could not find a suitable mate (i.e., if no courters were available).", fig.height=8,fig.width=8}
var_patts<-unique(gsub(".*highDiversityStrict_(.*)_\\d_summary.txt_\\d","\\1",rownames(highDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)

par(mfrow=c(2,length(var_patts)/2),oma=c(3,1,4,1),mar=c(3,1,4,1),xpd=TRUE)
varDat<-lapply(var_patts,function(pattern){
  dat<-highDiv[grep(paste0(pattern,"_\\d"),rownames(highDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main=pattern)
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})
par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0,0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=15,xpd = TRUE,ncol=2,cex=1.25)

```


#### Identifying and clarifying differences between the two models

One element of the simulation model is that the frequencies of the morphs do not start of at exactly equal frequencies. If we use the initial conditions from the simulation model to predict the equilibrial frequencies, we can see that the expected outcomes shift, and in some cases have a better qualitative match to the simulated outcomes (Table \@ref(tab:actualInitialFreqCals))

```{r getPolyHighGens}
polyHigh<-plot.pc.reps(pattern="highDiversity_polygyny_nm_\\d.*summary.txt",
                            path="single_locus",cols,make.plot=FALSE)
```

```{r actualInitialFreqCals}
freq0<-do.call(rbind,lapply(polyHigh,function(x,morph_results){
  gen0<-x[1,c("FreqCP","FreqCNp","FreqNcP","FreqNcNp")]
  expected<-morph_gens_ns(gens=100,
                         freqs=c(
                           CP =  round(gen0$FreqCP,2),
                           CN = round(gen0$FreqCNp,2),
                           NP = round(gen0$FreqNcP,2),
                           NN = round(gen0$FreqNcNp,2) 
                         ),
                         rs=c(0.75*8,0.75*8,8,8),
                         c=0.7, 
                         max_sneakers=2
                         )
  output<-cbind(gen0,expected, 
                x[nrow(x),c("FreqCP","FreqCNp","FreqNcP","FreqNcNp")])
  colnames(output)<-paste0(colnames(output),
                           c(rep("_initial",4),
                             rep("_expected",4),
                              rep("_observed",4)))
  return(output)
}, morph_results=morph_results))

rownames(freq0)<-gsub("^(.*_nm)_(\\d)_summary.txt_(\\d)$",
                      "Rep\\2_Pop\\3",
                      rownames(freq0))

kable(round(freq0,3),"latex",booktab=TRUE,
      caption="Initial allele frequencies from each of the simulations alongside the expected frequency, based on those starting conditions, from the mathematical model (after 100 generations).",
      col.names = gsub("(\\w+)_\\w+","\\1",colnames(freq0))) %>%
  add_header_above(c(" " = 1, "Actual initial" = 4,
                     "Expected final" = 4, "Actual final" = 4)) %>%
  landscape()
```


As we saw above, the mathematical model showed cyclic patterns (Fig. \@ref(fig:mathOverTime)). Interestingly, the same cycling pattern does not seem to occur in the dynamics of the simulation model (Fig. \@ref(fig:polyNMoverTime)).

```{r polyNMoverTime, fig.cap="Plots of the morph frequencies over time for the high diversity parameter set. Shown are the five replicate runs (each with 4 populations that started with identical initial conditions) of the polygyny version of the model, where males can mate with multiple females, when females are constrained to only mate if they find an acceptable mate. Each row is a replicate and each column is a population in the replicate."}

plot.morphs.reps(polyHigh,cols2=cols2, ncols=4)

```

That being said, if we run the mathematical model for a longer period of time (say, 1000 generations), the frequency of the noncourter-parents drifts to 0 and that causes the eventual fixation of the courter-parent morph (Fig. \@ref(fig:mathOverTime2)) -- suggesting that perhaps maintenance of variation isn't as expected as we thought. I am re-running the mathematical model predictions to see whether there are other regions of parameter space in which multiple morphs are maintained over a longer period of time (10,000 generations).

```{r mathOverTime2, fig.cap="The frequencies of the morphs cycle over time in the mathematical model, with polymorphism being maintained by the dominant morph shifting every few generations. In this case, the model was run for 1000 generations rather than 100.", fig.height=4}
gens<-1000
freqs<-c(CP=0.25,CN=0.25,NP=0.25,NN=0.25)
output<-data.frame(matrix(nrow=gens+1,ncol=4))
colnames(output)<-c("CP","CN","NP","NN")
output[1,]<-freqs
for (i in 1:(gens+1)){
  output[i+1,]<-one_gen(freqs=output[i,],rs=c(0.75*8,0.75*8,8,8),c=0.7, max_sneakers=2)
  if(all(is.na(output[i+1,]))){
    #then the population has crashed
    output[(i+1):(gens+1),]<-c(0,0,0,0)
    break
    
  }
}
plot(output$NN,type='l',ylim=c(0,1), col=cols2["NCNP"],lwd=2,
     xlab="Generation",
     ylab="Morph Frequency",
     bty='L')
lines(output$NP,type='l',col=cols2["NCP"], lty=2,lwd=2)
lines(output$CP,type='l',col=cols2["CP"], lty=3,lwd=2)
lines(output$CN,type='l',col=cols2["CNP"],lty=4,lwd=2)
legend("right",c("courter-parent","noncourter-parent", "courter-nonparent","noncourter-nonparent"),
       col=cols2,lty=c(3,2,4,1),lwd=2,ncol=1,bty='n')
```


Some other important differences between the simulation model and the mathematical model that could be underlying any differences between the two are that the simulation model includes recombination (and inheritance of traits via females) whereas the mathematical model does not. Additionally, the simulation model has a reasonable amount of stochasticity embedded within it, which the mathematical model does not include -- this is especially true for the parameter settings when females randomly mate if they do not find an acceptable male. 



