---
title: "Density-dependent ARTs Analysis: Initial results"
author: "Sarah P. Flanagan"
date:  "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: true
    dev: png
    fig_caption: true
    
  html_document: 
    toc: true  
header-includes: 
  \usepackage{float}
  \usepackage{caption}
  \captionsetup[figure]{labelfont=it,textfont=it}
  \captionsetup[table]{labelfont=it,textfont=it}
always_allow_html: yes
editor_options: 
  chunk_output_type: console
---

<style>
p.caption {
  font-size: 0.85em;
  font-style: italic;
  color: grey;
}
</style>

The purpose of this model is to understand how genetic architectures of alternative reproductive tactics impact their maintenance in populations. I'm using an individual-based simulation model with different selection scenarios, types of alternative tactics, and genetic architectures (genome-wide additive genetic variance, supergenes, expression networks).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200)
knitr::opts_knit$set(root.dir='../results/') #change
```
```{r opts, echo = FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "figs/"
)
```
```{r, echo=FALSE}
library(RColorBrewer)
library(scales)
source("../R/002_freq_functions.R")
cols<-c(courter="#33a02c",parent="#1f78b4",cpref="#b2df8a",ppref="#a6cee3")
cols2<-c(CP='#a1dab4',NCP='#41b6c4',CNP='#2c7fb8',NCNP='#253494')
```

## Overview of the model 

Males can be courters or not-courters and parents or not-parents. When the model is run with both traits, this results in four possible morphs: courter/parent, courter/not-parent, not-courter/parent, and not-courter/not-parent. Generations are non-overlapping and there is one reproductive bout per generation. In each generation, the population follows the following timeline: 

![Flowchart of simulation](../figs/density-dependent-flowchart.jpg)

Each step is explained in further detail below. 

\newpage

### 1. Choosing a nest

A female samples 50 males and chooses a male to nest with based on his courtship trait. If there are no courtship traits in the model, she chooses based on the male's parental trait. If she does not encouter an acceptable male, she nests with a randomly chosen male. If she encounters multiple equally-acceptable males, she randomly selects one of them.  

### 2. Fertilization

Once a female decides to nest, up to three males can fertilize the nest. Courters and parental males can contribute more sperm than non-courter and non-parental males: $r_{courter}=r_{parent}=8$ and $r_{non-courter}=r_{non-parent}=4$. A courter/non-parent has $r_{non-parent}$ and a non-courter/parent has $r_{parent}$. The male with whom the female is nesting gets $r_{parent}/\Sigma{n_{sperm}}$ and additional non-parental males (up to 2) get $(r_{non-parent}*0.5/\Sigma{n_{sperm}})$, where $\Sigma{n_{sperm}}$ is the total number of sperm contributed by all of the males, weighted by the sperm competition factor (0.5 is the default for all males except the nesting male). So, when a female mates with one courter and two non-parentals, $\Sigma{n_{sperm}}$ = $r_{courter}$ + 2$*$(0.5$*$$r_{sneaker}$), where $r_{courter}$ = 8 and $r_{sneaker}$ = 4, therefore $\Sigma{n_{sperm}}$ = `r sum(8,rep((0.5*4),2))`. 

That being said, every time a male mates he uses his sperm, so after one mating where a courter fertilizes 50% of the female's 4 eggs, he only has 6 sperm for his next mating. 

In these runs, males who are courters cannot sneak fertilizations.

### 3. Nest Survival

Before the offspring can survive, the nest has to survive. This step is only relevant when parental traits are in the model - if only the courtship trait is specified, then all progeny in the nest survive at this point. When males have the parental trait, if the female has given eggs to a non-parental male (because she chose based on courtship traits), then the nest has a 10% chance of surviving. If the female has given eggs to a parental male, the nest has a 90% chance of surviving. 

### 4. Density-dependent survival

Once the identities of the surviving nests are known--along with the identities of the parents, including sneaker parents--the offspring are created in the simulation. The offspring are created to fill the full carrying capacity of the population, with the numbers per nest being equally divided among all surviving offspring. This allows the population to maintain a constant population size (prior to viability selection) while maintaining the relative successes of the given genotypes/morphs. 

### 5. Viability selection 

Before becoming adults, the offspring experience viability selection. Courters and parental males are disfavored in viability selection, with a survival probability of `r exp(-0.5/(2*50))`. If an individual is both a courter and a parental male, the survival probability is `r exp(-0.5/(2*50))*exp(-0.5/(2*50))`. Non-courters and non-parental males have survival probabilities of 1. 

### Evaluating equilibrium

After 10000 generations, I begin tracking the change in frequency of the courter and parent traits, and do so for 2000 additional generations. I calculate the variance in the change in frequency over those 2000 generations. I declare an equilibrium ('stasis') has been reached if the last change in frequency of both traits is less than the variance in changes in frequency. 

### Replicates

I ran five independent instances of the model, each of which had four replicates spawned from identical starting conditions. This approach was chosen to reduce the effects of stochasticity in this type of model. In the multi-panel figures below, each row contains the four replicates that originated with identical starting conditions. 

## Unlinked additive genetic variance

In these cases, the traits are encoded by a number (50) of loci, whose alleles contribute additively to determine the trait value. These alleles are all freely recombining and are not adhered to any physical genomic location (aka this is a classical quantitative genetics approach). The overall trait value is compared to a population-level threshold (which is static, in these cases), and if the trait value is above the threshold the male takes the parent or courter morph and if it is below he does not. Below, I'm showing the results from 10 replicates of each scenario. 

### Courter trait

Females choose nests based on whether the male is a courter or not, and they all prefer courters all of the time (the female preference does not have a genetic basis and does not evolve). The only way that non-courters produce offspring is through sneaking, but all males can be sneakers (both courters and non-courters). Because parental care is not incorporated in this model, all nests survive.


```{r unlinked_courter,out.extra='',fig.cap="Frequency of the courter morph with unlinked QTLs (each line represents a different replicate)"}
cu<-plot.courter.reps(pattern="^courter_unlinked_novs_[1-5]_summary.txt",cols)
```

Of the `r length(cu)` replicates, `r length(cu)-length(unlist(lapply(cu,nrow))[unlist(lapply(cu,nrow))<9999])` reached a steady state by 10000 generations. The different runs all overlap, suggesting that they all converge on the same outcome: all males are courters. This result is also obvious by inspecting a table of courter frequencies and fitnesses in the final generation of the runs.


```{r unlinked_courter_final}
courter.final.freqs<-get.courter.freqs(cu)
#aggregate the data
court.ws<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterW"],rep["NonCourterW"]))})
court.freqs<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterFreq"],1-rep["CourterFreq"]))})
court.eq<-apply(courter.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```
```{r courter_barplot,eval=FALSE,echo=FALSE,out.extra='',fig.pos="H",fig.cap="Fitnesses of each morph with frequncies in the bars. E below the bars means the rep reached equilibrium, N means it did"}
#plot the final fitnesses
b<-barplot(court.ws,beside=TRUE,col=c(alpha(cols["courter"],0.75),alpha("darkgrey",0.75)),ylim=c(0,5),
        names.arg = rep("",10),ylab="Average Fitness",border = NA)
text(x=b[1,],y=1,label=court.freqs[1,],srt=90)
text(x=b[2,],y=1,label=court.freqs[2,],srt=90)
text(x=seq(2,30,3),y=-0.15,label=court.eq,xpd=T)
text(x=-1,y=-0.175,label="Equilibrium:",xpd=T)
legend("topright",bty='n',col=c(alpha(cols["courter"],0.75),alpha("darkgrey",0.75)),
       c("Courter","Non-courter"),pch=15)
```


```{r courter_table}
final.freqs<-courter.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of courters in final generation of runs with unlinked additive genetic variances",
      row.names = TRUE)
```



```{r,fig.pos="H",fig.cap="Evolution of courtship trait with unlinked loci. Green dotted lines are the standard deviations around the mean and the black dotted line is the threshold",fig.width=6,fig.height=10,eval=FALSE,echo=FALSE}
#We may also be interested in the evolution of the courter trait values, which reflects the additive effects of all loci. 

par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(cu,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$CourterAEmean,type="l",col=cols["courter"],
       ylim=c(min(rep.summ$CourterAEmean-rep.summ$CourterAEsd)-0.05,
              max(rep.summ$CourterAEmean+rep.summ$CourterAEsd)+0.05))
  lines(rep.summ$Generation,rep.summ$CourterAEmean+rep.summ$CourterAEsd,type="l",col=alpha(cols["courter"],0.5),lty=3)
  lines(rep.summ$Generation,rep.summ$CourterAEmean-rep.summ$CourterAEsd,type="l",col=alpha(cols["courter"],0.5),lty=3)
  abline(h=rep.summ$CourterThresh[1],lty=2,col="black")
})
```


### Parental trait

In this case, there are no courtship traits so females lay eggs in nests randomly. Non-parental males will sneak fertilizations in parental males' nests. Parental males provide care that allows nests to have a 90% chance of survival. The female preference does not have a genetic basis and does not evolve.

```{r parental_nogenetics, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of parent morph with unlinked loci (each color represents a different replicate)"}
pu<-plot.parent.reps("^parent_unlinked_novs_[1-5]_summary.txt",cols)
```

```{r unlinked_parent_final}
parent.final.freqs<-get.parent.freqs(pu)
#aggregate the data
parent.ws<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentW"],rep["NonParentW"]))})
parent.freqs<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentFreq"],1-rep["ParentFreq"]))})
parent.eq<-apply(parent.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```


All of the populations resulted in the parentage trait going to fixation, as can be also seen in a table of the final frequencies and fitnesses in each replicate: 

```{r parent_table}
final.freqs<-parent.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of parents in final generation",
      row.names = TRUE)
```


```{r parent_nogenetics_traits ,out.extra='',fig.pos="H",fig.cap="Evolution of parenting trait values with additive genetic variation. Blue dotted lines are the standard deviations around the mean and the black dotted line is the threshold",fig.width=6,fig.height=10,echo=FALSE,eval=FALSE}
#Let's take a look at how the parent trait values evolve
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(pu,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$ParentAEmean,type="l",col=cols["parent"],
       ylim=c(min(rep.summ$ParentAEmean-rep.summ$ParentAEsd)-0.05,
              max(rep.summ$ParentAEmean+rep.summ$ParentAEsd)+0.05))
  lines(rep.summ$Generation,rep.summ$ParentAEmean+rep.summ$ParentAEsd,type="l",col=alpha(cols["parent"],0.5),lty=3)
  lines(rep.summ$Generation,rep.summ$ParentAEmean-rep.summ$ParentAEsd,type="l",col=alpha(cols["parent"],0.5),lty=3)
  abline(h=rep.summ$CourterThresh[1],lty=2,col="black")
})
```

### Courtship and Parental Traits

Females choose nests based on males' courtship trait (they all only nest with courting males, and the female preference does not have a genetic basis and does not evolve), and then the survival of the nest depends on whether the courting male is also a parental male. If the chosen male is a parental male, the nest has a 90% chance of survival. Otherwise, it only has a 10% chance. Non-courters and non-parents can reproduce through sneaking.

```{r parentalCourtship_nogenetics, echo=FALSE,out.extra='',fig.keep=FALSE}
u<-plot.pc.reps("^parent-courter_unlinked_novs_[1-5]_summary.txt",cols,make.plot=FALSE)
```

Let's look at the morph frequencies for each run:


```{r courtParent_morphfreqs,echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the 4 morphs in each rep with unlinked genetic variation",fig.width=6,fig.height=8}
plot.morphs.reps(u,cols2,ncols = 4)
```


```{r unlinked_courterParent_final}
pc.final.freqs<-get.morph.freqs(u)
rownames(pc.final.freqs)<-NULL
kable(pc.final.freqs[,c(1,grep("Freq",colnames(pc.final.freqs)))],caption = "Frequency of morphs in final generation",
      row.names = TRUE)
```

Multiple morphs are maintained in `r length(apply(pc.final.freqs,1,function(r){ length(r[r==0])})[apply(pc.final.freqs,1,function(r){ length(r[r==0])})<3])` of the `r nrow(pc.final.freqs)` replicates. 


```{r courtParent_rs,echo=FALSE,eval=FALSE}
ymax<-round(max(pcng.sum$ParentW,pcng.sum$NonParentW,pcng.sum$CourterW,pcng.sum$NonCourterW)+1)
par(mar=c(4,4,1,4),mfrow=c(2,1))
plot(pcng.sum$ParentW,lwd=2,xlab="Generations",ylab="Parental Reproductive Success",
     bty="u",type="l",ylim = c(0,ymax),col=alpha(cols["parent"],0.75),las=1)
points(pcng.sum$NonParentW,lwd=2,lty=3,type="l",col=alpha(cols["parent"],0.75))
points(pcng.sum$NonParentW,lwd=2,lty=3,type="l",col=alpha("grey",0.75))
plot(pcng.sum$CourterW,xlab="Generations",ylab="Courter Reproductive Success",
     bty="u",lwd=2,type="l",col=alpha(cols["courter"],0.75),ylim = c(0,ymax))
points(pcng.sum$NonCourterW,lwd=2,lty=3,type="l",col=alpha(cols["courter"],0.75))
points(pcng.sum$NonCourterW,lwd=2,lty=3,type="l",col=alpha("grey",0.75))
legend("topright",bty='n',legend=c("Preferred","Not Preferred"),lty=c(1,2),col="darkgrey",lwd=2)
```

If we look at how the two traits evolved, it's interesting to note that the same outcomes can emerge from different averages relative to the thresholds. It appears that the courtship trait always evolves above the threshold but the parentage trait can evolve below (resulting in the courter/non-parent morph) or above (resulting in the courter/parent morph) the threshold. 

```{r courtparent_traits_unlinked,out.extra='',fig.pos="H",fig.cap="Evolution of the courter and parenting trait values in each rep with unlinked genetic variation. Dotted line shows the thresholds determining if males become courters (green) or parents (blue)",fig.width=6,fig.height=8}
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(u,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$ParentAEmean,type="l",col=cols["parent"],
       ylim=c(min(rep.summ$ParentAEmean,rep.summ$CourterAEmean)-0.1,
              max(rep.summ$ParentAEmean,rep.summ$CourterAEmean)+0.1),
       xlim=c(0,12000))
  points(rep.summ$Generation,rep.summ$CourterAEmean,type="l",col=cols["courter"])
  abline(h=rep.summ$ParentThresh[1],lty=2,col=cols["parent"],lwd=2)
  abline(h=rep.summ$CourterThresh[1],lty=2,col=cols["courter"],lwd=2)
})
```

## Linked additive genetic variance


In these cases, the traits are encoded by a number (50) of loci, whose alleles contribute additively to determine the trait value. These alleles anchored on chromosomes, equally distributed among the chromosomes (4). The overall trait value is compared to a population-level threshold (which is static, in these cases), and if the trait value is above the threshold the male takes the parent or courter morph and if it is below he does not. Below, I'm showing the results from 10 replicates of each scenario. 

### Courter trait


```{r linked_courter,out.extra='',fig.pos="H", fig.cap="Frequency of the courter morph with linked QTLs (each color represents a different replicate)"}
cl<-plot.courter.reps(pattern="^courter_linked_novs_[1-5]_summary.txt",cols)
```

When only the courtship trait is included in the simulation, `r length(unlist(lapply(cl,nrow))[unlist(lapply(cl,nrow))>=9999])` of the `r length(cl)` replicates reached an equilibrium by 10000 generations.

```{r linked_courter_final}
courter.final.freqs<-get.courter.freqs(cl)
#aggregate the data
court.ws<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterW"],rep["NonCourterW"]))})
court.freqs<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterFreq"],1-rep["CourterFreq"]))})
court.eq<-apply(courter.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```

This is further evidenced by looking at a table of trait values and fitnesses in the final generation of each replicate

```{r linked_courter_table}
final.freqs<-courter.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of courters with linked QTLs in final generation",
      row.names = TRUE)
```

Every replicate has the courtship trait go to fixation.


```{r,fig.pos="H",fig.cap="Evolution of courtship trait with linked loci. Green dotted lines are the standard deviations of the traits and the black dotted line is the threshold",fig.width=6,fig.height=10, echo=FALSE,eval=FALSE}
#Let's also evaluate the evolution of the courtship trait values:
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(cl,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$CourterAEmean,type="l",col=cols["courter"],
       ylim=c(min(rep.summ$CourterAEmean-rep.summ$CourterAEsd)-0.05,
              max(rep.summ$CourterAEmean+rep.summ$CourterAEsd)+0.05))
  lines(rep.summ$Generation,rep.summ$CourterAEmean+rep.summ$CourterAEsd,type="l",col=alpha(cols["courter"],0.5),lty=3)
  lines(rep.summ$Generation,rep.summ$CourterAEmean-rep.summ$CourterAEsd,type="l",col=alpha(cols["courter"],0.5),lty=3)
  abline(h=rep.summ$CourterThresh[1],lty=2,col="black")
})
#In every case, the trait values evolve above the threshold.
```



### Parental trait


```{r parental_genetics, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of parent morph with linked QTLs (each color represents a different replicate)"}
pl<-plot.parent.reps(pattern="^parent_linked_novs_[1-5]_summary.txt",cols)
```

```{r linked_parent_final}
parent.final.freqs<-get.parent.freqs(pl)
#aggregate the data
parent.ws<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentW"],rep["NonParentW"]))})
parent.freqs<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentFreq"],1-rep["ParentFreq"]))})
parent.eq<-apply(parent.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```

```{r parent_table_gen}
final.freqs<-parent.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of parents with linked QTLs in final generation",
      row.names = TRUE)
```

All of the replicates resulted in the parenting trait going to fixation. Of the `r length(pl)` replicates, `r length(unlist(lapply(pl,nrow))[unlist(lapply(pl,nrow))>=9999])` reached an equilibrium by 10000 generations.


```{r parent_linkedgenetics_traits ,out.extra='',fig.pos="H",fig.cap="Evolution of parenting trait values with linked QTLs. Blue dotted lines are the standard deviations around the mean and the black dotted line is the threshold",fig.width=6,fig.height=10, echo=FALSE,eval=FALSE}
#The evolution of the trait values may shed light on the variable outcomes:

par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(pl,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$ParentAEmean,type="l",col=cols["parent"],
       ylim=c(min(rep.summ$ParentAEmean-rep.summ$ParentAEsd)-0.05,
              max(rep.summ$ParentAEmean+rep.summ$ParentAEsd)+0.05))
  lines(rep.summ$Generation,rep.summ$ParentAEmean+rep.summ$ParentAEsd,type="l",col=alpha(cols["parent"],0.5),lty=3)
  lines(rep.summ$Generation,rep.summ$ParentAEmean-rep.summ$ParentAEsd,type="l",col=alpha(cols["parent"],0.5),lty=3)
  abline(h=rep.summ$CourterThresh[1],lty=2,col="black")
})
```


### Courtship and Parental Traits

Including both courtship and parental traits, along with linked loci, causes the outcomes to be less predictable -- same as in the case of unanchored/unlinked genetic loci.

```{r parentalCourtship_genetics, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the two morphs with linked QTLs (courter = green, parent = blue)"}
l<-plot.pc.reps("^parent-courter_linked_novs_[1-5]_summary.txt",cols,make.plot=FALSE)
```

Because of the variable outcomes, it is useful to visualize the frequencies of the four morphs with each rep in its own graph.


```{r courtParent_morphfreqs_linked,echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the 4 morphs in each rep with linked QTLs",fig.height=8,fig.width=6}
plot.morphs.reps(l,cols2,ncols = 4)
```


```{r linked_courterParent_final}
pc.final.freqs<-get.morph.freqs(l)
rownames(pc.final.freqs)<-NULL
kable(pc.final.freqs[,c(1,grep("Freq",colnames(pc.final.freqs)))],caption = "Frequency of morphs in final generation (linked loci)",
      row.names = TRUE)
```

Multiple morphs are maintained in `r length(apply(pc.final.freqs,1,function(r){ length(r[r==0])})[apply(pc.final.freqs,1,function(r){ length(r[r==0])})<3])` of the `r nrow(pc.final.freqs)` replicates, and those morphs contain either a parent or a courter. What's surprising is that the instance of a different morph going to fixation occurs in only one replicate, even though it had the same initial conditions as the others in its row. 

Looking at the trait value evolution may help us understand why some of the runs have different outcomes:

```{r courtparent_traits_linked,out.extra='',fig.pos="H",fig.cap="Evolution of the courter and parenting trait values in each rep with linked genetic variation",fig.width=6,fig.height=8}
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(l,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$ParentAEmean,type="l",col=cols["parent"],
       ylim=c(min(rep.summ$ParentAEmean,rep.summ$CourterAEmean)-0.1,
              max(rep.summ$ParentAEmean,rep.summ$CourterAEmean)+0.1))
  points(rep.summ$Generation,rep.summ$CourterAEmean,type="l",col=cols["courter"])
  abline(h=rep.summ$ParentThresh[1],lty=2,col=cols["parent"],lwd=2)
  abline(h=rep.summ$CourterThresh[1],lty=2,col=cols["courter"],lwd=2)
})
```


## Supergenes

In these cases, the loci (50) are all located in one region of a chromosome, and that region has reduced recombination. As above, the overall trait value is compared to a population-level threshold (which is static, in these cases), and if the trait value is above the threshold the male takes the parent or courter morph and if it is below he does not.

### Courter trait


```{r supergenes_courter,out.extra='',fig.pos="H", fig.cap="Frequency of the courter morph with supergenes (each color represents a different replicate)"}
cs<-plot.courter.reps(pattern="^courter_supergene_novs_[1-5]_summary.txt",cols)
```

Of the `r length(cs)` replicates, `r length(unlist(lapply(cs,nrow))[unlist(lapply(cs,nrow))>=9999])` reached an equilibrium by 10000 generations.

```{r supergenes_courter_final}
courter.final.freqs<-get.courter.freqs(cs)
#aggregate the data
court.ws<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterW"],rep["NonCourterW"]))})
court.freqs<-apply(courter.final.freqs,1,function(rep){ return(c(rep["CourterFreq"],1-rep["CourterFreq"]))})
court.eq<-apply(courter.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```

```{r supergene_courter_table}
final.freqs<-courter.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of courters with supergenes in final generation",
      row.names = TRUE)
```

```{r,fig.pos="H",fig.cap="Evolution of courtship trait with a supergene. Green dotted lines are the standard deviations around the mean and the black dotted line is the threshold",fig.width=6,fig.height=10,echo=FALSE,eval=FALSE}
#Looking at the evolution of the trait values: 
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(cu,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$CourterAEmean,type="l",col=cols["courter"],
       ylim=c(min(rep.summ$CourterAEmean-rep.summ$CourterAEsd)-0.05,
              max(rep.summ$CourterAEmean+rep.summ$CourterAEsd)+0.05))
  lines(rep.summ$Generation,rep.summ$CourterAEmean+rep.summ$CourterAEsd,type="l",col=alpha(cols["courter"],0.5),lty=3)
  lines(rep.summ$Generation,rep.summ$CourterAEmean-rep.summ$CourterAEsd,type="l",col=alpha(cols["courter"],0.5),lty=3)
  abline(h=rep.summ$CourterThresh[1],lty=2,col="black")
})
```


### Parental trait


```{r parental_supegene, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of parent morph with supergenes (each color represents a different replicate)"}
pss<-plot.parent.reps(pattern="^parent_supergene_novs_[1-5]_summary.txt",cols)
```


```{r supergenes_parent_final}
parent.final.freqs<-get.parent.freqs(pss)
#aggregate the data
parent.ws<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentW"],rep["NonParentW"]))})
parent.freqs<-apply(parent.final.freqs,1,function(rep){ return(c(rep["ParentFreq"],1-rep["ParentFreq"]))})
parent.eq<-apply(parent.final.freqs,1,function(rep){ if(rep["Generation"]==9999) return("E") else return("N")})
```

```{r parent_table_supergenes}
final.freqs<-parent.final.freqs[,-1]
kable(final.freqs,caption = "Frequency of parents with supergenes in final generation",
      row.names = TRUE)
```

Of the `r length(pss)` replicates, `r length(unlist(lapply(pss,nrow))[unlist(lapply(pss,nrow))>=9999])` reached an equilibrium by 10000 generations.



```{r parent_supergene_traits ,out.extra='',fig.pos="H",fig.cap="Evolution of parenting trait values with a supergene. Blue dotted lines are the standard deviations around the mean and the black dotted line is the threshold",fig.width=6,fig.height=10,echo=FALSE,eval=FALSE}
#Let's look at the trajectories of the parent traits.
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(pl,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$ParentAEmean,type="l",col=cols["parent"],
       ylim=c(min(rep.summ$ParentAEmean-rep.summ$ParentAEsd)-0.05,
              max(rep.summ$ParentAEmean+rep.summ$ParentAEsd)+0.05))
  lines(rep.summ$Generation,rep.summ$ParentAEmean+rep.summ$ParentAEsd,type="l",col=alpha(cols["parent"],0.5),lty=3)
  lines(rep.summ$Generation,rep.summ$ParentAEmean-rep.summ$ParentAEsd,type="l",col=alpha(cols["parent"],0.5),lty=3)
  abline(h=rep.summ$CourterThresh[1],lty=2,col="blue")
})
```

### Courtship and Parental Traits


```{r parentalCourtship_supergenes, echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the two morphs with linked QTLs (courter = green, parent = blue)"}
s<-plot.pc.reps("^parent-courter_supergene_novs_[1-5]_summary.txt",cols,make.plot=FALSE)
```

The different runs have different outcomes. To better see this, let's look at the morph frequencies in each rep separately.


```{r courtParent_morphfreqs_supergenes,echo=FALSE,out.extra='',fig.pos="H",fig.cap="Frequency of the 4 morphs in each rep with supergenes",fig.width=6,fig.height=8}
plot.morphs.reps(s,cols2,ncols = 4)
```


```{r supergenes_courterParent_final}
pc.final.freqs<-get.morph.freqs(s)
rownames(pc.final.freqs)<-NULL
kable(pc.final.freqs[,c(1,grep("Freq",colnames(pc.final.freqs)))],caption = "Frequency of morphs in final generation (supergenes)",
      row.names = TRUE)
```

Multiple morphs are maintained in `r length(apply(pc.final.freqs,1,function(r){ length(r[r==0])})[apply(pc.final.freqs,1,function(r){ length(r[r==0])})<3])` of the `r nrow(pc.final.freqs)` replicates, and those morphs contain either a parent or a courter. `r length(unlist(lapply(s,nrow))[lapply(s,nrow) < 9999])` of those reps with variation went extinct before 10000 generations.

```{r courtParent_rs_supergene,echo=FALSE,eval=FALSE}
ymax<-round(max(pcng.sum$ParentW,pcng.sum$NonParentW,pcng.sum$CourterW,pcng.sum$NonCourterW)+1)
par(mar=c(4,4,1,4),mfrow=c(2,1))
plot(pcng.sum$ParentW,lwd=2,xlab="Generations",ylab="Parental Reproductive Success",
     bty="u",type="l",ylim = c(0,ymax),col=alpha(cols["parent"],0.75),las=1)
points(pcng.sum$NonParentW,lwd=2,lty=3,type="l",col=alpha(cols["parent"],0.75))
points(pcng.sum$NonParentW,lwd=2,lty=3,type="l",col=alpha("grey",0.75))
plot(pcng.sum$CourterW,xlab="Generations",ylab="Courter Reproductive Success",
     bty="u",lwd=2,type="l",col=alpha(cols["courter"],0.75),ylim = c(0,ymax))
points(pcng.sum$NonCourterW,lwd=2,lty=3,type="l",col=alpha(cols["courter"],0.75))
points(pcng.sum$NonCourterW,lwd=2,lty=3,type="l",col=alpha("grey",0.75))
legend("topright",bty='n',legend=c("Preferred","Not Preferred"),lty=c(1,2),col="darkgrey",lwd=2)
```

Visualizing the evolution of the trait values for each trait separately again shows that the outcome for the morphs depends on whether the parental trait evolves above or below the threshold.

```{r courtparent_traits_supergene,out.extra='',fig.pos="H",fig.cap="Evolution of the courter and parenting trait values in each rep with a supergene",fig.width=6,fig.height=8}
par(mfrow=c(5,4),mar=c(1,1,1,1),oma=c(2,2,2,2))
out<-lapply(s,function(rep.summ){
  plot(rep.summ$Generation,rep.summ$ParentAEmean,type="l",col=cols["parent"],
       ylim=c(min(rep.summ$ParentAEmean,rep.summ$CourterAEmean)-0.1,
              max(rep.summ$ParentAEmean,rep.summ$CourterAEmean)+0.1))
  points(rep.summ$Generation,rep.summ$CourterAEmean,type="l",col=cols["courter"])
  abline(h=rep.summ$ParentThresh[1],lty=2,col=cols["parent"],lwd=2)
  abline(h=rep.summ$CourterThresh[1],lty=2,col=cols["courter"],lwd=2)
})
```

\newpage

## Preliminary conclusions

* Density-dependent survival successfully prevented populations from crashing.
* Genetic architecture is not playing a major role in the outcomes, at least for these parameter sets. However, recombination among linked genes may result in more variable outcomes (the most variable results are in the case with linked additive genetic variance).
* When both courtship and parenting traits exist, the morphs (combination of the two traits) that 'survive' depend on whether the parentage trait values evolve to be above the threshold (results in Courter/Parent fixation) or to be below the threshold (results in Courter/Non-parent fixation).
* Fixation of the traits/stasis is achieved after a small number of generations (<2000).

