---
title: "Sensitivity analysis with ARTs Analysis"
author: "Sarah P. Flanagan"
date:  "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: true
    dev: png
  html_document: 
    toc: true  
header-includes: \usepackage{float}
always_allow_html: yes
editor_options: 
  chunk_output_type: console
---

<style>
p.caption {
  font-size: 0.85em;
  background-color: #f5f5f5
}
</style>

The purpose of this model is to understand how genetic architectures of alternative reproductive tactics impact their maintenance in populations. I'm using an individual-based simulation model with different selection scenarios, types of alternative tactics, and genetic architectures (genome-wide additive genetic variance, supergenes, expression networks).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200)
knitr::opts_knit$set(root.dir='../fixedART-results/') #change

```
```{r opts, echo = FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "figs/"
)
```
```{r, echo=FALSE,echo=FALSE,message=FALSE}
library(RColorBrewer)
library(scales)
library(sm)
library(kableExtra)
source("~/Research/ARTs/R/002_freq_functions.R")
cols<-c(courter="#33a02c",parent="#1f78b4",cpref="#b2df8a",ppref="#a6cee3")
cols2<-c(CP='#a1dab4',NCP='#41b6c4',CNP='#2c7fb8',NCNP='#253494')
```

Here, I'm presenting the initial sensitivity analysis results. I tested the effects of four parameter settings: mating system, the ratio of reproductive allocation between courters and non-courters, the survival of parental male nests, and the survival of non-parental male nests. 

By 'mating system' I refer to whether males are able to mate with multiple females (i.e., be the nesting male for multiple females). The default is monogamy, meaning males can only be the nesting male to one female. 

The ratio of reproductive allocation between courters and non-courters comes into play during the fertilization stage of the model. At this point, females have mated with a male, and her eggs will be preferentially fertilized by her mate but could also be fertilized by non-courting males. Each male has a maximum reproductive success or reproductive allocation that is given by their courtship status ($r_{courter}$ or $r_{non-courter}$) and gets used up by each additional fertilization. This sets a bound on reproductive success for each male. When a female mates, she receives a total of $\Sigma{n_{sperm}}$ from all of the males who attempt to fertilize her eggs (e.g., with one courter and two non-courters, $\Sigma{n_{sperm}}$ = $r_{courter}$ + 2$*$(0.5$*$$r_{non-courter}$)). The ratio of $r_{courter}$ to $r_{non-courter}$ is the parameter set that I varied in these sensitivity analyses.

The final two traits that I tested were the survival of a nest when it was with parental male and the survival of a nest when it was with a non-parental male. I tested these two parameters separately. When the parental nest survival parameter was tested the non-parental nest survival was constant at 0.1. When the non-parental nest survival parameter was tested the parental nest survival was constant at 0.9.

I tested these parameter settings with three genetic architectures and with just a courter trait, just a parent trait, and both traits: 

```{r}
params<-data.frame(parameter=c("polygyny","relative reproductive allocation (courter:non-courter)", "parental nest survival","non-parental nest survival", "supergene proportion"),
                   default = c("monogamy","8:4 (2)","0.9","0.1","0.1?"),
                   tested_settings = c("polygyny","2:8 (0.25), 4:8 (0.5), 4:4 (1), 8:2 (4)","0.5, 0.6, 0.7, 0.8","0.2, 0.3, 0.4, 0.5","0.5,0.25,0.05"))
kable(params, format="latex", booktabs = T, caption="The parameters, their defaults, and their tested settings") %>% 
  kable_styling(latex_options =c("scale_down","HOLD_position"))
```

Here, I present the initial analyses of those parameter combinations.


# Courter Trait 


```{r, eval=FALSE}
crt_unlinked_summ<-summarize_params(base_pattern="^courter_unlinked",cols,
                                    summ_list="CourterFreq",type="courter")
write.csv(crt_unlinked_summ,"crt_unlinked_summ.csv",row.names = FALSE)
crt_linked_summ<-summarize_params(base_pattern="^courter_linked",cols,
                                  summ_list="CourterFreq",type="courter")
write.csv(crt_linked_summ,"crt_linked_summ.csv",row.names = FALSE)
crt_supergene_summ<-summarize_params(base_pattern="^courter_supergene",cols,
                                     summ_list="CourterFreq",type="courter")
write.csv(crt_supergene_summ,"crt_supergene_summ.csv",row.names = FALSE)
```
```{r}
crt_unlinked_summ<-read.csv("crt_unlinked_summ.csv")
crt_linked_summ<-read.csv("crt_linked_summ.csv")
crt_supergene_summ<-read.csv("crt_supergene_summ.csv")
# Put it all together
crt_summ<-cbind(crt_unlinked_summ,crt_linked_summ[,3:4],crt_supergene_summ[,3:4])
colnames(crt_summ)[3:4]<-paste(colnames(crt_summ)[3:4]," Unlinked",sep="")
colnames(crt_summ)[5:6]<-paste(colnames(crt_summ)[5:6]," Linked",sep="")
colnames(crt_summ)[7:8]<-paste(colnames(crt_summ)[7:8]," Supergene",sep="")
rownames(crt_summ)<-1:nrow(crt_summ)
kable(crt_summ, "latex", booktabs = T, caption = "Summaries of final Courter frequencies",row.names=FALSE) %>% 
  kable_styling(latex_options =c("scale_down","HOLD_position"))
```

When just the courter trait is present, regardless of the parameter settings or genetic architectures, the courting male morph goes to fixation in every run.

### Supergene proportion

This doesn't reflect the segmentation faults that I know happened...
```{r}
cs_prop<-plot.courter.reps(pattern="^courter_supergene_prop.*summary.txt.gz",path="sensitivity",cols,make.plot=TRUE)
```


# Parent Trait 


```{r, eval=FALSE}
prt_unlinked_summ<-summarize_params(base_pattern="^parent_unlinked",cols,
                                    summ_list="ParentFreq",type="parent")
write.csv(prt_unlinked_summ,"prt_unlinked_summ.csv",row.names = FALSE)
prt_linked_summ<-summarize_params(base_pattern="^parent_linked",cols,
                                  summ_list="ParentFreq",type="parent")
write.csv(prt_linked_summ,"prt_linked_summ.csv",row.names = FALSE)
prt_supergene_summ<-summarize_params(base_pattern="^parent_supergene",cols,
                                     summ_list="ParentFreq",type="parent")
write.csv(prt_supergene_summ,"prt_supergene_summ.csv",row.names = FALSE)
```
```{r}
prt_unlinked_summ<-read.csv("prt_unlinked_summ.csv")
prt_linked_summ<-read.csv("prt_linked_summ.csv")
prt_supergene_summ<-read.csv("prt_supergene_summ.csv")
# Put it all together
prt_summ<-cbind(prt_unlinked_summ,prt_linked_summ[,3:4],prt_supergene_summ[,3:4])
colnames(prt_summ)[3:4]<-paste(colnames(prt_summ)[3:4]," Unlinked",sep="")
colnames(prt_summ)[5:6]<-paste(colnames(prt_summ)[5:6]," Linked",sep="")
colnames(prt_summ)[7:8]<-paste(colnames(prt_summ)[7:8]," Supergene",sep="")
rownames(prt_summ)<-NULL
kable(prt_summ, "latex", booktabs = T, row.names = FALSE,caption = "Summaries of final Parent frequencies") %>% 
  kable_styling(latex_options =c("scale_down","HOLD_position"))
```

When just the parent trait is present, regardless of the parameter settings or genetic architectures, the parental male morph goes to fixation in every run.

# Courter and Parent Traits

When both traits are present, the genetic architecture plays a role, so we'll go through one at a time.

```{r,eval=TRUE}
morphs_unlinked_summ<-summarize_params(base_pattern="^parent-courter_unlinked",cols,
                                    summ_list=c("FreqNcNp","FreqNcP","FreqCNp","FreqCP"),type="morphs")
write.csv(morphs_unlinked_summ,"morphs_unlinked_summ.csv",row.names = FALSE)

morphs_linked_summ<-summarize_params(base_pattern="^parent-courter_linked",cols,
                                  summ_list=c("FreqNcNp","FreqNcP","FreqCNp","FreqCP"),type="morphs")
write.csv(morphs_linked_summ,"morphs_linked_summ.csv",row.names = FALSE)

morphs_supergene_summ<-summarize_params(base_pattern="^parent-courter_supergene",cols,
                                     summ_list=c("FreqNcNp","FreqNcP","FreqCNp","FreqCP"),type="morphs")
write.csv(morphs_supergene_summ,"morphs_supergene_summ.csv",row.names = FALSE)

```


## Unlinked QTLs

We'll start by looking at the unlinked QTLs.

```{r}
morphs_unlinked_summ<-read.csv("morphs_unlinked_summ.csv")
kable(morphs_unlinked_summ, "latex", booktabs = T, row.names = FALSE,caption="Summaries of final morph frequencies when QTLs are unanchored")%>% 
  kable_styling(latex_options =c("scale_down","HOLD_position"))
```


The outcomes are not as consistent with both traits present, and it may be easier to visualize the ones that are variable with plots. 

```{r}
cpu_baseline<-up<-plot.pc.reps("^parent-courter_unlinked_.*_summary.txt",path="baseline",cols,make.plot=FALSE)
cpu_baseline_freqs<-get.morph.freqs(cpu_baseline)
cpu_baseline_freqs<-cpu_baseline_freqs[,-1]
cpu_baseline_freqs$Polygyny<-FALSE
cpu_baseline_freqs$CourterRS<-8
cpu_baseline_freqs$NoncourterRS<-4
cpu_baseline_freqs$ParentSurvival<-0.9
cpu_baseline_freqs$NonparentSurvival<-0.1

base_keep_cols<-grep("mean|sd|W|Thresh|Progeny|Pref",colnames(cpu_baseline_freqs),invert = TRUE,value = TRUE)[1:10]
```

```{r echo=FALSE}
params$unlinked<-NA
```

### Mating System

First, we'll look at the morph frequencies when the mating system was either monogamy or polygyny. In the plot below (and in the other similarly formatted plots), each panel represents the frequency of the moprhs in the final generation of the runs. Each point represents a single run, so although the frequencies of each morph for a given run depend on each other (all must sum to 1), the frequencies for a single run are not linked in this diagram (i.e., you can't tell which Courter-Parent frequencies correspond to which Courter-Nonparent frequency, but you can surmise that if the Courter-Parent frequency = 1 that the Courter-Nonparent frequency = 0).

```{r}
cpu_polygyny<-plot.pc.reps(pattern="^parent-courter_unlinked_polygyny_.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_polygyny,cols2,ncols = 4)
cpu_polygyny_freqs<-get.morph.freqs(cpu_polygyny)[,-1]
cpu_polygyny_freqs$Polygyny<-TRUE
#merge with baseline
cpu_polygyny_freqs<-rbind(cpu_polygyny_freqs[,c(base_keep_cols,"Polygyny")],cpu_baseline_freqs[,c(base_keep_cols,"Polygyny")])
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with and without polygyny",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_polygyny_freqs,"Polygyny",cols2,"Mating System",x_lim = c(-0.25,1.25),yname_pos = -0.35,xaxt='n')
axis(1,at=c(0,1),labels=c("Monogamy","Polygyny"),lwd=0)
```

Looking at these figures, the outcomes are consistent across mating systems: in all cases, either the Courter-Parent or the Courter-Nonparent morphs are fixed. 

```{r echo=FALSE}
params[1,"unlinked"]<-"No effect"
```

### Reproductive Allocations


```{r}
cpu_rs<-plot.pc.reps(pattern="^parent-courter_unlinked_crs.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_rs,cols2,ncols = 4)
cpu_rs_freqs<-get.morph.freqs(cpu_rs)[,-1]
cpu_rs_freqs$CourterRS<-gsub(".*crs(\\d)_ncrs\\d.*","\\1",rownames(cpu_rs_freqs))
cpu_rs_freqs$NoncourterRS<-gsub(".*crs\\d_ncrs(\\d).*","\\1",rownames(cpu_rs_freqs))
#merge with baseline
cpu_rs_freqs<-rbind(cpu_rs_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")],cpu_baseline_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")])

cpu_rs_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_rs_freqs$CourterRS)/as.numeric(cpu_rs_freqs$NoncourterRS))
```


For the ratio of reproductive allocations, the parameter settings do affect the outcomes. In the cases when the Courter has a higher allocation -- both 8 to 4 (ratio = 2) and 8 to 2 (ratio = 4) -- the outcomes are variable, with either the Courter-Parent or Courter-Nonparent morphs being fixed. When the non-courters have a much higher reproductive allocation than courters (8 offspring compared to 2, ratio = 0.25), the Courter-Parent morph is never fixed, and instead the Courter-Nonparent morph is fixed. This pattern is also found when Courters and Noncourters have equivalent allocations (4 and 4). Intriguingly, at the ratio of 0.5, when Noncourters have an allocation of 8 and Courters have an allocation of 4, variation is maintained in a handful of the simulations, with a very small number of Noncourter-Nonparent morphs still occurring in the population along with Courter-Parent morphs. 


```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with different ratios of courter to non-courter reproductive success",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_rs_freqs,"Courter2NonRS",cols2,"Ratio of Reproductive Allocation\nCourter:Non-Courter",xaxt='n',x_lim = c(0.15,5.5),yname_pos = -0.2)
axis(1,at=levels(as.factor(as.numeric(cpu_rs_freqs$Courter2NonRS))),labels = levels(cpu_rs_freqs$Courter2NonRS))

```


```{r echo=FALSE}
params[2,"unlinked"]<-"Nuanced effect. High ratios, courters are always fixed. Low allocations, Courter-Nonparent morph is usually fixed, except at ratio 0.25"
```

### Parental nest survival

Parental nest survival does not appear to have a major effect on the outcomes when QTLs are unanchored. Occasionally, at several of the lower survival rates (0.5, 0.7, 0.8) a very small proportion of males have the Noncourter-Nonparent morph when the vast majority of males have the Courter-Parent morph.

```{r}
cpu_psurv<-plot.pc.reps(pattern="^parent-courter_unlinked_psurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_psurv,cols2,ncols = 4)
cpu_psurv_freqs<-get.morph.freqs(cpu_psurv)[,-1]
cpu_psurv_freqs$ParentSurvival<-gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_psurv_freqs))
cpu_psurv_freqs<-rbind(cpu_psurv_freqs[,c(base_keep_cols,"ParentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"ParentSurvival")])
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_psurv_freqs,"ParentSurvival",cols2,"Parent nest survival")
```

```{r echo=FALSE}
params[3,"unlinked"]<-"No major effect"
```

### Non-parental nest survival

No major effects observed, although when non-parent nest survival was 0.3 every run had the outcome of fixed Courter-Parent morph, whereas in other cases the Courter-Nonparent morph was also fixed occasionally. In the case where survival was 0.5, one or two runs maintained tiny amounts of polymorphism, with near-zero positive frequencies of Noncourter-Nonparent morphs occurring alongside the Courter-Parent morphs.

```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_unlinked_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"NonparentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"NonparentSurvival")])
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"NonparentSurvival",cols2,"Non-Parent nest survival")
```


```{r echo=FALSE}
params[4,"unlinked"]<-"No major effect"
```

## Linked QTLs

```{r}
morphs_linked_summ<-read.csv("morphs_linked_summ.csv")
kable(morphs_linked_summ, "latex", booktabs = T, row.names = FALSE,caption="Summaries of final morph frequencies when QTLs are anchored on chromosomes") %>% 
  kable_styling(latex_options =c("scale_down","HOLD_position"))
```


In some of the cases with linked QTLs, polymorphism is maintained in some runs. Therefore, I'm also showing genetic variation and trait variation for each set of parameter settings. Note that the trait variation plots may show higher frequencies of some traits than the summary plots and tables because they include females to demonstrate the total population genetic composition.


```{r}
cpu_baseline<-up<-plot.pc.reps("^parent-courter_linked_.*_summary.txt",path="baseline",cols,make.plot=FALSE)
cpu_baseline_freqs<-get.morph.freqs(cpu_baseline)
cpu_baseline_freqs<-cpu_baseline_freqs[,-1]
cpu_baseline_freqs$Polygyny<-FALSE
cpu_baseline_freqs$CourterRS<-8
cpu_baseline_freqs$NoncourterRS<-4
cpu_baseline_freqs$ParentSurvival<-0.9
cpu_baseline_freqs$NonparentSurvival<-0.1

base_keep_cols<-grep("mean|sd|W|Thresh|Progeny|Pref",colnames(cpu_baseline_freqs),invert = TRUE,value = TRUE)[1:10]
```
```{r echo=FALSE}
params$linked<-NA
```

Here, the distribution of allele frequencies at QTLs for the Courter trait and the Parent trait are shown for the baseline parameter settings. 

```{r,fig.cap="Allele frequencies at QTLs in the final generation of all of the runs - baseline",fig.height=10,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_.*",path="baseline",4,cols)
```


```{r,fig.cap="Trait values in the final generation of all the runs: baseline with linked QTLs",fig.height=10,fig.width=8}
plot_final_traits(pattern="^parent-courter_linked_.*traits.txt",path="baseline",ncols=4,cols=cols,cols2 = cols2)
```

These graphs demonstrate that substantial genetic variation can be maintained at QTLs even when the traits themselves have become fixed. These trait combinations reflect what I've previously observed, that a single morph becomes fixed in the population, with the Courter-Parent morph being fixed when the parental trait evolves to be positive and the Courter-Nonparent morph becoming fixed if the parental trait evolves to have negative values.

### Mating System

```{r}
cpu_polygyny<-plot.pc.reps(pattern="^parent-courter_linked_polygyny_.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_polygyny,cols2,ncols = 4)
cpu_polygyny_freqs<-get.morph.freqs(cpu_polygyny)[,-1]
cpu_polygyny_freqs$Polygyny<-TRUE
#merge with baseline
cpu_polygyny_freqs<-rbind(cpu_polygyny_freqs[,c(base_keep_cols,"Polygyny")],cpu_baseline_freqs[,c(base_keep_cols,"Polygyny")])
```

With linked QTLs, the mating system has an effect. Polygyny allows variation to be maintained, with some runs resulting in a population of *mostly* Courter-Parents, but some small numbers of Courter-Nonparents, Nonparent-Courters, and/or NonCourter-Nonparents.

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with and without polygyny",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_polygyny_freqs,"Polygyny",cols2,"Mating System",x_lim = c(-0.25,1.25),yname_pos = -0.35,xaxt='n')
axis(1,at=c(0,1),labels=c("Monogamy","Polygyny"),lwd=0)
```

From the way these results are presented, it's challenging to determine whether all three morphs are maintained in the same run. To get at that, we can plot each run's morph frequencies over time:

```{r,fig.cap="Frequencies over time in each run with linked QTLs, both traits, and polygyny",fig.height=8,fig.width=8}
plot.morphs.reps(cpu_polygyny,cols2,ncols=4)
```

and the final generation trait values:

```{r,fig.cap="Trait values in the final generation of all the runs: polygyny with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_polygyny.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

Substantial trait variation is maintained when the courter trait and parent trait evolve overlapping trait values that span a substantial range.

If we take a look at the genetic variation, we can see that the genetic variation is not substantially higher even though more trait variation is present (panels are in the same order in the two graphs -- so the middle panels with allele frequencies are the same as the middle panels for trait variation).


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - polygyny",fig.height=6,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_polygyny.*",path="sensitivity",4,cols)
```


```{r echo=FALSE}
params[1,"linked"]<-"Allows polymorphism sometimes"
```


### Reproductive Allocations

In general, high ratios generally result in Courter-Parent morph to be fixed, although at very high ratios some polymorphism with Noncourter-Nonparent is observed. At the lowest ratio, when Noncourters have 8 offspring and courters have 2, in almost all cases the Courter-Nonparent morph is fixed, with one exception where the Courter-Parent morph is fixed. 

```{r}
cpu_rs<-plot.pc.reps(pattern="^parent-courter_linked_crs.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_rs,cols2,ncols = 4)
cpu_rs_freqs<-get.morph.freqs(cpu_rs)[,-1]
cpu_rs_freqs$CourterRS<-gsub(".*crs(\\d)_ncrs\\d.*","\\1",rownames(cpu_rs_freqs))
cpu_rs_freqs$NoncourterRS<-gsub(".*crs\\d_ncrs(\\d).*","\\1",rownames(cpu_rs_freqs))
#merge with baseline
cpu_rs_freqs<-rbind(cpu_rs_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")],cpu_baseline_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")])

cpu_rs_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_rs_freqs$CourterRS)/as.numeric(cpu_rs_freqs$NoncourterRS))
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with different ratios of courter to non-courter reproductive success",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_rs_freqs,"Courter2NonRS",cols2,"Ratio of Reproductive Allocation\nCourter:Non-Courter",xaxt='n',x_lim = c(0.15,5.5),yname_pos = -0.2)
axis(1,at=levels(as.factor(as.numeric(cpu_rs_freqs$Courter2NonRS))),labels = levels(cpu_rs_freqs$Courter2NonRS))

```

QTL allele frequencies are not incredibly informative for these runs, so I will not present those. however, I will present the trait values for each parameter set.


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 2, noncourter RS = 8",fig.height=8,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs2_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 2, noncourter RS = 8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs2_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 4, noncourter RS = 8",fig.height=8,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs4_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 4, noncourter RS = 8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs4_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 8, noncourter RS = 2",fig.height=8,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 2 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs8_ncrs2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

This is interesting because in these cases for multiple morphs to be maintained the parental trait has to be non-positive but not necessarily overlapping the courter trait values. 


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 8, noncourter RS = 8",fig.height=8,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs8_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

And here, no variation is maintained and the Courter-Parent morph is always maintained.


```{r echo=FALSE}
params[2,"linked"]<-"Nuanced effect. High ratios generally result in Courter-Parent morph to be fixed, although at very high ratios some polymorphism with Noncourter-Nonparent is observed. At the lowest ratio, usually Courter-Nonparent is fixed instead of Courter-Parent"
```

### Parental nest survival


```{r}
cpu_psurv<-plot.pc.reps(pattern="^parent-courter_linked_psurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_psurv,cols2,ncols = 4)
cpu_psurv_freqs<-get.morph.freqs(cpu_psurv)[,-1]
cpu_psurv_freqs$ParentSurvival<-gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_psurv_freqs))
cpu_psurv_freqs<-rbind(cpu_psurv_freqs[,c(base_keep_cols,"ParentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"ParentSurvival")])
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_psurv_freqs,"ParentSurvival",cols2,"Parent nest survival")
```

In general, the Courter-Parent or Courter-Nonparent traits become fixed, although with parental nest survival at 0.7 only the Courter-Parent morph became fixed (though this may be an artefact of not running huge numbers of runs). Because these outcomes are not qualitatively different from the baseline outcomes, I'll just show the final trait values for the runs with parental nest survival at 0.7.


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.5",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.5 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_psurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.6",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.6.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.6 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_psurv0.6.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.7",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.7.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.7 with linked QTLs",fig.height=5,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_psurv0.7.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

It appears that when trait values overlap substantially, polymorphism can be maintained.

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.8",fig.height=10,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.8 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_psurv0.8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r echo=FALSE}
params[3,"linked"]<-"No major effect"
```

### Non-parental nest survival


In general, the Courter-Parent or Courter-Nonparent morph becomes fixed, although occasionally the Courter-Nonparent morph is fixed instead. The highest survival rate (0.5) allows the Courter-Nonparent morph to be fixed more frequently. For now, we won't look at each run individually.

```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_linked_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"NonparentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"NonparentSurvival")])
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"NonparentSurvival",cols2,"Non-Parent nest survival")
```


```{r echo=FALSE}
params[4,"linked"]<-"High survival allows higher frequency of Courter-Nonparent morph"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.2",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.2.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.2 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.3",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.3.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.3 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.3.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.4",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.4.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.4 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.4.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.5",fig.height=8,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.5 with linked QTLs",fig.height=4,fig.width=7,eval=FALSE}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```



## QTLs in a supergene

When the QTLs are anchored in a supergene, more variation in allele frequencies occurs at the baseline, but the overall trait variation is the same as baseline parameters for the other genetic architectures.

```{r}
morphs_supergene_summ<-read.csv("morphs_supergene_summ.csv")
kable(morphs_supergene_summ, "latex", booktabs = T, row.names = FALSE,caption="Summaries of final morph frequencies when QTLs are in a supergene")%>%
  kable_styling(latex_options =c("scale_down","HOLD_position"))
```


```{r}
cpu_baseline<-plot.pc.reps("^parent-courter_supergene_.*_summary.txt",path="baseline",cols,make.plot=FALSE)
cpu_baseline_freqs<-get.morph.freqs(cpu_baseline)
cpu_baseline_freqs<-cpu_baseline_freqs[,-1]
cpu_baseline_freqs$Polygyny<-FALSE
cpu_baseline_freqs$CourterRS<-8
cpu_baseline_freqs$NoncourterRS<-4
cpu_baseline_freqs$ParentSurvival<-0.9
cpu_baseline_freqs$NonparentSurvival<-0.1
cpu_baseline_freqs$SupergeneProportion<-0.1

base_keep_cols<-grep("mean|sd|W|Progeny|Pref|Polygyny",colnames(cpu_baseline_freqs),invert = TRUE,value = TRUE)
```

```{r echo=FALSE}
params$supergene<-NA
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - baseline",fig.height=10,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_.*",path="baseline",4,cols)
```

```{r,fig.cap="Trait values in the final generation of all the runs: polygyny with a supergene",fig.height=10,fig.width=8,eval=FALSE}
plot_final_traits(pattern="^parent-courter_supergene_.*traits.txt",path="baseline",ncols=4,cols=cols,cols2 = cols2)
```


### Mating System

Similar to the case with linked QTLs, polygyny with QTLs in a supergene allows polymorphism to be maintained in some runs, this time with Courter-Parent, Courter-Nonparent, and Noncourter-Parent morphs all being maintained in the population.

```{r}
cpu_polygyny<-plot.pc.reps(pattern="^parent-courter_supergene_polygyny_.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_polygyny,cols2,ncols = 4)
cpu_polygyny_freqs<-get.morph.freqs(cpu_polygyny)[,-1]
cpu_polygyny_freqs$Polygyny<-TRUE
#merge with baseline
cpu_polygyny_freqs<-rbind(cpu_polygyny_freqs[,c(base_keep_cols,"Polygyny")],cpu_baseline_freqs[,c(base_keep_cols,"Polygyny")])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with and without polygyny",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_polygyny_freqs,"Polygyny",cols2,"Mating System",x_lim = c(-0.25,1.25),yname_pos = -0.35,xaxt='n')
axis(1,at=c(0,1),labels=c("Monogamy","Polygyny"),lwd=0)
```


If we look at the trait values, we see that the trait values fall into three categories, essentially, rather than forming a continuous distribution of trait values. 

```{r,fig.cap="Trait values in the final generation of all the runs: polygyny with a supergene",fig.height=6,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_polygyny.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - polygyny",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_polygyny.*",path="sensitivity",4,cols)
```

When variation is maintained, the genetic variation maintained is much greater as well.


```{r echo=FALSE}
params[1,"supergene"]<-"Allows polymorphism sometimes"
```


### Reproductive Allocations

Now polymorphism can be maintained when the ratios are very low. In one set of simulations when Noncourters had the reproductive advantage, the Courter-Parent morphs coexisted with Courter-Nonparents and Noncourter-Parents. Otherwise, similar patterns were observed with either Courter-Parent morphs of Courter-Nonparent morphs becoming fixed. 

```{r}
cpu_rs<-plot.pc.reps(pattern="^parent-courter_supergene_crs.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_rs,cols2,ncols = 4)
cpu_rs_freqs<-get.morph.freqs(cpu_rs)[,-1]
cpu_rs_freqs$CourterRS<-gsub(".*crs(\\d)_ncrs\\d.*","\\1",rownames(cpu_rs_freqs))
cpu_rs_freqs$NoncourterRS<-gsub(".*crs\\d_ncrs(\\d).*","\\1",rownames(cpu_rs_freqs))
#merge with baseline
cpu_rs_freqs<-rbind(cpu_rs_freqs[,base_keep_cols[1:14]],cpu_baseline_freqs[,base_keep_cols[1:14]])

cpu_rs_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_rs_freqs$CourterRS)/as.numeric(cpu_rs_freqs$NoncourterRS))
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different ratios of courter to non-courter reproductive success",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_rs_freqs,"Courter2NonRS",cols2,"Ratio of Reproductive Allocation\nCourter:Non-Courter",xaxt='n',x_lim = c(0.15,5.5),yname_pos = -0.2)
axis(1,at=levels(as.factor(as.numeric(cpu_rs_freqs$Courter2NonRS))),labels = levels(cpu_rs_freqs$Courter2NonRS))

```


Genetic composition?

```{r,eval=FALSE}
mrks<-read.delim("sensitivity/parent-courter_supergene_crs2_ncrs8_2_markers.txt")
qtls<-read.delim("sensitivity/parent-courter_supergene_crs2_ncrs8_2_qtlinfo.txt")
qtls<-qtls[,complete.cases(t(qtls))] #remove nas
qtl_af<-cbind(mrks[,1:2],mrks[,colnames(mrks) %in% paste("Marker",qtls[1,-1],sep="")])
qtl_af<-qtl_af[qtl_af$Pop=="Pop0",]

courter_qtls<-paste("Marker",qtls[1,grep("Courter",colnames(qtls))],sep="")
parent_qtls<-paste("Marker",qtls[1,grep("Parent",colnames(qtls))],sep="")

qtl_dat<-do.call(rbind,apply(qtl_af,1,function(qtlr,parent_inx){
  freqs<-qtlr[3:length(qtlr)]
  gen<-rep(qtlr[1],length(freqs))
  trait<-rep("Courter",length(freqs))
  trait[parent_inx]<-"Parent"
  return(data.frame(freqs=freqs,generation=gen,trait=trait))
},parent_inx=which(colnames(qtl_af) %in% parent_qtls)-2))
qtl_dat$generation<-as.numeric(as.character(qtl_dat$generation))
qtl_dat$freqs<-as.numeric(as.character(qtl_dat$freqs))

```
```{r plotDensities, eval=FALSE,echo=FALSE}

par(mar=c(5,4,4,4),oma=c(2,2,2,2),xpd=TRUE)
plot(c(1,length(intervals)),c(0,1),type='n',xlab="Generations",ylab="Allele Frequencies",xaxt='n',bty="L")
axis(1,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  dC<-density(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"],from=0,to=1)
  #dP<-density(qtl_dat[qtl_dat$trait=="Parent" & qtl_dat$generation ==int,"freqs"])
  polygon(dC$y+cnts,dC$x,col = alpha(cols["courter"],0.5)) 
  #rug(int,jitter(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"]))
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))


plot(c(0,1),c(1,length(intervals)),type='n',ylab="Generations",xlab="Allele Frequencies",yaxt='n',bty="L")
axis(2,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  dC<-density(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"],from=0,to=1)
  #dP<-density(qtl_dat[qtl_dat$trait=="Parent" & qtl_dat$generation ==int,"freqs"])
  polygon(dC$x,dC$y+cnts,col = alpha(cols["courter"],0.5))
  rug(jitter(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"]))
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))
```
```{r,eval=FALSE}
library(scales)
intervals<-c(seq(0,11999,500),11999)

par(mfrow=c(2,1),mar=c(2,2,1,1),oma=c(2,2,2,2),xpd=TRUE)
plot(c(1,length(intervals)),c(0,length(seq(0,1,0.05))),type='n',xlab="",ylab="",yaxt='n',xaxt='n',bty="L")
axis(2,at=0:(length(seq(0,1,0.05))-1),labels=seq(0,1,0.05))
axis(1,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  h<-hist(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"],breaks=seq(0,1,0.05),plot = FALSE)
  frq<-h$counts/sum(h$counts)
  barplot(frq,axes=FALSE,space=0,col = alpha(cols["courter"],0.5),offset=cnts,add = TRUE,horiz = TRUE,border=cols["courter"])
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))

plot(c(1,length(intervals)),c(0,length(seq(0,1,0.05))),type='n',xlab="Generations",ylab="Parent Allele Frequencies",yaxt='n',xaxt='n',bty="L")
axis(2,at=0:(length(seq(0,1,0.05))-1),labels=seq(0,1,0.05))
axis(1,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  h<-hist(qtl_dat[qtl_dat$trait=="Parent" & qtl_dat$generation ==int,"freqs"],breaks=seq(0,1,0.05),plot = FALSE)
  frq<-h$counts/sum(h$counts)
  barplot(frq,axes=FALSE,space=0,col = alpha(cols["parent"],0.5),offset=cnts,add = TRUE,horiz = TRUE,border=cols["parent"])
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))

mtext("Generations",1,outer=TRUE)
mtext("Allele Frequencies",2,outer=TRUE)

```

I think that looking into the trait values is more informative than investigating allele frequencies at this stage, so I'll focus on the trait value plots for now. Similar to when polymorphism is maintained with polygyny, the trait values become fixed at three different values to allow the various morphs to be maintained.

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 2, noncourter RS = 8",fig.height=4,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs2_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 2, noncourter RS = 8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs2_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 4, noncourter RS = 8",fig.height=6,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs4_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 4, noncourter RS = 8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs4_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 8, noncourter RS = 2",fig.height=4,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 2 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs8_ncrs2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 8, noncourter RS = 8",fig.height=4,fig.width=8.5,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs8_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```



```{r echo=FALSE}
params[2,"supergene"]<-"Low ratios allow polymorphism in some cases"
```


### Parental nest survival


```{r}
cpu_psurv<-plot.pc.reps(pattern="^parent-courter_supergene_psurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_psurv,cols2,ncols = 4)
cpu_psurv_freqs<-get.morph.freqs(cpu_psurv)[,-1]
cpu_psurv_freqs$ParentSurvival<-gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_psurv_freqs))
cpu_psurv_freqs<-rbind(cpu_psurv_freqs[,base_keep_cols[c(seq(1,12),15)]],cpu_baseline_freqs[,base_keep_cols[c(1:12,15)]])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_psurv_freqs,"ParentSurvival",cols2,"Parent nest survival")
```

Here, the lowest parental nest survival rate (0.5) resulted in all cases having the Courter-Nonparent morph becoming fixed, whereas in the other runs either the Courter-Parent or the Courter-Nonparent could become fixed. 

```{r echo=FALSE}
params[3,"supergene"]<-"Low survival results in coevolution of Courter and Non-parent traits"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.5",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.5 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.6",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.6.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.6 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.6.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.7",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.7.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.7 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.7.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.8",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

### Non-parental nest survival

High survival of the non-parental nests allows polymorphism to be maintained -- Courter-Parent, Courter-Nonparent, and Noncourter-Parent morphs are all maintained at appreciable frequencies when survival was 0.5. In the other cases, either the Courter-Parent or the Courter-Nonparent morphs became fixed.

```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_supergene_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,base_keep_cols[c(1:12,16)]],cpu_baseline_freqs[,base_keep_cols[c(1:12,16)]])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"NonparentSurvival",cols2,"Non-Parent nest survival")
```

For now, I'll just plot the non-parent nest survival = 0.5 set of outcomes. We see that there are again three fixed trait values with a supergene.



```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.2",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.2.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.2 with a supergene",fig.height=4,fig.width=8,eval=FALSE}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.3",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.3.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.3 with a supergene",fig.height=4,fig.width=8,eval=FALSE}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.3.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.4",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.4.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.4 with a supergene",fig.height=4,fig.width=8,eval=FALSE}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.4.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.5",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.5 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_cd supergene_npsurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r echo=FALSE}
params[4,"supergene"]<-"High survival allows polymorphism"
```

### Supergene proportions

```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_supergene_prop.*summary.txt.gz",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$SupergeneProportion<-as.numeric(gsub(".*prop(\\d.\\d+)_.*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"SupergeneProportion")],cpu_baseline_freqs[,c(base_keep_cols,"SupergeneProportion")])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"SupergeneProportion",cols2,"Supergene Proportion")
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene proportion = 0.05",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_prop0.05.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: supergene proportion = 0.05",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_prop0.05.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene proportion = 0.25",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_prop0.25.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: supergene proportion = 0.25",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_prop0.25.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene proportion = 0.5",fig.height=5,fig.width=8,eval=FALSE}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_prop0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: supergene proportion = 0.5",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_prop0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


# Summary

The genetic architecture of traits impacts whether variation can be maintained, and the extent of genetic variation that is maintained. When trait variation is maintained with linked QTLs, the variation in trait values (which are equivalent to the sum of the allelic effects of QTLs) is continuous. However, with a supergene, trait variation is discrete and is generally related to increased variation in allele frequencies at the QTLs. The effects of the relative reproductive allocation between courters and non-courters is seen in all of the genetic architectures. When courters are the higher allocation, the Courter-Parent morph is usually fixed but for low ratios other morphs can be maintained -- including polymorphism when the QTLs are anchored on chromosomes. This makes sense to me.

The other parameters only have effects when QTLs are anchored to chromosomes. Parental nest survival, non-parental nest survival, and polygyny generally allowed polymorphism to be maintained when there was some genetic architecture and for some parameter combinations. In some cases this polymorphism is partly maintained through females, who do not express the traits themselves but carry the genes for the traits. 


```{r}
kable(params, "latex", booktabs = T) %>% 
  kable_styling(latex_options =c("scale_down", "HOLD_position")) %>%
  column_spec(1, width= "15em") %>%
  column_spec(3, width= "15em") %>%
  column_spec(4, width= "20em") %>%
  column_spec(5, width= "20em") %>%
  column_spec(6, width= "20em")
```

Summary plot
```{r}

cpu_all_freqs<-cpu_rs_freqs
cpu_all_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_all_freqs$CourterRS)/as.numeric(cpu_all_freqs$NoncourterRS))
cpu_all_freqs$Parent2NonSurvival<-0.9/0.1
cpu_psurv_freqs$Courter2NonRS<-8/2
cpu_psurv_freqs$Parent2NonSurvival<-as.numeric(cpu_psurv_freqs$ParentSurvival)/0.1
cpu_all_freqs<-rbind(cpu_psurv_freqs[,c(base_keep_cols,"Courter2NonRS","Parent2NonSurvival")],cpu_all_freqs[,c(base_keep_cols,"Courter2NonRS","Parent2NonSurvival")])

cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_supergene_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs$Courter2NonRS<-8/2
cpu_npsurv_freqs$Parent2NonSurvival<-0.9/as.numeric(cpu_npsurv_freqs$NonparentSurvival)
cpu_all_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"Courter2NonRS","Parent2NonSurvival")],cpu_all_freqs[,c(base_keep_cols,"Courter2NonRS","Parent2NonSurvival")])

#Plotting setup
nx<-length(unique(cpu_all_freqs$Courter2NonRS)) #number of values on x axis
ny<-length(unique(cpu_all_freqs$Parent2NonSurvival)) #number of values on y axis
cpu_all_freqs$Rep<-as.numeric(factor(rownames(cpu_all_freqs[order(cpu_all_freqs$Courter2NonRS),]))) # make rep in order of RS
cpu_all_freqs$RSloc<-as.numeric(as.factor(cpu_all_freqs$Courter2NonRS))
cpu_all_freqs$Survloc<-as.numeric(as.factor(cpu_all_freqs$Parent2NonSurvival))
rep_ns<-tapply(cpu_all_freqs$Rep,cpu_all_freqs$RSloc,function(x) length(x))

cpu_all_freqs$paramsets<-gsub(".*/parent-courter_supergene_(.*)_\\d+_summary.txt_\\d+","\\1",rownames(cpu_all_freqs))

newlocs<-do.call(rbind,lapply(unique(cpu_all_freqs$paramsets),function(p,cpu_all_freqs){  
  newlocs<-cpu_all_freqs[cpu_all_freqs$paramsets==p,]$RSloc+seq(-0.4,0.4,length.out = nrow(cpu_all_freqs[cpu_all_freqs$paramsets==p,]))
  return(cbind(Rep=cpu_all_freqs[cpu_all_freqs$paramsets==p,"Rep"],newlocs))
},cpu_all_freqs=cpu_all_freqs))

cpu_all_freqs<-merge(cpu_all_freqs,newlocs,by="Rep")

png("~/Research/ARTs/figs/RS_Survival.png",height=7,width=7,units="in",res=300)

plot(c(1,1),c(nx,ny),xlim=c(0,nx+1),ylim=c(0,ny+1),axes = FALSE,
     xlab="Relative reproductive allocation (courter/non-courter)",
     ylab="Relative nest survival (parent/non-parent)",type='n')
abline(h=seq(0.5,(ny+1.5)),col="grey",xpd=FALSE)
axis(2,at=1:ny,labels = as.numeric(levels(as.factor(cpu_all_freqs$Parent2NonSurvival))),lty=0,las=1)
abline(v=seq(0.5,(nx+1.5)),col="grey",xpd=FALSE)
axis(1,at=1:nx,labels=as.numeric(levels(as.factor(as.numeric(cpu_all_freqs$Courter2NonRS)))),lty=0)

points(cpu_all_freqs$newlocs,cpu_all_freqs$Survloc+0.2,
       col=alpha(cols2["CP"],cpu_all_freqs$FreqCP),pch=18)
points(cpu_all_freqs$newlocs,cpu_all_freqs$Survloc+0.075,
       col=alpha(cols2["CNP"],cpu_all_freqs$FreqCNp),pch=17)
points(cpu_all_freqs$newlocs,cpu_all_freqs$Survloc-0.075,
       col=alpha(cols2["NCP"],cpu_all_freqs$FreqNcP),pch=16)
points(cpu_all_freqs$newlocs,cpu_all_freqs$Survloc-0.2,
       col=alpha(cols2["NCNP"],cpu_all_freqs$FreqNcNp),pch=15)
par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0, 0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=c(18,17,16,15),xpd = TRUE,ncol=2)
dev.off()
```


Further follow-up ideas:

* Do simple FST comparisons between the morphs -- can the QTLs be identified?
* How does the supergene evolve over time?
* Do some loci evolve to have a larger effect on the traits than others?
* Run other sensitivity paramters, plus the key parameters of interest

(I will wait on these other analysis ideas for analysing my key parameters of interest)
