---
title: "Sensitivity analysis with ARTs Analysis"
author: "Sarah P. Flanagan"
date:  "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document: 
    toc: true  
  pdf_document:
    toc: true
    dev: png
header-includes: \usepackage{float}
always_allow_html: yes
editor_options: 
  chunk_output_type: console
---

<style>
p.caption {
  font-size: 0.85em;
  background-color: #f5f5f5
}
</style>

The purpose of this model is to understand how genetic architectures of alternative reproductive tactics impact their maintenance in populations. I'm using an individual-based simulation model with different selection scenarios, types of alternative tactics, and genetic architectures (genome-wide additive genetic variance, supergenes, expression networks).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200)
knitr::opts_knit$set(root.dir='../fixedART-results/') #change

```
```{r opts, echo = FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "figs/"
)
```
```{r, echo=FALSE,echo=FALSE}
library(RColorBrewer)
library(scales)
library(sm)
source("~/Research/ARTs/R/002_freq_functions.R")
cols<-c(courter="#33a02c",parent="#1f78b4",cpref="#b2df8a",ppref="#a6cee3")
cols2<-c(CP='#a1dab4',NCP='#41b6c4',CNP='#2c7fb8',NCNP='#253494')
```

Here, I'm presenting the initial sensitivity analysis results. I tested the effects of four parameter settings: mating system, the ratio of reproductive allocation between courters and non-courters, the survival of parental male nests, and the survival of non-parental male nests. 

By 'mating system' I refer to whether males are able to mate with multiple females (i.e., be the nesting male for multiple females). The default is monogamy, meaning males can only be the nesting male to one female. 

The ratio of reproductive allocation between courters and non-courters comes into play during the fertilization stage of the model. At this point, females have mated with a male, and her eggs will be preferentially fertilized by her mate but could also be fertilized by non-courting males. Each male has a maximum reproductive success or reproductive allocation that is given by their courtship status ($r_{courter}$ or $r_{non-courter}$) and gets used up by each additional fertilization. This sets a bound on reproductive success for each male. When a female mates, she receives a total of $\Sigma{n_{sperm}}$ from all of the males who attempt to fertilize her eggs (e.g., with one courter and two non-courters, $\Sigma{n_{sperm}}$ = $r_{courter}$ + 2$*$(0.5$*$$r_{non-courter}$)). The ratio of $r_{courter}$ to $r_{non-courter}$ is the parameter set that I varied in these sensitivity analyses.

The final two traits that I tested were the survival of a nest when it was with parental male and the survival of a nest when it was with a non-parental male. I tested these two parameters separately. When the parental nest survival parameter was tested the non-parental nest survival was constant at 0.1. When the non-parental nest survival parameter was tested the parental nest survival was constant at 0.9.

I tested these parameter settings with three genetic architectures and with just a courter trait, just a parent trait, and both traits: 

```{r}
params<-data.frame(parameter=c("polygyny","relative reproductive allocation (courter:non-courter)", "parental nest survival","non-parental nest survival"),
                   default = c("monogamy","8:4 (2)","0.9","0.1"),
                   tested_settings = c("polygyny","2:8 (0.25), 4:8 (0.5), 4:4 (1), 8:2 (4)","0.5, 0.6, 0.7, 0.8","0.2, 0.3, 0.4, 0.5"))
kable(params,caption="The parameters, their defaults, and their tested settings",format="pandoc")
```

Here, I present the initial analyses of those parameter combinations.


# Courter Trait 


```{r, eval=FALSE}
crt_unlinked_summ<-summarize_params(base_pattern="^courter_unlinked",cols,
                                    summ_list="CourterFreq",type="courter")
write.csv(crt_unlinked_summ,"crt_unlinked_summ.csv",row.names = FALSE)
crt_linked_summ<-summarize_params(base_pattern="^courter_linked",cols,
                                  summ_list="CourterFreq",type="courter")
write.csv(crt_linked_summ,"crt_linked_summ.csv",row.names = FALSE)
crt_supergene_summ<-summarize_params(base_pattern="^courter_supergene",cols,
                                     summ_list="CourterFreq",type="courter")
write.csv(crt_supergene_summ,"crt_supergene_summ.csv",row.names = FALSE)
```
```{r}
crt_unlinked_summ<-read.csv("crt_unlinked_summ.csv")
crt_linked_summ<-read.csv("crt_linked_summ.csv")
crt_supergene_summ<-read.csv("crt_supergene_summ.csv")
# Put it all together
crt_summ<-cbind(crt_unlinked_summ,crt_linked_summ[,3:4],crt_supergene_summ[,3:4])
colnames(crt_summ)[3:4]<-paste(colnames(crt_summ)[3:4]," Unlinked",sep="")
colnames(crt_summ)[5:6]<-paste(colnames(crt_summ)[5:6]," Linked",sep="")
colnames(crt_summ)[7:8]<-paste(colnames(crt_summ)[7:8]," Supergene",sep="")
rownames(crt_summ)<-1:nrow(crt_summ)
kable(crt_summ,caption = "Summaries of final morph frequencies",row.names=FALSE)
```

When just the courter trait is present, regardless of the parameter settings or genetic architectures, the courting male morph goes to fixation in every run.

# Parent Trait 


```{r, eval=FALSE}
prt_unlinked_summ<-summarize_params(base_pattern="^parent_unlinked",cols,
                                    summ_list="ParentFreq",type="parent")
write.csv(prt_unlinked_summ,"prt_unlinked_summ.csv",row.names = FALSE)
prt_linked_summ<-summarize_params(base_pattern="^parent_linked",cols,
                                  summ_list="ParentFreq",type="parent")
write.csv(prt_linked_summ,"prt_linked_summ.csv",row.names = FALSE)
prt_supergene_summ<-summarize_params(base_pattern="^parent_supergene",cols,
                                     summ_list="ParentFreq",type="parent")
write.csv(prt_supergene_summ,"prt_supergene_summ.csv",row.names = FALSE)
```
```{r}
prt_unlinked_summ<-read.csv("prt_unlinked_summ.csv")
prt_linked_summ<-read.csv("prt_linked_summ.csv")
prt_supergene_summ<-read.csv("prt_supergene_summ.csv")
# Put it all together
prt_summ<-cbind(prt_unlinked_summ,prt_linked_summ[,3:4],prt_supergene_summ[,3:4])
colnames(prt_summ)[3:4]<-paste(colnames(prt_summ)[3:4]," Unlinked",sep="")
colnames(prt_summ)[5:6]<-paste(colnames(prt_summ)[5:6]," Linked",sep="")
colnames(prt_summ)[7:8]<-paste(colnames(prt_summ)[7:8]," Supergene",sep="")
rownames(prt_summ)<-NULL
kable(prt_summ,row.names = FALSE,caption = "Summaries of final morph frequencies")
```

When just the parent trait is present, regardless of the parameter settings or genetic architectures, the parental male morph goes to fixation in every run.

# Courter and Parent Traits

When both traits are present, the genetic architecture plays a role, so below the three genetic architectures are presented in their own tables.

```{r,eval=FALSE}
morphs_unlinked_summ<-summarize_params(base_pattern="^parent-courter_unlinked",cols,
                                    summ_list=c("FreqNcNp","FreqNcP","FreqCNp","FreqCP"),type="morphs")
write.csv(morphs_unlinked_summ,"morphs_unlinked_summ.csv",row.names = FALSE)

morphs_linked_summ<-summarize_params(base_pattern="^parent-courter_linked",cols,
                                  summ_list=c("FreqNcNp","FreqNcP","FreqCNp","FreqCP"),type="morphs")
write.csv(morphs_linked_summ,"morphs_linked_summ.csv",row.names = FALSE)

morphs_supergene_summ<-summarize_params(base_pattern="^parent-courter_supergene",cols,
                                     summ_list=c("FreqNcNp","FreqNcP","FreqCNp","FreqCP"),type="morphs")
write.csv(morphs_supergene_summ,"morphs_supergene_summ.csv",row.names = FALSE)

```

```{r}
morphs_unlinked_summ<-read.csv("morphs_unlinked_summ.csv")
kable(morphs_unlinked_summ,row.names = FALSE,caption="Summaries of final morph frequencies when QTLs are unanchored")
```

```{r}
morphs_linked_summ<-read.csv("morphs_linked_summ.csv")
kable(morphs_linked_summ,row.names = FALSE,caption="Summaries of final morph frequencies when QTLs are anchored on chromosomes")
```

```{r}
morphs_supergene_summ<-read.csv("morphs_supergene_summ.csv")
kable(morphs_supergene_summ,row.names = FALSE,caption="Summaries of final morph frequencies when QTLs are in a supergene")
```


The outcomes are not as consistent with both traits present, and it may be easier to visualize with plots.

## Unlinked QTLs

We'll start by looking at the unlinked QTLs.

```{r}
cpu_baseline<-up<-plot.pc.reps("^parent-courter_unlinked_.*_summary.txt",path="baseline",cols,make.plot=FALSE)
cpu_baseline_freqs<-get.morph.freqs(cpu_baseline)
cpu_baseline_freqs<-cpu_baseline_freqs[,-1]
cpu_baseline_freqs$Polygyny<-FALSE
cpu_baseline_freqs$CourterRS<-8
cpu_baseline_freqs$NoncourterRS<-4
cpu_baseline_freqs$ParentSurvival<-0.9
cpu_baseline_freqs$NonparentSurvival<-0.1

base_keep_cols<-grep("mean|sd|W|Thresh|Progeny|Pref",colnames(cpu_baseline_freqs),invert = TRUE,value = TRUE)[1:10]
```

```{r echo=FALSE}
params$unlinked<-NA
```

### Polygyny

```{r}
cpu_polygyny<-plot.pc.reps(pattern="^parent-courter_unlinked_polygyny_.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_polygyny,cols2,ncols = 4)
cpu_polygyny_freqs<-get.morph.freqs(cpu_polygyny)[,-1]
cpu_polygyny_freqs$Polygyny<-TRUE
#merge with baseline
cpu_polygyny_freqs<-rbind(cpu_polygyny_freqs[,c(base_keep_cols,"Polygyny")],cpu_baseline_freqs[,c(base_keep_cols,"Polygyny")])
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with and without polygyny",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_polygyny_freqs,"Polygyny",cols2,"Mating System",x_lim = c(-0.25,1.25),yname_pos = -0.35,xaxt='n')
axis(1,at=c(0,1),labels=c("Monogamy","Polygyny"),lwd=0)
```

Looking at these figures, the outcomes are consistent across mating systems: in all cases, either the Courter-Parent or the Courter-Nonparent morphs are fixed. 

```{r echo=FALSE}
params[1,"unlinked"]<-"No effect"
```

### Reproductive Allocations


```{r}
cpu_rs<-plot.pc.reps(pattern="^parent-courter_unlinked_crs.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_rs,cols2,ncols = 4)
cpu_rs_freqs<-get.morph.freqs(cpu_rs)[,-1]
cpu_rs_freqs$CourterRS<-gsub(".*crs(\\d)_ncrs\\d.*","\\1",rownames(cpu_rs_freqs))
cpu_rs_freqs$NoncourterRS<-gsub(".*crs\\d_ncrs(\\d).*","\\1",rownames(cpu_rs_freqs))
#merge with baseline
cpu_rs_freqs<-rbind(cpu_rs_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")],cpu_baseline_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")])

cpu_rs_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_rs_freqs$CourterRS)/as.numeric(cpu_rs_freqs$NoncourterRS))
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with different ratios of courter to non-courter reproductive success",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_rs_freqs,"Courter2NonRS",cols2,"Ratio of Reproductive Allocation\nCourter:Non-Courter",xaxt='n',x_lim = c(0.15,5.5),yname_pos = -0.2)
axis(1,at=levels(as.factor(as.numeric(cpu_rs_freqs$Courter2NonRS))),labels = levels(cpu_rs_freqs$Courter2NonRS))

```

For the ratio of reproductive allocations, the parameter settings do affect the outcomes. In the cases when the Courter has a higher allocation -- both 8 to 4 (ratio = 2) and 8 to 2 (ratio = 4) -- the outcomes are variable, with either the Courter-Parent or Courter-Nonparent morphs being fixed. When the non-courters have a much higher reproductive allocation than courters (8 offspring compared to 2, ratio = 0.25), the Courter-Parent morph is never fixed, and instead the Courter-Nonparent morph is fixed. This pattern is also found when Courters and Noncourters have equivalent allocations (4 and 4). Intriguingly, at the ratio of 0.5, when Noncourters have an allocation of 8 and Courters have an allocation of 4, variation is maintained in a handful of the simulations, with a very small number of Noncourter-Nonparent morphs still occurring in the population along with Courter-Parent morphs. 

```{r echo=FALSE}
params[2,"unlinked"]<-"Nuanced effect. High ratios, courters are always fixed. Low allocations, Courter-Nonparent morph is usually fixed, except at ratio 0.25"
```

### Parental nest survival


```{r}
cpu_psurv<-plot.pc.reps(pattern="^parent-courter_unlinked_psurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_psurv,cols2,ncols = 4)
cpu_psurv_freqs<-get.morph.freqs(cpu_psurv)[,-1]
cpu_psurv_freqs$ParentSurvival<-gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_psurv_freqs))
cpu_psurv_freqs<-rbind(cpu_psurv_freqs[,c(base_keep_cols,"ParentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"ParentSurvival")])
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_psurv_freqs,"ParentSurvival",cols2,"Parent nest survival")
```

Parental nest survival does not appear to have a major effect on the outcomes. Occasionally, at several of the lower survival rates (0.5, 0.7, 0.8) a very small proportion of males have the Noncourter-Nonparent morph when the vast majority of males have the Courter-Parent morph.

```{r echo=FALSE}
params[3,"unlinked"]<-"No major effect"
```

### Non-parental nest survival


```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_unlinked_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"NonparentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"NonparentSurvival")])
```

```{r,fig.cap="Frequency of morphs with unlinked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"NonparentSurvival",cols2,"Non-Parent nest survival")
```

No major effects observed, although when non-parent nest survival was 0.3 every run had the outcome of fixed Courter-Parent morph, whereas in other cases the Courter-Nonparent morph was also fixed occasionally. In the case where survival was 0.5, one or two runs maintained tiny amounts of polymorphism, with near-zero positive frequencies of Noncourter-Nonparent morphs occurring alongside the Courter-Parent morphs.

```{r echo=FALSE}
params[4,"unlinked"]<-"No major effect"
```

## Linked QTLs

In some of the cases with linked QTLs, polymorphism is maintained in some runs. Therefore, I'm also showing genetic variation and trait variation for each set of parameter settings.


```{r}
cpu_baseline<-up<-plot.pc.reps("^parent-courter_linked_.*_summary.txt",path="baseline",cols,make.plot=FALSE)
cpu_baseline_freqs<-get.morph.freqs(cpu_baseline)
cpu_baseline_freqs<-cpu_baseline_freqs[,-1]
cpu_baseline_freqs$Polygyny<-FALSE
cpu_baseline_freqs$CourterRS<-8
cpu_baseline_freqs$NoncourterRS<-4
cpu_baseline_freqs$ParentSurvival<-0.9
cpu_baseline_freqs$NonparentSurvival<-0.1

base_keep_cols<-grep("mean|sd|W|Thresh|Progeny|Pref",colnames(cpu_baseline_freqs),invert = TRUE,value = TRUE)[1:10]
```
```{r echo=FALSE}
params$linked<-NA
```

```{r,fig.cap="Allele frequencies at QTLs in the final generation of all of the runs - baseline",fig.height=10,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_.*",path="baseline",4,cols)
```

Here, the distribution of allele frequencies at QTLs for the Courter trait and the Parent trait are shown for the baseline parameter settings. These graphs demonstrate that substantial genetic variation can be maintained at QTLs even when the traits themselves have become fixed.

```{r,fig.cap="Trait values in the final generation of all the runs: baseline with linked QTLs",fig.height=10,fig.width=8}
plot_final_traits(pattern="^parent-courter_linked_.*traits.txt",path="baseline",ncols=4,cols=cols,cols2 = cols2)
```

These trait combinations reflect what I've previously observed, that a single morph becomes fixed in the population, with the Courter-Parent morph being fixed when the parental trait evolves to be positive and the Courter-Nonparent morph becoming fixed if the parental trait evolves to have negative values.

### Polygyny

```{r}
cpu_polygyny<-plot.pc.reps(pattern="^parent-courter_linked_polygyny_.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_polygyny,cols2,ncols = 4)
cpu_polygyny_freqs<-get.morph.freqs(cpu_polygyny)[,-1]
cpu_polygyny_freqs$Polygyny<-TRUE
#merge with baseline
cpu_polygyny_freqs<-rbind(cpu_polygyny_freqs[,c(base_keep_cols,"Polygyny")],cpu_baseline_freqs[,c(base_keep_cols,"Polygyny")])
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with and without polygyny",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_polygyny_freqs,"Polygyny",cols2,"Mating System",x_lim = c(-0.25,1.25),yname_pos = -0.35,xaxt='n')
axis(1,at=c(0,1),labels=c("Monogamy","Polygyny"),lwd=0)
```

With linked QTLs, the mating system has an effect. Polygyny allows variation to be maintained, with every run resulting in a population of *mostly* Courter-Parents, but some small numbers of Courter-Nonparents, Nonparent-Courters, and/or NonCourter-Nonparents. From the way these results are presented, it's challenging to determine whether all three morphs are maintained in the same run. To get at that, we can plot each run's outcomes:

```{r,fig.cap="Frequencies over time in each run with linked QTLs, both traits, and polygyny",fig.height=8,fig.width=8}
plot.morphs.reps(cpu_polygyny,cols2,ncols=4)
```

It appears that in one set of runs, all three morphs are maintained simultaneously.

```{r echo=FALSE}
params[1,"linked"]<-"Allows polymorphism sometimes"
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - polygyny",fig.height=10,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_polygyny.*",path="sensitivity",4,cols)
```

Interestingly, genetic variation is found in the runs that don't have polymorphism either (but remember that allele frequencies are not necessarily reflective of additive genetic variation).

```{r,fig.cap="Trait values in the final generation of all the runs: polygyny with linked QTLs",fig.height=7,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_polygyny.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

Substantial trait variation is maintained when the courter trait and parent trait evolve overlapping trait values that span a substantial range.


### Reproductive Allocations


```{r}
cpu_rs<-plot.pc.reps(pattern="^parent-courter_linked_crs.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_rs,cols2,ncols = 4)
cpu_rs_freqs<-get.morph.freqs(cpu_rs)[,-1]
cpu_rs_freqs$CourterRS<-gsub(".*crs(\\d)_ncrs\\d.*","\\1",rownames(cpu_rs_freqs))
cpu_rs_freqs$NoncourterRS<-gsub(".*crs\\d_ncrs(\\d).*","\\1",rownames(cpu_rs_freqs))
#merge with baseline
cpu_rs_freqs<-rbind(cpu_rs_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")],cpu_baseline_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")])

cpu_rs_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_rs_freqs$CourterRS)/as.numeric(cpu_rs_freqs$NoncourterRS))
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with different ratios of courter to non-courter reproductive success",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_rs_freqs,"Courter2NonRS",cols2,"Ratio of Reproductive Allocation\nCourter:Non-Courter",xaxt='n',x_lim = c(0.15,5.5),yname_pos = -0.2)
axis(1,at=levels(as.factor(as.numeric(cpu_rs_freqs$Courter2NonRS))),labels = levels(cpu_rs_freqs$Courter2NonRS))

```

In general, high ratios generally result in Courter-Parent morph to be fixed, although at very high ratios some polymorphism with Noncourter-Nonparent is observed. At the lowest ratio, when Noncourters have 8 offspring and courters have 2, in almost all cases the Courter-Nonparent morph is fixed, with one exception where the Courter-Parent morph is fixed.

```{r echo=FALSE}
params[2,"linked"]<-"Nuanced effect. High ratios generally result in Courter-Parent morph to be fixed, although at very high ratios some polymorphism with Noncourter-Nonparent is observed. At the lowest ratio, usually Courter-Nonparent is fixed instead of Courter-Parent"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 2, noncourter RS = 8",fig.height=8,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs2_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 2, noncourter RS = 8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs2_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 4, noncourter RS = 8",fig.height=8,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs4_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 4, noncourter RS = 8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs4_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 8, noncourter RS = 2",fig.height=8,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 2 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs8_ncrs2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

This is interesting because in these cases for multiple morphs to be maintained the parental trait has to be non-positive but not necessarily overlapping the courter trait values. 


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - courter RS = 8, noncourter RS = 8",fig.height=8,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_crs8_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

And here, no variation is maintained and the Courter-Parent morph is always maintained.



### Parental nest survival


```{r}
cpu_psurv<-plot.pc.reps(pattern="^parent-courter_linked_psurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_psurv,cols2,ncols = 4)
cpu_psurv_freqs<-get.morph.freqs(cpu_psurv)[,-1]
cpu_psurv_freqs$ParentSurvival<-gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_psurv_freqs))
cpu_psurv_freqs<-rbind(cpu_psurv_freqs[,c(base_keep_cols,"ParentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"ParentSurvival")])
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_psurv_freqs,"ParentSurvival",cols2,"Parent nest survival")
```

In general, the Courter-Parent or Courter-Nonparent traits become fixed, although with parental nest survival at 0.7 only the Courter-Parent morph became fixed (though this may be an artefact of not running huge numbers of runs).

```{r echo=FALSE}
params[3,"linked"]<-"No major effect"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.5",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.5 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_psurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.6",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.6.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.6 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_psurv0.6.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.7",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.7.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.7 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_psurv0.7.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.8",fig.height=10,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_psurv0.8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.8 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_psurv0.8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

### Non-parental nest survival


```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_linked_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"NonparentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"NonparentSurvival")])
```

```{r,fig.cap="Frequency of morphs with linked QTLs in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"NonparentSurvival",cols2,"Non-Parent nest survival")
```

In general, the Courter-Parent or Courter-Nonparent morph becomes fixed, although occasionally the Courter-Nonparent morph is fixed instead. The highest survival rate (0.5) allows the Courter-Nonparent morph to be fixed more frequently.

```{r echo=FALSE}
params[4,"linked"]<-"High survival allows higher frequency of Courter-Nonparent morph"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.2",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.2.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.2 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.3",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.3.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.3 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.3.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.4",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.4.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.4 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.4.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.5",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_linked_npsurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.5 with linked QTLs",fig.height=4,fig.width=7}
plot_final_traits(pattern="^parent-courter_linked_npsurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```



## QTLs in a supergene

We'll start by looking at the unlinked QTLs.

```{r}
cpu_baseline<-up<-plot.pc.reps("^parent-courter_supergene_.*_summary.txt",path="baseline",cols,make.plot=FALSE)
cpu_baseline_freqs<-get.morph.freqs(cpu_baseline)
cpu_baseline_freqs<-cpu_baseline_freqs[,-1]
cpu_baseline_freqs$Polygyny<-FALSE
cpu_baseline_freqs$CourterRS<-8
cpu_baseline_freqs$NoncourterRS<-4
cpu_baseline_freqs$ParentSurvival<-0.9
cpu_baseline_freqs$NonparentSurvival<-0.1

base_keep_cols<-grep("mean|sd|W|Thresh|Progeny|Pref",colnames(cpu_baseline_freqs),invert = TRUE,value = TRUE)[1:10]
```

```{r echo=FALSE}
params$supergene<-NA
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - baseline",fig.height=10,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_.*",path="baseline",4,cols)
```

```{r,fig.cap="Trait values in the final generation of all the runs: polygyny with a supergene",fig.height=10,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_.*traits.txt",path="baseline",ncols=4,cols=cols,cols2 = cols2)
```


### Polygyny

```{r}
cpu_polygyny<-plot.pc.reps(pattern="^parent-courter_supergene_polygyny_.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_polygyny,cols2,ncols = 4)
cpu_polygyny_freqs<-get.morph.freqs(cpu_polygyny)[,-1]
cpu_polygyny_freqs$Polygyny<-TRUE
#merge with baseline
cpu_polygyny_freqs<-rbind(cpu_polygyny_freqs[,c(base_keep_cols,"Polygyny")],cpu_baseline_freqs[,c(base_keep_cols,"Polygyny")])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with and without polygyny",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_polygyny_freqs,"Polygyny",cols2,"Mating System",x_lim = c(-0.25,1.25),yname_pos = -0.35,xaxt='n')
axis(1,at=c(0,1),labels=c("Monogamy","Polygyny"),lwd=0)
```

Similar to the case with linked QTLs, polygyny allows polymorphism to be maintained in some runs, this time with Courter-Parent, Courter-Nonparent, and Noncourter-Parent morphs all being maintained in the population.

```{r echo=FALSE}
params[1,"supergene"]<-"Allows polymorphism sometimes"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - polygyny",fig.height=8,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_polygyny.*",path="sensitivity",4,cols)
```

When variation is maintained, the genetic variation maintained is much greater as well.

Trait values are recorded for each population at the final generation and written to the *traits.txt files.

```{r,fig.cap="Trait values in the final generation of all the runs: polygyny with a supergene",fig.height=6,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_polygyny.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


### Reproductive Allocations


```{r}
cpu_rs<-plot.pc.reps(pattern="^parent-courter_supergene_crs.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_rs,cols2,ncols = 4)
cpu_rs_freqs<-get.morph.freqs(cpu_rs)[,-1]
cpu_rs_freqs$CourterRS<-gsub(".*crs(\\d)_ncrs\\d.*","\\1",rownames(cpu_rs_freqs))
cpu_rs_freqs$NoncourterRS<-gsub(".*crs\\d_ncrs(\\d).*","\\1",rownames(cpu_rs_freqs))
#merge with baseline
cpu_rs_freqs<-rbind(cpu_rs_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")],cpu_baseline_freqs[,c(base_keep_cols,"CourterRS","NoncourterRS")])

cpu_rs_freqs$Courter2NonRS<-as.factor(as.numeric(cpu_rs_freqs$CourterRS)/as.numeric(cpu_rs_freqs$NoncourterRS))
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different ratios of courter to non-courter reproductive success",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_rs_freqs,"Courter2NonRS",cols2,"Ratio of Reproductive Allocation\nCourter:Non-Courter",xaxt='n',x_lim = c(0.15,5.5),yname_pos = -0.2)
axis(1,at=levels(as.factor(as.numeric(cpu_rs_freqs$Courter2NonRS))),labels = levels(cpu_rs_freqs$Courter2NonRS))

```

Now polymorphism can be maintained when the ratios are very low. In one set of simulations when Noncourters had the reproductive advantage, the Courter-Parent morphs coexisted with Courter-Nonparents and Noncourter-Parents. Otherwise, similar patterns were observed with either Courter-Parent morphs of Courter-Nonparent morphs becoming fixed. 

```{r echo=FALSE}
params[2,"supergene"]<-"Low ratios allow polymorphism in some cases"
```

Genetic composition?

```{r}
mrks<-read.delim("sensitivity/parent-courter_supergene_crs2_ncrs8_2_markers.txt")
qtls<-read.delim("sensitivity/parent-courter_supergene_crs2_ncrs8_2_qtlinfo.txt")
qtls<-qtls[,complete.cases(t(qtls))] #remove nas
qtl_af<-cbind(mrks[,1:2],mrks[,colnames(mrks) %in% paste("Marker",qtls[1,-1],sep="")])
qtl_af<-qtl_af[qtl_af$Pop=="Pop0",]

courter_qtls<-paste("Marker",qtls[1,grep("Courter",colnames(qtls))],sep="")
parent_qtls<-paste("Marker",qtls[1,grep("Parent",colnames(qtls))],sep="")

qtl_dat<-do.call(rbind,apply(qtl_af,1,function(qtlr,parent_inx){
  freqs<-qtlr[3:length(qtlr)]
  gen<-rep(qtlr[1],length(freqs))
  trait<-rep("Courter",length(freqs))
  trait[parent_inx]<-"Parent"
  return(data.frame(freqs=freqs,generation=gen,trait=trait))
},parent_inx=which(colnames(qtl_af) %in% parent_qtls)-2))
qtl_dat$generation<-as.numeric(as.character(qtl_dat$generation))
qtl_dat$freqs<-as.numeric(as.character(qtl_dat$freqs))

```
```{r plotDensities, eval=FALSE,echo=FALSE}

par(mar=c(5,4,4,4),oma=c(2,2,2,2),xpd=TRUE)
plot(c(1,length(intervals)),c(0,1),type='n',xlab="Generations",ylab="Allele Frequencies",xaxt='n',bty="L")
axis(1,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  dC<-density(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"],from=0,to=1)
  #dP<-density(qtl_dat[qtl_dat$trait=="Parent" & qtl_dat$generation ==int,"freqs"])
  polygon(dC$y+cnts,dC$x,col = alpha(cols["courter"],0.5)) 
  #rug(int,jitter(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"]))
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))


plot(c(0,1),c(1,length(intervals)),type='n',ylab="Generations",xlab="Allele Frequencies",yaxt='n',bty="L")
axis(2,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  dC<-density(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"],from=0,to=1)
  #dP<-density(qtl_dat[qtl_dat$trait=="Parent" & qtl_dat$generation ==int,"freqs"])
  polygon(dC$x,dC$y+cnts,col = alpha(cols["courter"],0.5))
  rug(jitter(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"]))
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))
```
```{r}
library(scales)
intervals<-c(seq(0,11999,500),11999)

par(mfrow=c(2,1),mar=c(2,2,1,1),oma=c(2,2,2,2),xpd=TRUE)
plot(c(1,length(intervals)),c(0,length(seq(0,1,0.05))),type='n',xlab="",ylab="",yaxt='n',xaxt='n',bty="L")
axis(2,at=0:(length(seq(0,1,0.05))-1),labels=seq(0,1,0.05))
axis(1,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  h<-hist(qtl_dat[qtl_dat$trait=="Courter" & qtl_dat$generation ==int,"freqs"],breaks=seq(0,1,0.05),plot = FALSE)
  frq<-h$counts/sum(h$counts)
  barplot(frq,axes=FALSE,space=0,col = alpha(cols["courter"],0.5),offset=cnts,add = TRUE,horiz = TRUE,border=cols["courter"])
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))

plot(c(1,length(intervals)),c(0,length(seq(0,1,0.05))),type='n',xlab="Generations",ylab="Parent Allele Frequencies",yaxt='n',xaxt='n',bty="L")
axis(2,at=0:(length(seq(0,1,0.05))-1),labels=seq(0,1,0.05))
axis(1,at=1:length(intervals),labels = intervals)
d<-mapply(function(int,cnts,qtl_dat){
  h<-hist(qtl_dat[qtl_dat$trait=="Parent" & qtl_dat$generation ==int,"freqs"],breaks=seq(0,1,0.05),plot = FALSE)
  frq<-h$counts/sum(h$counts)
  barplot(frq,axes=FALSE,space=0,col = alpha(cols["parent"],0.5),offset=cnts,add = TRUE,horiz = TRUE,border=cols["parent"])
},intervals,c(1:length(intervals)),MoreArgs=list(qtl_dat=qtl_dat))

mtext("Generations",1,outer=TRUE)
mtext("Allele Frequencies",2,outer=TRUE)

```



Let's look at the final allele frequencies across all reps

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 2, noncourter RS = 8",fig.height=4,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs2_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 2, noncourter RS = 8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs2_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 4, noncourter RS = 8",fig.height=6,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs4_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 4, noncourter RS = 8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs4_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 8, noncourter RS = 2",fig.height=4,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 2 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs8_ncrs2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - supergene and courter RS = 8, noncourter RS = 8",fig.height=4,fig.width=8.5}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_crs8_ncrs8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: courter RS = 8, noncourter RS = 8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_crs8_ncrs8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

### Parental nest survival


```{r}
cpu_psurv<-plot.pc.reps(pattern="^parent-courter_supergene_psurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_psurv,cols2,ncols = 4)
cpu_psurv_freqs<-get.morph.freqs(cpu_psurv)[,-1]
cpu_psurv_freqs$ParentSurvival<-gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_psurv_freqs))
cpu_psurv_freqs<-rbind(cpu_psurv_freqs[,c(base_keep_cols,"ParentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"ParentSurvival")])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_psurv_freqs,"ParentSurvival",cols2,"Parent nest survival")
```

Here, the lowest parental nest survival rate (0.5) resulted in all cases having the Courter-Nonparent morph becoming fixed, whereas in the other runs either the Courter-Parent or the Courter-Nonparent could become fixed.

```{r echo=FALSE}
params[3,"supergene"]<-"Low survival results in coevolution of Courter and Non-parent traits"
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.5",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.5 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.6",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.6.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.6 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.6.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.7",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.7.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.7 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.7.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - parent nest survival = 0.8",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_psurv0.8.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: parent nest survival = 0.8 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_psurv0.8.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

### Non-parental nest survival


```{r}
cpu_npsurv<-plot.pc.reps(pattern="^parent-courter_supergene_npsurv.*_summary.txt",path="sensitivity",cols,make.plot=FALSE)
#plot.morphs.reps(cpu_npsurv,cols2,ncols = 4)
cpu_npsurv_freqs<-get.morph.freqs(cpu_npsurv)[,-1]
cpu_npsurv_freqs$NonparentSurvival<-as.numeric(gsub(".*psurv(\\d.\\d).*","\\1",rownames(cpu_npsurv_freqs)))
cpu_npsurv_freqs<-rbind(cpu_npsurv_freqs[,c(base_keep_cols,"NonparentSurvival")],cpu_baseline_freqs[,c(base_keep_cols,"NonparentSurvival")])
```

```{r,fig.cap="Frequency of morphs with a supergene in scenarios with different survival rates for nests of the parental morph",fig.height=8,fig.width=8}
plot_morph_freqs(cpu_npsurv_freqs,"NonparentSurvival",cols2,"Non-Parent nest survival")
```

High survival of the non-parental nests allows polymorphism to be maintained -- Courter-Parent, Courter-Nonparent, and Noncourter-Parent morphs are all maintained at appreciable frequencies when survival was 0.5. In the other cases, either the Courter-Parent or the Courter-Nonparent morphs became fixed.

```{r echo=FALSE}
params[4,"supergene"]<-"High survival allows polymorphism"
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.2",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.2.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.2 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.2.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```


```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.3",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.3.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.3 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.3.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.4",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.4.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.4 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.4.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```

```{r,fig.cap="Allele frequencies in the final generation of all of the runs - non-parental nest survival = 0.5",fig.height=5,fig.width=8}
all_af_dat<-plot_final_af(base_pattern="^parent-courter_supergene_npsurv0.5.*",path="sensitivity",4,cols)
```
```{r,fig.cap="Trait values in the final generation of all the runs: non-parent nest survival = 0.5 with a supergene",fig.height=4,fig.width=8}
plot_final_traits(pattern="^parent-courter_supergene_npsurv0.5.*traits.txt",path="sensitivity",ncols=4,cols=cols,cols2 = cols2)
```



# Summary

```{r}
kable(params)
```

The genetic architecture of traits impacts whether variation can be maintained, and the extent of genetic variation that is maintained. Further follow-up:

* Do simple FST comparisons between the morphs?
* How does the supergene evolve over time?
* Other sensitivity paramters, plus the key parameters of interest
