---
title: "Genetic architecthure of alternative reproductive tactics"
author:
- affilnum: 1
  corresponding: yes
  email: spflanagan.phd@gmail.com
  name: Sarah P. Flanagan
- affilnum: 2
  name: Suzanne Alonzo
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    template: manuscript.latex
    toc: yes
    toc_depth: 2
  html_document: null
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document: null
zotero: 
  library: "My Library"
  csl-style: proceedings-of-the-royal-society-b
capsize: normalsize
documentclass: article
editor_options:
  chunk_output_type: console
  markdown: 
    wrap: 72
fontsize: 11pt
abstract: |
header-includes: |
  \usepackage{lipsum} \usepackage{float} \floatplacement{figure}{H} \usepackage{luamplib} \usepackage{unicode-math} \usepackage{luatex} \usepackage{luatexversion}   \usepackage[utf8]{inputenc}
affiliation:
- affil: School of Biological Sciences, University of Canterbury, 4800 Private Bag,
    Christchurch 8140 New Zealand
  affilnum: 1
- affil: University of California Santa Cruz, Santa Cruz, California, USA
  affilnum: 2
preprint: no
spacing: singlespacing
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200,fig.height = 7,fig.width = 7)
knitr::opts_knit$set(root.dir='../results/') #change

```
```{r opts, echo = FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "figs/"
)
```
```{r librarysetup, echo=FALSE,echo=FALSE,message=FALSE}
library(RColorBrewer)
library(scales)
library(sm)
library(kableExtra)
library(plotly)
library(dplyr)
library(vegan)
library(tidyr)
tmp<-sapply(list.files(path="../../spfTools/R/",full.names = TRUE),source)
source("../R/002_freq_functions.R")

source("../morph_predictions/check_freqs.R")
source("../morph_predictions/morph_gens_ns.R")

cols<-c(courter="#33a02c",parent="#1f78b4",cpref="#b2df8a",ppref="#a6cee3")
cols2<-c(CP='#d7191c',NCP='#fdae61',CNP='#abd9e9',NCNP='#2c7bb6')
```
```{r baseCols}
base_keep_cols<- c("Pop","PopSize","NumMal","NumFem","ParentThresh","ParentFreq","CourterThresh","CourterFreq","FreqNcNp","FreqCNp",
                   "FreqNcP","FreqCP","Courter2NonRS","Parent2NonSurvival")
poly_keep_cols<- c("Pop","PopSize","NumMal","NumFem","ParentThresh","ParentFreq","CourterThresh","CourterFreq","FreqNcNp","FreqCNp",
                    "FreqNcP","FreqCP","Polygyny")

```


# Introduction

Both genetic and phenotypic variation are required for evolution to occur, and as such the maintenance of variation in populations has long been of interest to evolutionary biologists and conservation geneticists. Evolution is often expected to remove variation from populations, with specific patterns of selection being predicted to maintain variation at a high frequency -- including overdominance, sexually antagonistic selection [@connallonEvolutionaryConsequencesSexSpecific2018; @connallonResolutionSexualAntagonism2011], and frequency dependent selection [@alonzoMateChoiceGames2001]. Yet, substantial phenotypic variation is observed, with ubiquitous examples such as the prevalence of sexual dimorphism and the unexpectedly frequent existence of alternative reproductive tactics, in which individuals of the same species and sex display different morphs to maximize reproductive fitness in unique ways [@oliveiraAlternativeReproductiveTactics2008]. 

The genetic basis of traits experiencing selection is also critical to understanding the maintenance of variation. Evolution can be constrained by factors including pleiotropy, genetic correlations, and physical linkage. As such, the relaxation of these constraints can facilitate sequence evolution by relaxing positive selection on genes [@harrisonSexualSelectionDrives2015]. However, the interaction of selection that facilitates variation with the genetic architecture and genetic constraints remains under-explored. In particular, the majority of predictions relating selection to genetic constraints arise from one- or two-locus models, which can be difficult to extrapolate to genome-wide predictions. In the modern genomic era, having predictions for large genome-wide datasets is of critical importance to develop testable hypotheses. [something about genome-wide predictions]. 

One genetic mechanism that is increasingly shown to play an important role in evolution are structural variants, such as inversion polymorphisms that harbour so-called 'supergenes'. These variants are regions of the genome in which some individuals carry an inversion, in which genes with direct phenotypic effects exist. The mismatched directionality reduces recombination between the different variants, facilitating the maintenance of both genetic and phenotypic variations. These supergenes have been shown to underpin stable polymporhisms including ant morphs, immune response genes (e.g., toll-like receptors), xxx, and alternative reproductive tactics (e.g., in white-collared sparrows [@thomasChromosomalPolymorphismLinked2008] and ruffs [@kupperSupergeneDeterminesHighly2016; @lamichhaneyStructuralGenomicChanges2016]). These polymorphisms are often maintained in populations through frequency dependent selection, and in the case of sexually dimorphic variants, might resolve sexual conflict. **Need to look into the literature for more examples and to add detail**.

Substantial advances in our understanding of the evolution of these traits have been made with whole-genome sequencing and transcriptome studies. Furthermore, recent mathematical and simulation models have helped clarify the allelic variation that is expected to be maintained under sex-specific selection pressures [@kasimatisLimitsGenomicDivergence2019a], OTHER EXAMPLES. However, providing direct linkages between theoretical predictions of evolutionary mechanisms, genotype-to-phenotype maps, and realistic datasets are currently lacking. Here, we fill this gap using individual-based simulation models of alternative reproductive tactics in males and three types of genetic architectures for the traits of interest.


# Methods

Our model focusses on a single population of diploid males and females
with non-overlapping generations. We model two independent traits,
courtship and parental care, both of which are expressed only in males,
but are also inherited by females. As such, males are one of four
morphs: courter/parents; courter/non-parents; non-courters/parents; or
non-courters/non-parents. The model is inspired by fish species like the
ocellated wrasse, *Symphodus ocellatus*, and as such we will use
terminology appropriate to that species. Nonetheless, our model can be
generalized to any species in which one sex (but not the other) has two
traits that impact mating success and the survival of their offspring.

Each generation has three distinct phases: (1) mate choice and
fertilization; (2) parental care; and (3) viability selection. In the
mate choice phase of the model, females always prefer to mate with males
displaying the courtship trait. If a female decides to mate with a male,
he receives all of her eggs to care for. The male is not guaranteed
paternity of his mates' entire brood, however, because non-courting
males are able to fertilize a proportion of the eggs. The chosen parent
does receive first-male priority. The consequences of this setup are
that the only mechanism by which non-courters can produce offspring is
through these sneak fertilizations. In the parental care phase of the
model, males with the parental care trait have 100% nest survival while
males without the parental care trait have 0% nest survival, independent
of the male's courtship trait. Consequently, non-parents will only be
sustained in the population by sneaking fertilizations in the parental
males' nests. Finally, viability selection is imposed on offspring as
they mature to adulthood, such that males carrying the courtship trait
and/or the parental care trait have a lower probability of survival than
females and males without either courtship or parenting. Males who are
both courters and parents have a multiplicative effect of viability
selection. **Who is allowed to sneak?**

We implemented our model first as a mathematical model and then as an
individual-based simulation model. The purpose of the mathematical model
was to generate baseline predictions for when multiple morphs are
maintained.


## Frequency-dependent population dynamics model

In the mathematical model, we model the proportion of the population
that belong to morphs at each stage of the lifecycle (Fig.
\@ref(fig:mathModel)). In the first stage, we consider the proportion of
the population's nests that will be allocated to each morph. The maximum
number of nests that could be laid is proportional to the number of
males in the population ($N_m$), if each male has the same likelihood of
attracting a female to deposit eggs with him. However, in our model,
males are not equally likely to be chosen by females, and therefore the
maximum number of nests is scaled by the probability of females mating
with each type of male morph. In the case of our model, the total number
of nests laid in the population becomes $$
\frac{N_f}{N_mf_{CP}+N_m f_{CNp}}
$$ because we have constrained females to only mate with courting males.

```{r mathModel, fig.cap="Diagram of the mathematical model describing the inheritance of alternative male reproductive morphs given courtship and parenting traits."}
knitr::include_graphics("../figs/frequency_dependence.png")
```

Likewise, the proportion of nests each type of male receives (${p_n}_M$)
is determined by the number of morphs in the population ($N_mf_M$ for
the $M$ morph), the probability of a female choosing that morph
(${w_s}_M$), and the ratio of the number of nests to the number of
females ($n/N_f$).

Within a given nest, the offspring could either be fertilized by the
nest-holding male or non-courting males 'sneaking' fertilizations.
Therefore, each morph has a proportion of eggs fertilized within their
own nests (${p_{fn}}_M$) and a proportion of eggs fertilized in other
males' nests (${p_{fs}}_M$). These proportions were determined by the
sperm competition coefficient, $c$, which penalized non-courting males,
and the relative reproductive investment for that morph ($r_M$).
Analogously, the proportion of offspring in a nest surviving due to
parental care investment can be split into the proportion surviving in a
male's own nest (${p_{sn}}_M$) or in another male's nest (${p_{ss}}_M$),
which is simply the proportions from the fertilization step weighted by
the probability of the nest surviving if the morph is a parent ($w_n=1$)
or of the morph is a non-parent ($w_n=0$). These proportions
(${p_{sn}}_M$ and ${p_{ss}}_M$) are summed to determine the the overall
proportion of embryos in the population that survive the parenting phase
in each morph's nests (${p_s}_M$).

The final stage of the lifecycle is for offspring surviving in the nest
to undergo viability selection by multiplying the viability selection
coefficient (${w_v}_M$) by the surviving proportion (${p_s}_M$) and the
original frequency of the morph in the population (**Note: does pv need
the frequency multiplication??)**. Finally, the frequency of each morph
in the next generation was calculated by dividing the surviving
proportion (${p_v}_M$ by the sum of all proportions.

To evaluate the outcomes of this mathematical model, we iterated forward
in time for 100 generations from 1659 combinations of initial morph
frequencies and repeated those iterations across 5 values for the sperm
competition coefficient ($c$ = {0..1}) and 21 values for the relative
reproductive coefficient ($r_M$ = {0...2}). The iterations were
implemented in R v. 4.1.1 [@rcoreteam2021] using some features of tidyR
[@wickham2021]. To facilitate visualisation of the results, we
calculated the Shannon's diversity index for the morphs remaining in the
100th generation using the R package vegan [@oksanen2015]. We then
visualised the diversity of the population using contour plots produced
by plotly [@sievert2020] in a shiny app that relies on R packages shiny
[@chang2021], shinydashboard [@chang2018], rsconnect [@mcpherson2021],
dplyr [@wickham2021], and rmarkdown [@allaire2021; @xie2020; @xie2018].
The interactive app is available at
<https://spflanagan.shinyapps.io/morph_predictions/>.

**this model has no recombination/females do not pass on a morph**

## Individual-based simulation model

To evaluate the role of genetic architecture in the maintenance of
alternative reproductive tactics, we wrote an individual-based
simulation model in C++. The simulation model follows the same lifecycle
as described above, but rather than tracking proportions of the
population belonging to particular morphs, we tracked individuals within
the population. Each individual contained diploid genotypes, which
included both non-coding loci and quantitative trait loci (QTLs). We
included two types of QTLs: ones that impacted the individual's courting
trait and ones that impacted the individual's parenting trait. The QTLs
were each assigned a randomly-drawn effect on the trait, and the QTLs
additively determined the trait value for each trait (i.e., the effects
of all QTLs were summed to determine the trait value). In this model, we
assume no environmental effects impacted the trait value. These trait
values were then compared to a threshold to dichotomize each trait. The
threshold for each trait was the mean trait value during the
initialization of the model (i.e., at generation 0). 

We implemented three different genetic architectures: a single gene per
trait, multiple QTLs randomly placed throughout chromosomes, and
supergenes. The single gene scenario is most analogous to the
mathematical model described above. When the traits have multiple QTLs
determining their trait values, each QTL is assigned a location on a
chromosome and the QTL's effect on the trait is associated with a single
nucleotide polymorphism (SNP). These SNPs undergo recombination and
mutation during mating. The supergene scenario forced the QTLs to
cluster into a small proportion of one of the chromosomes. These QTLs
still experience mutation but experienced reduced recombination.

The lifecycle:

When we ran the model, we initialized four replicate populations with
the identical starting conditions to evaluate the impact of
stochasticity on the model. These identical conditions included
randomly-generated values for the allelic effects of QTLs, exact number
of individuals of each sex and morph, and thresholds for determining
morph traits (i.e., the values at which an individual is deemed a
courter vs a non-courter and a parent vs a non-parent). For each set of
four populations, we ran 10,000 burn-in generations followed by 2,000
'experimental' generations. Doing so ensured that the population reached
a steady state in which we could examine any stable polymorphisms or
cyclical fluctuations in morph frequencies.

Within each of those generations, the population underwent four main
phases: (1) mate choice and fertilization; (2) nest survival; (3)
viability selection on juveniles; (4) maturation to adulthood. The main
difference between the individual-based simulation model and the
mathematical model were the introduction of stochasticity during mate
choice and the introduction of recombination and mutation during
reproduction. During the mate choice phase, females searched through a
set number of males and chose the first attractive male (i.e., the first
courter male she encountered). If a female did not find an acceptable
mate, she either did not mate ('no mating' or 'nm' scenarios) or she 
chose a male at random with from the population and allocated
all of her offspring to him, depending on the parameter settings.
Females always preferred courting males.

Once a female chose a male, fertilization of the eggs occurred, in which
the chosen male had first male priority (i.e., the chosen male allocated
all of his reproductive effort to his own nest) and up to 3 non-courting
males could 'sneak' fertilizations, with their contributions discounted
by a factor of $c$. After the males were assigned their paternity
shares, the nests survived if the chosen male was also a parent.
Subsequently, the total number of offspring in the population were
created based on the proportion of population reproductive success
achieved by each individual male and female (i.e., each generation, the
number of offspring created was equal to the carrying capacity).

When fertilization occurred, the parents' genotypes (both neutral
markers and QTLs) had the opportunity to recombine. Recombination
occurred independently on the chromosome inherited from the mother and
the chromosome inherited from the father. The number of recombination
events for a given chromosome was determined by drawing from a Poisson
distribution with the recombination rate as the mean, which we set to
0.2. For each recombination event, breakpoints were chosen by randomly
drawing from the total number of markers on a chromosome. These
breakpoints defined the segments of the chromosome that experienced
recombination, and each segment was randomly chosen to be assigned the
genotypes and effects from the parents' maternal or paternal chromosome.
When the genetic architecture was that of a supergene, if the
breakpoints fell within the region containing the aggregated QTLs for
the traits, the offspring was considered not viable.

After recombination occurred, if the progeny survived, the genotypes
underwent mutation, which occurred at a rate of $mu$. For a given
individual, the number of mutations introduced was determined as
$2\mu*n_SNPs*n_chrom$. The location of each mutation was randomly
selected and the allele at that location was randomly chosen from the
set of existing alleles. The mutation also impacted the phenotypic
effects of QTLs by altering the original effect by a value drawn from a
normal distribution with a mean of 0 and a standard deviation of
$sigma_\mu$. Finally, offspring were also randomly assigned a sex.

The next stage in the lifecycle was viability selection, which impacted
only male offspring. For each male offspring, a survival probability was
calculated as $e^{\frac{-morph2}{2\omega_v}}$. Viability selection acted
multiplicatively on courter-parents (their survival probability was
$e^{\frac{-morph2}{2\omega_v}}*e^{\frac{-morph2}{2\omega_v}}$). A random
value between 0 and 1 was drawn, and if the number was larger than the
survival probability for an individual, that individual did not survive
to adulthood. If, at this point, the number of surviving offpsring was
larger than the carrying capacity, offspring were randomly selected to
become adults. When offspring matured to adulthood, the previous
generation's adults were replaced, and the lifecycle began again.

# Results

## Frequency-dependent population dynamics model

The initial frequency of each morph impacted whether multiple morphs were maintained. In no cases was the courter-nonparent morph maintained. This result makes intuitive sense, because if a female chooses to nest with a courter-nonparent, the nest will always die (given our parameter settings).  [What happens if one morph is absent?] 

The parameters controlling sperm competition ($c$), relative reproductive allocation ($r$), and the number of sneakers allowed also impacted the diversity of the population after both 100 generations and 1000 generations.  Some initial starting conditions took more than 100 generations to reach a consistent (i.e., equilibrial) state, but the contributions of these three parameters remained qualitatively consistent across 100 and 1000 generations. Reproductive allocations corresponding to a range of approximately 0.5 to 1 (meaning that courters invest half as much to just as much as non-courters in producing sperm) resulted in the maximum amount of diversity, as long as sperm competition ($c$) was above 0.2 (meaning that, if a non-courting male produced 8 sperm and courting males produce 4, the non-courting male would fertilize `r (8*0.2)/(4+(2*0.2*8))` or more of the offspring). This importance of the sperm competition coefficient was influenced by the number of sneakers allowed, with larger contributions from an individual non-courting male (e.g., larger $c$) being required with fewer sneakers (Fig. \@ref(fig:mathParams)).

```{r mathParams, fig.cap="Contour plots showing the diversity of the populations given equal initial starting frequencies of all four morphs, the relative reproductive investments (x-axis), sperm competition coefficients (y-axis), and number of sneakers allowed per nest (1 through 5, plots going from top to bottom).", fig.width=4.5, fig.height=7, warning=FALSE,message=FALSE}

results<-readRDS("../morph_predictions/morph_results_10000_equalStart.RDS")
results$diversity<-vegan::diversity(round(results[,c("CP","CN","NP","NN")],4))

contours<-by(results, results$num_sneak, function(sub_calcs){
  
  
  data_wide<-tidyr::spread(
    sub_calcs[,c("r","c","diversity")],
    r,
    diversity
  )
  rownames(data_wide)<-data_wide[,1]
  data_wide<-data_wide[,-1]
  
  fig <- plot_ly(
    x = as.numeric(colnames(data_wide)), 
    y = as.numeric(rownames(data_wide)), 
    z = as.matrix(data_wide), 
    type = "contour"
  )
  # add axis labels
  x<-list(title="r")
  y<-list(title="c")
  fig <- fig %>% layout(xaxis=x,yaxis=y)
  # add label to contour names
  fig <- fig %>% colorbar(title = "Diversity")
  # plot
  return(fig)
})


multi_fig <- subplot(contours, nrows = 5, shareY = TRUE, shareX = TRUE)

multi_fig %>% layout(autosize = F, width = 450, height = 650)


```

```{r}
divRes<-results[results$diversity>0,]
ndiv<-nrow(divRes)

pcourt<-nrow(divRes[divRes$CP>=0.5,])/ndiv
pnp<-nrow(divRes[divRes$NP>=0.5,])/ndiv
pnn<-nrow(divRes[divRes$NN>=0.5,])/ndiv


```


The regions of the parameter space the have non-zero diversity values correspond to scenarios where three of the four morphs were retained (courter-parents, non-courter-parents, and non-courter-non-parents). The dominant morph differed with various parameter combinations. Assuming equal starting frequencies, after 10000 generations, `r pcourt*100`% of parameter combinations that retained polymorphism resulted in the courter-parent morph comprising $\le$ 50% of the population. The remaining `r (1-pcourt)*100`% of parameter combinations resulting in polymorphism had more equal representation of all three morphs, with the average proportion of the population being `r mean(divRes[divRes$CP<=0.5,"CP"])*100` % ($pm$ `r sem(divRes[divRes$CP<=0.5,"CP"])*100` %) courter-parents, `r mean(divRes[divRes$CP<=0.5,"NP"])*100` % ($pm$ `r sem(divRes[divRes$CP<=0.5,"NP"])*100` %) non-courter-parents, and `r mean(divRes[divRes$CP<=0.5,"NN"])*100` % ($pm$ `r sem(divRes[divRes$CP<=0.5,"NN"])*100` %) non-courter-non-parents.    

## Individual-based simulation model

The results from the  mathematical model were used to select parameter settings for the individual-based simulation model, with one set expected to produce low-diversity populations (i.e., those where only one morph were retained) and another set expected to produce high-diversity populations (i.e., where 2 or more morphs were retained). The chosen parameter settings are shown in Table \@ref(tab:SimParamsTable). 


```{r SimParamsTable}
SimParams<-data.frame(
  parameter = c("r (in shiny app)",
                "r_CP",
                "r_CN",
                "r_NP",
                "r_NN",
                "c",
                "number of sneakers"),
  LowDiv = c( 2, 
             8,
             8,
             4,
             4,
             0.5,
             2),
   HighDiv = c( 0.75, 
             6,
             6,
             8,
             8,
             0.7,
             2)
)

kable(SimParams, "latex",booktab=TRUE,
      col.names = c("Parameter", "Low Diversity Value","High Diversity Value"),
      caption="Table of parameters selected to create a high-diversity population. All morphs started at equal frequencies.")  %>% 
  kableExtra::kable_styling(latex_options = "HOLD_position")

```



### Impacts of mating system on the model



As expected, in the low diversity parameter space, the majority of simulations resulted in fixation or near-fixation of the courter-parent morph (Fig. \@ref(fig:MatingTypesFig)). A more surprising result was the inconsistency of results from the mating types within the the high diversity parameter space. Polygyny resulted in the fixation of the courter/parent morph regardless of whether females mated at random if a suitable mate was found, and monogamy resulted in the fixation of the non-courter/parent morph in every case. Only monogamy with no mating allowed resulted in polymorphism, with the courter/parent morph coexisting with either the non-courter/parent or the non-courter/non-parent morph (Fig. \@ref(fig:MatingTypesFig)). These results can be explained by the nature of density-dependent survival in the individual-based model that are not consistent with the mathematical model; in the simulation model, offspring are created in proportion to the surviving offspring after mating has occurred, increasing the males' reproductive success above their specified reproductive investments. This allows males to escape the costs and increase their reproductive success beyond what is possible in the mathematical model. Random mating if non-preferred males are unable to be found also enables non-courter/parents to have increased reproductive success compared the the mathematical model predictions, which is why they are fixed in the monogamy case. If stricter costs on courting males are imposed, specifying $r=0.25$, meaning that $r_{courter}=2$ and $r_{non-courter}=8$, the scenarios without random mating do result in the maintenance of 2 or 3 morphs in most cases for both monogamy and polygyny (Fig. \@ref(fig:MatingTypesFig)), as expected.


```{r MatingTypesFig, fig.cap="The frequencies of morphs in the final generation of simulation runs a single gene underlying each trait, with parameters expected to yield low diversity (top row), high diversity (centre row), and high diversity with stronger costs on courter reproductive success (bottom row). The frequency of the courter/parent morph in the final generation is printed on top of each bar. In 'monogamy' cases, the males were constrained to only accept eggs from one female into their nests, whereas under 'polygyny' males could mate with multiple females. In both cases females mated once. Parameter settings lablled with 'nm' were runs where females did not mate if they could not find a suitable mate (i.e., if no courters were available). Within each parameter set, the numbers below the bars identify sets of simulation runs that were initiated with identical starting conditions."}



par(mfrow=c(3,4),oma=c(3,3,4,1),mar=c(3,1,4,1),xpd=TRUE)

# Low diversity
lowDiv<-get.morph.freqs(plot.pc.reps(pattern="lowDiversity.*summary.txt",path="single_locus",cols,make.plot=FALSE))
var_patts<-unique(gsub(".*lowDiversity_(.*)_\\d_summary.txt_\\d","\\1",rownames(lowDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)
varDat<-lapply(var_patts,function(pattern){
  dat<-lowDiv[grep(paste0(pattern,"_\\d"),rownames(lowDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main=pattern)
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})

# High diversity
highDiv<-get.morph.freqs(plot.pc.reps(pattern="highDiversity_.*summary.txt",path="single_locus",cols,make.plot=FALSE))
var_patts<-unique(gsub(".*highDiversity_(.*)_\\d_summary.txt_\\d","\\1",rownames(highDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)

varDat<-lapply(var_patts,function(pattern){
  dat<-highDiv[grep(paste0(pattern,"_\\d"),rownames(highDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main="")
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})

# Strict High Diversity
highDiv<-get.morph.freqs(plot.pc.reps(pattern="highDiversityStrict.*summary.txt",path="single_locus",cols,make.plot=FALSE))
var_patts<-unique(gsub(".*highDiversityStrict_(.*)_\\d_summary.txt_\\d","\\1",rownames(highDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)


varDat<-lapply(var_patts,function(pattern){
  dat<-highDiv[grep(paste0(pattern,"_\\d"),rownames(highDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main="")
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})

# outer legend
par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0,0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=15,xpd = TRUE,ncol=2,cex=1.25)


```

One striking result is that stochasticity appears to play a major role in which morphs co-exist with the courter-parent morph, even when initial starting conditions are identical (Fig. \@ref(fig:MatingTypesFig)). Across independent iterations of the model, variation in the initial starting frequencies of each morph (which centre around 0.25 for each morph, but with some variation) can explain some of the differences in outcomes. Variation in outcomes within replicate runs can be attributed to the many sources of random chance incorporated in the model: stochasticity associated with finite population sizes (i.e., genetic drift), the randomness of recombination, and the finite mate search process (in which females might not find a suitable mate, even if one exists in the population) all contribute to the within-replicate variation observed. Furthermore, these sources of variation differentiate the results of the simulation model from those of the mathematical model. 

### The role of polygenic inheritance of traits

We incorporated polygenic inheritance of the male traits in two ways, both of which reduced the impacts of stochasticity on the final morph frequencies. 
In the low diversity parameter space, the outcomes remained consistent with the single-locus model (and the expectations from the mathematical model): courter/parents dominated the populations, with complete fixation of the morph in the majority of runs, and any variation existing in the form of 1-2 individuals (courter/parent frequencies were $\ge$ 99% in all cases). 
These results were consistent across all genetic architectures, including QTLs distributed across 2, 4, and 8 chromosomes and QTLs clustered in supergenes spanning from 5% to 50% of a single chromosome. 
In the high diversity parameter space, polygenic loci distributed throughout the genome resulted in the maintenance of the courter/parent trait at a frequency of `r round(mean(final_freqs$CP[grep("qtls.*highDiversity",final_freqs$file)])*100,2)`% ($\pm$ `r round(sem(final_freqs$CP[grep("qtls.*highDiversity",final_freqs$file)])*100,2)`%), with the non-courter/parent being the dominant morph. 
These proportions are remarkably consistent across genetic architectures. Most simulations with supergenes resulted in very similar final morph frequencies (`r round(mean(final_freqs$CP[grep("supergene.*highDiversity",final_freqs$file)])*100,2)`% $\pm$ `r round(sem(final_freqs$CP[grep("supergene.*highDiversity",final_freqs$file)])*100,2)`%), and only occasionally resulted in either three morphs being retained or the non-courter/non-parent morph replacing the non-courter parent. 
The size of the supergene did not substantially impact the final morph frequencies, nor did the total number of QTLs underlying each locus. *Do an analysis of this?*


### Will GWAS detect the actual QTLs?

As expected, the detection of true QTLs in a GWAS was more likely when fewer QTLs underpin the trait, especially if those QTLs are aggregated on fewer chromosomes (Fig. \@ref(fig:GWASoutcomes)). Somewhat unexpectedly, the QTLs for parent traits were much more likely to be detected by GWAS than the courter QTLs. In the majority of simulations, the final population only contained parental males, but both courters and non-courters persisted in the population. The finding the parent QTLs are more likely to be detected is likely because that trait experienced stronger selection on it (i.e., being a non-parent was more costly than being a non-courter).

What about supergenes?

### Sexual conflict over reproduction

Should we have any mention of Fsts? Perhaps in the sexual conflict context


# Discussion

Summary of findings 
- genetic architecture matters
- maybe counter intuitively it reduces stochasticity, likely through genetic correlations and/or linkage disequilibrium?


# Acknowledgments

We acknowledge Ngāi Tūāhuriri, upon whose lands the analyses and writing
was conducted. This work was conducted in part while SPF was a
postdoctoral fellowship at the National Institute for Mathematical and
Biological Synthesis, sponsored by the National Science Foundation
through NSF Award \#DBI-1300426, with additional support from The
University of Tennessee, Knoxville. 
SPF was partially funded by the Marsden fund grant number UOC1904. 

# Data Accessibility

All code has been archived on Zenodo (DOI: ) and is maintained on github
(). Simulation results are archived on dryad ().

# Author Contributions

# References
