---
title: "Genetic architecthure of alternative reproductive tactics"
author:
- affilnum: 1
  corresponding: yes
  email: spflanagan.phd@gmail.com
  name: Sarah P. Flanagan
- affilnum: 2
  name: Suzanne Alonzo
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: no
    toc_depth: 2
  html_document: null
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document: null
bibliography: references.bib
csl: proceedings-of-the-royal-society-b.csl
capsize: normalsize
documentclass: article
editor_options:
  chunk_output_type: console
  markdown: 
    wrap: 72
fontsize: 11pt
abstract: |
header-includes: |
  \usepackage{lipsum} \usepackage{float} \floatplacement{figure}{H}
affiliation:
- affil: School of Biological Sciences, University of Canterbury, 4800 Private Bag,
    Christchurch 8140 New Zealand
  affilnum: 1
- affil: University of California Santa Cruz, Santa Cruz, California, USA
  affilnum: 2
preprint: no
spacing: singlespacing
---

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200,fig.height = 7,fig.width = 7)
knitr::opts_knit$set(root.dir='../results/') #change

```

```{r opts, echo = FALSE, warning=FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "../figs/"
)
```

```{r librarysetup, echo=FALSE,echo=FALSE,message=FALSE, warning=FALSE}
library(RColorBrewer)
library(scales)
library(sm)
library(kableExtra)
library(plotly)
library(dplyr)
library(vegan)
library(tidyr)
tmp<-sapply(list.files(path="../../spfTools/R/",full.names = TRUE),source)
source("../R/002_freq_functions.R")

source("../morph_predictions/check_freqs.R")
source("../morph_predictions/morph_gens_ns.R")

cols<-c(courter="#33a02c",parent="#1f78b4",cpref="#b2df8a",ppref="#a6cee3")
cols2<-c(CP='#d7191c',NCP='#fdae61',CNP='#abd9e9',NCNP='#2c7bb6')
```

```{r baseCols}
base_keep_cols<- c("Pop","PopSize","NumMal","NumFem","ParentThresh","ParentFreq","CourterThresh","CourterFreq","FreqNcNp","FreqCNp",
                   "FreqNcP","FreqCP","Courter2NonRS","Parent2NonSurvival")
poly_keep_cols<- c("Pop","PopSize","NumMal","NumFem","ParentThresh","ParentFreq","CourterThresh","CourterFreq","FreqNcNp","FreqCNp",
                    "FreqNcP","FreqCP","Polygyny")
divcols<-c("#e78ac3","#ffd92f")

```

# Introduction

Both genetic and phenotypic variation are required for selection to
occur, and as such the maintenance of variation in populations has long
been of interest to evolutionary biologists and conservation
geneticists. In general, evolution by natural selection is predicted to
decrease variation in populations, while only a few specific forms of
selection canÂ maintain variation in populations, such as overdominance,
sexually antagonistic selection
[@connallonEvolutionaryConsequencesSexSpecific2018;
@connallonResolutionSexualAntagonism2011], and frequency dependent
selection [@alonzoMateChoiceGames2001]. Yet, substantial phenotypic
variation is observed, with ubiquitous examples such as the prevalence
of sexual dimorphism. Across a wide range of taxa, discrete alternative
reproductive tactics are frequently maintained, in which individuals of
the same species and sex display different morphs to maximize
reproductive fitness in unique ways
[@oliveiraAlternativeReproductiveTactics2008]. These stable
polymorphisms are typically maintained in populations through negative
frequency-dependent selection (e.g., female morphs in damselflies
[@iserbytNegativeFrequencydependentSelection2013]; Gouldian finches
[@prykeFrequencydependentPhysiologicalTradeoffs2007]), the interaction
of frequency-dependent and density-dependent factors (e.g.,
side-blotched lizards [@alonzoMateChoiceGames2001]), and variable
selection such as partitioning of reproductive success over time (e.g.,
the feminized dwarf spider morph mates earlier, whereas the alternative
morph is able to mate with previously-mated females later in the season
[@hendrickxMasculinizingSupergeneUnderlies2022]). *Add a sentence to
link alternatives with genetic basis more explicitly*

The genetic basis of polymorphic traits is also critical to
understanding the maintenance of variation. Evolution can be constrained
by factors including pleiotropy, genetic correlations, and physical
linkage. As such, the reduction or removal of these constraints can
facilitate sequence evolution by relaxing positive selection on genes
[@harrisonSexualSelectionDrives2015; @dapperRelaxedSelectionRapid2020].
When polymorphic traits create conflict between sexes or morphs, these
genetic constraints can prevent the resolution of conflict and therefore
result in the maintenance of diversity
([@pennellContrastingEffectsIntralocus2016]). Often, these constraints
are overcome through mechanisms such as gene duplications (e.g.,
[@desmaraisEscapeAdaptiveConflict2008]), differential gene expression
[@mankPopulationGeneticsSexual2017], or even the evolution of asexuality
[@brownSocialDilemmasSupergenes2011]. Most work on such intralocus
conflict is related to sexual antagonism and sexual conflict, but
intralocus conflict can emerge among morphs of the same sex, as they
also share the majority of their genome
[@morrisIntralocusTacticalConflict2013]. Intriguingly, in ruffs, which
have three male alternative tactics, the evolution of the inversion
polymorphism encoding the male morphs appears to have resolved conflict
among the male morphs, while simultaneously exacerbating intralocus
sexual conflict [@giraldo-deckIntralocusConflictsAssociated2022].

Despite the fact that alternative reproductive tactics are by definition
complex, multivariate phenotypes, population genetic predictions based
on intralocus conflict tend to be based on one- or two-locus models. As
such, generating predictions for expected evolutionary outcomes -- and
expected patterns to identify in genome-wide datasets -- is not
necessarily straightforward. *Expand this a bit more to transition to
supergenes : touch on the fact that simple pop gen models using the same
genetic controls that are different between alternatives, while
multilocus models predict... and then supergenes is one hypothesized
solution to this evolutionary puzzle*

One genetic mechanism that is increasingly being considered as an
important mechanism in evolution are structural variants, such as
inversion polymorphisms that harbour so-called 'supergenes'. These
variants are regions of the genome in which some individuals carry an
inversion, in which genes with direct phenotypic effects exist. The
mismatched directionality reduces recombination between the different
variants, facilitating the maintenance of both genetic and phenotypic
variations. These supergenes have been shown to underpin stable
polymporhisms including multiple social morphs in fire ants
[@yanEvolutionSupergeneThat2020], colony sex ratio skew in ants
[\@lagunas-roblesLinkedSupergenesUnderlie2021], sex-specific migratory
behaviours [\@pearseSexdependentDominanceMaintains2019], sperm length
and swim speed [\@kimSexlinkedSupergeneControls2017;
\@kniefSexchromosomeInversionCauses2017], venom composition in
rattlesnakes [\@zancolliWhenOnePhenotype2019], and sexually dimorphic
alternative reproductive tactics
[\@thomasChromosomalPolymorphismLinked2008;
\@kupperSupergeneDeterminesHighly2016;
\@lamichhaneyStructuralGenomicChanges2016;
\@hendrickxMasculinizingSupergene2022]. The potential for structural
variants to underpin sexually dimorphic polymorphisms is intriguing, as
most research to date has focused on the role of differential gene
expression in underlying sexual dimorphism (e.g.,
[@harrisonSexualSelectionDrives2015]). For example, in canaries, sexual
dichromatism appears to be mediated by differential expression of a
single gene involved in carotenoid breakdown
[\@gazdaGeneticMechanismSexual2020]. Androgens are a likely mechanism
for modulating these gene expression differences, for example in
primates male-biased genes are more likely to be proximal to androgen
response elements [\@andersonRelationshipSexualDimorphism]. Unlike the
canaries, though, examples where supergenes apparently underpin sexually
dimorphic phenotypes tend to involve multiple morphs of one sex, usually
males. Nearly all examples include a female-mimic morph, including ruffs
[\@kupperSupergeneDeterminesHighly2016;
\@lamichhaneyStructuralGenomicChanges2016], dwarf spiders
[\@hendrickxMasculinizingSupergene2022], and white-throated sparrows
[\@thomasChromosomalPolymorphismLinked2008]. In the ruffs, a third male
morph exists, with an intermediate phenotype between the dominant or
'classic' male morph and the female-mimic
[\@kupperSupergeneDeterminesHighly2016;
\@lamichhaneyStructuralGenomicChanges2016]. In some cases, a
heterozygous genotype is lethal (e.g., ruffs
[\@kupperSupergeneDeterminesHighly2016;
\@lamichhaneyStructuralGenomicChanges2016]), but not in others (e.g.,
dwarf spiders [\@hendrickxMasculinizingSupergene2022]). In the
white-throated sparrow and the ruff, these structural variants are
successful in coding for these variable phenotypes because they include
androgen receptors [@maneySupergenesSteroids2022]-- the white-throated
sparrow supergene contains the estrogen receptor gene which directly
influences the aggressive aspects of the dominant male phenotype
[\@merrittSupergenelinkedEstrogenReceptor2020].

An open question is to what extent do supergenes facilitate the
maintenance of multiple morphs, versus whether the evolution of multiple
morphs might facilitate the evolution of supergenes. This question is
difficult to address empirically, and is likely to have consequences for
the maintenance of genetic variation at loci contributing to traits. For
example, if a supergene exists and a mutation or arises within it that
impacts the morph traits, the genetic variation within the supergene
itself could have eroded due to relaxation of selection and reduced
recombination. As such, only the mutation would be likely to show
significant genetic variation -- this could make a supergene with
functional effects difficult to detect empirically. This is especially
true in cases such as the white-throated sparrow, in which the gene with
a major effect is an estrogen receptor
[\@merrittSupergenelinkedEstrogenReceptor2020], which likely impacts the
expression of many genes. Furthermore, it is currently unknown to what
extent genetic architectures interact with the frequency- and
density-dependent mechanisms that maintain multiple morphs in
populations. *Expand: the questions/problems our theory aims to
answer/address are laid out a bit more explicitly and fully* *possibly
this is a good way to describe how folks have looked into supergenes,
and set up our genome-wide analyses*

To tackle these problems, we have developed a simulation model
incorporating explicit genetic architectures. We first identify regions
of the frequency-dependent parameter space that should result in the
maintenance of multiple morphs in a single-locus context using a
mathematical model, and subsequently compare the outcomes of multilocus
genetic architectures to those baselines. Our results suggest that
morphs encoded by many genes are more robust to stochastic elements that
impact the morph frequencies over time, and that the number of morphs
that are predicted to persist changes when multilocus genotypes are
involved. When the morphs have arrived at equilibrial frequencies,
selection on male traits is generally relaxed, and this corresponds with
a decreased ability to reliably detect causative variants using typical
genomic analysis approaches (i.e., genome-wide associations). *Possibly
less focus on our results and instead lay out our questions and analyses
a bit more*

# Methods

Our model considers a biological scenario in which males display to
attract females, and with obligate male parental care. In our model,
males possess two independent traits, courtship and parental care, both
of which are expressed only in males (but are inherited by females).
Courting males display within a territory, and females choose a male to
mate with based on these displays. Once a female chooses a male, the
survival of the clutch is dependent on whether the chosen male is also a
parenting male. We assume that all females in the population always
prefer males with the courting trait, but females do not know when
choosing whether males will be parents. Males who do not court can only
reproduce by sneaking, and males who display but do not provide any
parental care will not produce viable offspring. Thus, our model has
four morphs: courter/parents (CP); courter/non-parents (CN);
non-courters/parents (NP); or non-courters/non-parents (NN), although
the courter/non-parent morph will never produce offspring. Our model is
diploid, and assumes non-overlapping generations. While this model might
not directly apply to all examples of alternative reproductive tactics,
it is biologically reasonable and common even if not universally true.
We expect that this model can be generalized to species in which one sex
(but not the other) has two fixed traits that impact mating success and
the survival of their offspring.

Each generation has three distinct phases: (1) mate choice and
fertilization; (2) parental care; and (3) viability selection. In the
mate choice phase of the model, we assume females always prefer to mate
with males displaying the courtship trait. If a female decides to mate
with a male, he receives all of her eggs to care for. The male is not
guaranteed paternity of his mates' entire brood, however, because
non-courting males are able to fertilize a proportion of the eggs. The
chosen parent does receive first-male priority. The consequences of this
setup are that the only mechanism by which non-courters can produce
offspring is through these sneak fertilizations. In the parental care
phase of the model, males with the parental care trait have 100% nest
survival while males without the parental care trait have 0% nest
survival, independent of the male's courtship trait. Consequently,
non-parents will only be sustained in the population by sneaking
fertilizations in the parental males' nests. Finally, viability
selection is imposed on offspring as they mature to adulthood, such that
males carrying the courtship trait and/or the parental care trait have a
lower probability of survival than females and males without either
courtship or parenting. Males who are both courters and parents have a
multiplicative effect of viability selection.

We implemented our model first as a mathematical model and then as an
individual-based simulation model. The purpose of the mathematical model
was to generate baseline predictions for when multiple morphs are
maintained.

## Mathematical model

*Need to adjust the terms/explanations, and adjust the terminology to
avoid 'nests laid', which is confusing*

In the mathematical model, we model the proportion of the population
that belong to morphs at each stage of the lifecycle (Fig.
\@ref(fig:mathModel)). In the first stage, we consider the proportion of
the population's nests that will be allocated to each morph. The maximum
number of nests that could be laid is proportional to the number of
males in the population ($N_m$), if each male has the same likelihood of
attracting a female to deposit eggs with him. However, in our model,
males are not equally likely to be chosen by females, and therefore the
maximum number of nests is scaled by the probability of females mating
with each type of male morph. In the case of our model, the total number
of nests laid in the population becomes $$
\frac{N_f}{N_mf_{CP}+N_m f_{CNp}}
$$ because we have constrained females to only mate with courting males.

```{r mathModel, fig.cap="Diagram of the mathematical model describing the inheritance of alternative male reproductive morphs given courtship and parenting traits.", out.width='100%'}
knitr::include_graphics("frequencyDependence.png", dpi=150)
```

*make sure fig. 1 equations are complete and also in the main text*

Likewise, the proportion of nests each type of male receives (${p_n}_M$)
is determined by the number of morphs in the population ($N_mf_M$ for
the $M$ morph), the probability of a female choosing that morph
(${w_s}_M$), and the ratio of the number of nests to the number of
females ($n/N_f$).

Within a given nest, the offspring could either be fertilized by the
nest-holding male or non-courting males 'sneaking' fertilizations.
Therefore, each morph has a proportion of eggs fertilized within their
own nests (${p_{fn}}_M$) and a proportion of eggs fertilized in other
males' nests (${p_{fs}}_M$). These proportions were determined by the
sperm competition coefficient, $c$, which penalized non-courting males,
and the relative reproductive investment for that morph ($r_M$).
Analogously, the proportion of offspring in a nest surviving due to
parental care investment can be split into the proportion surviving in a
male's own nest (${p_{sn}}_M$) or in another male's nest (${p_{ss}}_M$),
which is simply the proportions from the fertilization step weighted by
the probability of the nest surviving if the morph is a parent ($w_n=1$)
or of the morph is a non-parent ($w_n=0$). These proportions
(${p_{sn}}_M$ and ${p_{ss}}_M$) are summed to determine the the overall
proportion of embryos in the population that survive the parenting phase
in each morph's nests (${p_s}_M$).

The final stage of the lifecycle is for offspring surviving in the nest
to undergo viability selection by multiplying the viability selection
coefficient (${w_v}_M$) by the surviving proportion (${p_s}_M$) and the
original frequency of the morph in the population. Finally, the
frequency of each morph in the next generation was calculated by
dividing the surviving proportion (${p_v}_M$ by the sum of all
proportions.

To evaluate the outcomes of this mathematical model, we iterated forward
in time for 100 generations from 1659 combinations of initial morph
frequencies and repeated those iterations across 5 values for the sperm
competition coefficient ($c$ = {0..1}) and 21 values for the relative
reproductive coefficient ($r_M$ = {0...2}). The iterations were
implemented in R v. 4.1.1 [@rcoreteam2021] using some features of tidyR
[@wickham2021].

We were interested in identifying the regions of parameter space that
resulted in the largest diversity of male morphs. To ameliorate the
difficulty of visualising and summarizing four frequencies, we leveraged
existing ecological mathematical tools to summarize diversity of a
system into a single metric. Specifically, we calculated the Shannon's
diversity index. We calculated this index in the 100th generation using
the R package vegan [@oksanen2015] and subsequently visualised the
diversity of the population using contour plots produced by plotly
[@sievert2020] in a shiny app that relies on R packages shiny
[@chang2021], shinydashboard [@chang2018], rsconnect [@mcpherson2021],
dplyr [@wickham2021], and rmarkdown [@allaire2021; @xie2020; @xie2018].
The interactive app is available at
<https://spflanagan.shinyapps.io/morph_predictions/>.

Note that this model does not incorporate recombination or inheritance
via females. Nonetheless, this model provides a baseline predictions for
parameter values likely to facilitate morph diversity within a
population. These limitations are addressed in the individual-based
simulation model.

## Individual-based simulation model

To evaluate the role of genetic architecture in the maintenance of
alternative reproductive tactics, we wrote an individual-based
simulation model in C++. The simulation model follows a lifecycle that
is largely the same as the mathematical model (Fig.
\@ref(fig:mathModel)), but rather than tracking proportions of the
population belonging to particular morphs, we tracked individuals within
the population (Fig. \@ref(Fig:simModel)). Each individual contained
diploid genotypes, which included both non-coding loci and quantitative
trait loci (QTLs). We included two types of QTLs: ones that determined
the individual's courting trait and ones that determined the
individual's parenting trait. The QTLs were each assigned a
randomly-drawn effect on the trait, and the QTLs additively determined
the trait value for each trait (i.e., the effects of all QTLs were
summed to determine the trait value). In this model, we assume no
environmental effects impacted the trait value. These trait values were
then compared to a threshold to dichotomize each trait. The threshold
for each trait was the mean trait value during the initialization of the
model (i.e., at generation 0).

```{r simModel, fig.cap="Overview of the individual-based simulation model."}
knitr::include_graphics("density-dependent-flowchart.jpg", dpi=150)
```

We implemented three different genetic architectures for these QTLs: a
single gene per trait, multiple QTLs randomly placed throughout
chromosomes, and supergenes. The single gene scenario is most analogous
to the mathematical model described above. When the traits have multiple
QTLs determining their trait values, each QTL is assigned a location on
a chromosome and the QTL's effect on the trait is associated with a
single nucleotide polymorphism (SNP). These SNPs undergo recombination
and mutation during mating. The supergene scenario forced the QTLs to
cluster into a small proportion of one of the chromosomes. In all cases,
we assume all QTLs are autosomal, and we do not model sex chromosomes at
all.

When we ran the model, we initialized four replicate populations with
the identical starting conditions to evaluate the impact of
stochasticity on the model. These identical conditions included
randomly-generated values for the allelic effects of QTLs, exact number
of individuals of each sex and morph, and thresholds for determining
morph traits (i.e., the values at which an individual is deemed a
courter vs a non-courter and a parent vs a non-parent). For each set of
four populations, we ran 10,000 burn-in generations followed by 2,000
'experimental' generations. Doing so ensured that the population reached
a steady state in which we could examine any stable polymorphisms or
cyclical fluctuations in morph frequencies.

Within each of those generations, the population underwent four main
phases: (1) mate choice and fertilization; (2) nest survival; (3)
viability selection on juveniles; (4) maturation to adulthood. The main
difference between the individual-based simulation model and the
mathematical model were the introduction of stochasticity during mate
choice and the introduction of recombination and mutation during
reproduction. During the mate choice phase, females searched through a
set number of males and chose the first attractive male (i.e., the first
courter male she encountered). If a female did not find an acceptable
mate, she either did not mate ('no mating' or 'nm' scenarios) or she
chose a male at random with from the population and allocated all of her
offspring to him, depending on the parameter settings. Females always
preferred courting males.

Once a female chose a male, fertilization of the eggs occurred, in which
the chosen male had first male priority (i.e., the chosen male allocated
all of his reproductive effort to his own nest) and up to 3 non-courting
males could 'sneak' fertilizations, with their contributions discounted
by a factor of $c$. After the males were assigned their paternity
shares, the nests survived if the chosen male was also a parent.
Subsequently, the total number of offspring in the population were
created based on the proportion of population reproductive success
achieved by each individual male and female (i.e., each generation, the
number of offspring created was equal to the carrying capacity).

When fertilization occurred, the parents' genotypes (both neutral
markers and QTLs) had the opportunity to recombine. Recombination
occurred independently on the chromosome inherited from the mother and
the chromosome inherited from the father. The number of recombination
events for a given chromosome was determined by drawing from a Poisson
distribution with the recombination rate as the mean, which we set to
0.2. For each recombination event, breakpoints were chosen by randomly
drawing from the total number of markers on a chromosome. These
breakpoints defined the segments of the chromosome that experienced
recombination, and each segment was randomly chosen to be assigned the
genotypes and effects from the parents' maternal or paternal chromosome.
When the genetic architecture was that of a supergene, if the
breakpoints fell within the region containing the aggregated QTLs for
the traits, the offspring was considered not viable.

After recombination occurred, if the progeny survived, the genotypes
underwent mutation, which occurred at a rate of $mu$. For a given
individual, the number of mutations introduced was determined as
$2\mu*n_SNPs*n_chrom$. The location of each mutation was randomly
selected and the allele at that location was randomly chosen from the
set of existing alleles. The mutation also impacted the phenotypic
effects of QTLs by altering the original effect by a value drawn from a
normal distribution with a mean of 0 and a standard deviation of
$sigma_\mu$. Finally, offspring were also randomly assigned a sex.

The next stage in the lifecycle was viability selection, which impacted
only male offspring. For each male offspring, a survival probability was
calculated as $e^{\frac{-morph2}{2\omega_v}}$. Viability selection acted
multiplicatively on courter-parents (their survival probability was
$e^{\frac{-morph2}{2\omega_v}}*e^{\frac{-morph2}{2\omega_v}}$). A random
value between 0 and 1 was drawn, and if the number was larger than the
survival probability for an individual, that individual did not survive
to adulthood. If, at this point, the number of surviving offspring was
larger than the carrying capacity, offspring were randomly selected to
become adults. When offspring matured to adulthood, the previous
generation's adults were replaced, and the lifecycle began again.

Each set of genetic architectures and population parameters was also was
run with four types of mating systems, all of which required females to
lay their eggs with only a single male: males accept eggs from a single
female, and females deposit their eggs with a randomly-selected unmated
male in the population should they fail to find a preferred mate
('monogamy'); males accept eggs from multiple females, and females
deposit their eggs with a randomly-selected unmated male in the
population should they fail to find a preferred mate ('polygyny'); males
accept eggs from a single female, and females who fail to find a
preferred mate do not produce offspring ('monogamy no mating'); males
accept eggs from multiple females, and females who fail to find a
preferred mate do not produce offspring ('polygyny no mating').

### Model analysis

The outcomes of simulation model were first assessed by visualising the
proportions of each morph that persisted to the final generation of the
model. The first priority was to compare the outcomes of the simulation
model with a single locus for each trait to the results from the
mathematical model. Next, the four mating systems were compared in the
single-locus model. From this comparison, we were then able to narrow
the parameter space for the comparisons of the two different multilocus
genetic architectures. The many different formats for these two genetic
architectures were then compared in terms of the morph composition in
the final generation, and the amount of variation among replicate runs.

To identify how selection acts on the two male traits, selection
gradients on the courter trait and parent trait (both the binary
variable of courter/non-courter and parent/non-parent) were estimated as
the slope of a linear regression of lifetime reproductive success on
each trait values. The patterns of selection were compared across
genetic architectures and parameter settings.

### Analysing multilocus genotypes

To provide a link between our theoretical models and the realities of
empirical data, we analyzed the genetic diversity of the populations at
the final generation of the individual based simulations. First we used
vcfR to estimate observed heterozygosity for all markers in the genome
and separately for the QTLs. Allele frequencies between morphs were also
compared using the Gst estimator in vcfR
[@knausVcfrPackageManipulate2017]. Additionally, an association between
male phenotypes and genetic markers was tested for using GWASpoly
[@rosyaraSoftwareGenomeWideAssociation2016]. For both the Gst estimates
and the GWAS p-values, any major peaks identified were scanned to
identify if they were within a 50 basepair window of one of the true
QTLs. Finally, we used vcftools [@danecekVariantCallFormat2011] to test
for signals of balancing selection at the QTL loci compared to non-QTL
loci.

To investigate whether linkage disequilibrium persisted over time within
our various genetic architectures, we used vcftools
[\@danecekVariantCallFormat2011] to estimate linkage disequilibrium
between all loci. We then estimated summary statistics for each locus
and compared linkage disequilibrium between QTLs and compared it to
linkage disequilibrium among non-QTLs. Because we had far fewer QTLs
than non-QTL loci in all of the iterations, we also used re-sampling of
the non-QTLs to estimate the average linkage disequilibrium, using 999
resampled estimates with the actual number of QTLs in each replicate.

# Results

## Frequency-dependent population dynamics model

The initial frequency of each morph impacted whether multiple morphs
were maintained. In no cases was the courter-nonparent morph maintained.
This result makes intuitive sense, because if a female chooses to nest
with a courter-nonparent, the nest will always die (given our parameter
settings).

The parameters controlling sperm competition ($c$), relative
reproductive allocation ($r$), and the number of sneakers allowed also
impacted the diversity of the population after both 100 generations and
1000 generations. Some initial starting conditions took more than 100
generations to reach a consistent (i.e., equilibrial) state, but the
contributions of these three parameters remained qualitatively
consistent across 100 and 1000 generations. Reproductive allocations
corresponding to a range of approximately 0.5 to 1 (meaning that
courters invest half as much to just as much as non-courters in
producing sperm) resulted in the maximum amount of diversity, as long as
sperm competition ($c$) was above 0.2 (meaning that, if a non-courting
male produced 8 sperm and courting males produce 4, the non-courting
male would fertilize `r (8*0.2)/(4+(2*0.2*8))` or more of the
offspring). This importance of the sperm competition coefficient was
influenced by the number of sneakers allowed, with larger contributions
from an individual non-courting male (e.g., larger $c$) being required
with fewer sneakers (Fig. \@ref(fig:mathParams)).

```{r mathParams, fig.cap="Contour plots showing the diversity of the populations given equal initial starting frequencies of all four morphs, the relative reproductive investments (x-axis), sperm competition coefficients (y-axis), and number of sneakers allowed per nest (1 through 5, plots going from top to bottom).", fig.width=4.5, fig.height=7, warning=FALSE,message=FALSE}

results<-readRDS("../morph_predictions/morph_results_10000_equalStart.RDS")
results$diversity<-vegan::diversity(round(results[,c("CP","CN","NP","NN")],4))

contours<-by(results, results$num_sneak, function(sub_calcs){
  
  
  data_wide<-tidyr::spread(
    sub_calcs[,c("r","c","diversity")],
    r,
    diversity
  )
  rownames(data_wide)<-data_wide[,1]
  data_wide<-data_wide[,-1]
  
  fig <- plot_ly(
    x = as.numeric(colnames(data_wide)), 
    y = as.numeric(rownames(data_wide)), 
    z = as.matrix(data_wide), 
    type = "contour"
  )
  # add axis labels
  x<-list(title="r")
  y<-list(title="c")
  fig <- fig %>% layout(xaxis=x,yaxis=y)
  # add label to contour names
  fig <- fig %>% colorbar(title = "Diversity")
  # plot
  return(fig)
})


multi_fig <- subplot(contours, nrows = 5, shareY = TRUE, shareX = TRUE)

multi_fig %>% layout(autosize = F, width = 450, height = 650)


```

```{r}
divRes<-results[results$diversity>0,]
ndiv<-nrow(divRes)

pcourt<-nrow(divRes[divRes$CP>=0.5,])/ndiv
pnp<-nrow(divRes[divRes$NP>=0.5,])/ndiv
pnn<-nrow(divRes[divRes$NN>=0.5,])/ndiv


```

The regions of the parameter space the have non-zero diversity values
correspond to scenarios where three of the four morphs were retained
(courter-parents, non-courter-parents, and non-courter-non-parents). The
dominant morph differed with various parameter combinations. Assuming
equal starting frequencies, after 10000 generations, `r pcourt*100`% of
parameter combinations that retained polymorphism resulted in the
courter-parent morph comprising $\le$ 50% of the population. The
remaining `r (1-pcourt)*100`% of parameter combinations resulting in
polymorphism had more equal representation of all three morphs, with the
average proportion of the population being
`r mean(divRes[divRes$CP<=0.5,"CP"])*100` % ($pm$
`r sem(divRes[divRes$CP<=0.5,"CP"])*100` %) courter-parents,
`r mean(divRes[divRes$CP<=0.5,"NP"])*100` % ($pm$
`r sem(divRes[divRes$CP<=0.5,"NP"])*100` %) non-courter-parents, and
`r mean(divRes[divRes$CP<=0.5,"NN"])*100` % ($pm$
`r sem(divRes[divRes$CP<=0.5,"NN"])*100` %) non-courter-non-parents.

## Individual-based simulation model

The results from the mathematical model were used to select parameter
settings for the individual-based simulation model, with one set
expected to produce low-diversity populations (i.e., those where only
one morph were retained) and another set expected to produce
high-diversity populations (i.e., where 2 or more morphs were retained).
The chosen parameter settings are shown in Table
\@ref(tab:SimParamsTable).

```{r SimParamsTable}
SimParams<-data.frame(
  parameter = c("r (in shiny app)",
                "r_CP",
                "r_CN",
                "r_NP",
                "r_NN",
                "c",
                "number of sneakers"),
  LowDiv = c( 2, 
             8,
             8,
             4,
             4,
             0.5,
             2),
   HighDiv = c( 0.75, 
             6,
             6,
             8,
             8,
             0.7,
             2)
)

kable(SimParams, "latex",booktab=TRUE,
      col.names = c("Parameter", "Low Diversity Value","High Diversity Value"),
      caption="Table of parameters selected to create a high-diversity population. All morphs started at equal frequencies.")  %>% 
  kableExtra::kable_styling(latex_options = "HOLD_position")

```

### Impacts of mating system on the model

As expected, in the low diversity parameter space, the majority of
simulations resulted in fixation or near-fixation of the courter-parent
morph (Fig. \@ref(fig:MatingTypesFig)). A more surprising result was the
inconsistency of results from the mating types within the the high
diversity parameter space. Polygyny resulted in the fixation of the
courter/parent morph regardless of whether females mated at random if a
suitable mate was found, and monogamy resulted in the fixation of the
non-courter/parent morph in every case. Only monogamy with no mating
allowed resulted in polymorphism, with the courter/parent morph
coexisting with either the non-courter/parent or the
non-courter/non-parent morph (Fig. \@ref(fig:MatingTypesFig)). These
results can be explained by the nature of density-dependent survival in
the individual-based model that are not consistent with the mathematical
model; in the simulation model, offspring are created in proportion to
the surviving offspring after mating has occurred, increasing the males'
reproductive success above their specified reproductive investments.
This allows males to escape the costs and increase their reproductive
success beyond what is possible in the mathematical model. Random mating
if non-preferred males are unable to be found also enables
non-courter/parents to have increased reproductive success compared the
the mathematical model predictions, which is why they are fixed in the
monogamy case. If stricter costs on courting males are imposed,
specifying $r=0.25$, meaning that $r_{courter}=2$ and
$r_{non-courter}=8$, the scenarios without random mating do result in
the maintenance of 2 or 3 morphs in most cases for both monogamy and
polygyny (Fig. \@ref(fig:MatingTypesFig)), as expected.

```{r MatingTypesFig, fig.cap="The frequencies of morphs in the final generation of simulation runs a single gene underlying each trait, with parameters expected to yield low diversity (top row), high diversity (centre row), and high diversity with stronger costs on courter reproductive success (bottom row). The frequency of the courter/parent morph in the final generation is printed on top of each bar. In 'monogamy' cases, the males were constrained to only accept eggs from one female into their nests, whereas under 'polygyny' males could mate with multiple females. In both cases females mated once. Parameter settings lablled with 'nm' were runs where females did not mate if they could not find a suitable mate (i.e., if no courters were available). Within each parameter set, the numbers below the bars identify sets of simulation runs that were initiated with identical starting conditions."}



par(mfrow=c(3,4),oma=c(3,3,4,1),mar=c(3,1,4,1),xpd=TRUE)

# Low diversity
lowDiv<-get.morph.freqs(plot.pc.reps(pattern="lowDiversity.*summary.txt",path="single_locus",cols,make.plot=FALSE))
var_patts<-unique(gsub(".*lowDiversity_(.*)_\\d_summary.txt_\\d","\\1",rownames(lowDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)
varDat<-lapply(var_patts,function(pattern){
  dat<-lowDiv[grep(paste0(pattern,"_\\d"),rownames(lowDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main=pattern)
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})

# High diversity
highDiv<-get.morph.freqs(plot.pc.reps(pattern="highDiversity_.*summary.txt",path="single_locus",cols,make.plot=FALSE))
var_patts<-unique(gsub(".*highDiversity_(.*)_\\d_summary.txt_\\d","\\1",rownames(highDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)

varDat<-lapply(var_patts,function(pattern){
  dat<-highDiv[grep(paste0(pattern,"_\\d"),rownames(highDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main="")
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})

# Strict High Diversity
highDiv<-get.morph.freqs(plot.pc.reps(pattern="highDiversityStrict.*summary.txt",path="single_locus",cols,make.plot=FALSE))
var_patts<-unique(gsub(".*highDiversityStrict_(.*)_\\d_summary.txt_\\d","\\1",rownames(highDiv)))
var_patts<-grep("v0",grep("RM", var_patts, invert=TRUE, value=TRUE), invert=TRUE, value=TRUE)


varDat<-lapply(var_patts,function(pattern){
  dat<-highDiv[grep(paste0(pattern,"_\\d"),rownames(highDiv)),]
  bp<-barplot(t(as.matrix(dat[,c("FreqNcNp", "FreqCNp", "FreqNcP", "FreqCP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_summary.txt.*$","\\1",rownames(dat)),
        las=2,
        border=NA,
        main="")
  text(x=bp,y=0.5,srt=90,labels = dat$FreqCP)

  return(dat)
})

# outer legend
par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0,0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=15,xpd = TRUE,ncol=2,cex=1.25)


```

One striking result is that stochasticity appears to play a major role
in which morphs co-exist with the courter-parent morph, even when
initial starting conditions are identical (Fig.
\@ref(fig:MatingTypesFig)). Across independent iterations of the model,
variation in the initial starting frequencies of each morph (which
centre around 0.25 for each morph, but with some variation) can explain
some of the differences in outcomes. Variation in outcomes within
replicate runs can be attributed to the many sources of random chance
incorporated in the model: stochasticity associated with finite
population sizes (i.e., genetic drift), the randomness of recombination,
and the finite mate search process (in which females might not find a
suitable mate, even if one exists in the population) all contribute to
the within-replicate variation observed. Furthermore, these sources of
variation differentiate the results of the simulation model from those
of the mathematical model.

### The role of polygenic inheritance of traits

```{r}
final_freqs<-read.csv("../results/morph_freqs_summary.csv")
final_freqs$rep<-gsub("^.*Diversity_(.*)_pheno.csv","\\1",final_freqs$file)

```

We incorporated polygenic inheritance of the male traits in two ways,
both of which reduced the impacts of stochasticity on the final morph
frequencies. In the low diversity parameter space, the outcomes remained
consistent with the single-locus model (and the expectations from the
mathematical model): courter/parents dominated the populations, with
complete fixation of the morph in the majority of runs, and any
variation existing in the form of 1-2 individuals (courter/parent
frequencies were $\ge$ 99% in all cases). These results were consistent
across all genetic architectures, including QTLs distributed across 2,
4, and 8 chromosomes and QTLs clustered in supergenes spanning from 5%
to 50% of a single chromosome. In the high diversity parameter space,
polygenic loci distributed throughout the genome resulted in the
maintenance of the courter/parent trait at a frequency of
`r round(mean(final_freqs$CP[grep("qtls.*highDiversity",final_freqs$file)])*100,2)`%
($\pm$
`r round(sem(final_freqs$CP[grep("qtls.*highDiversity",final_freqs$file)])*100,2)`%),
with the non-courter/parent being the dominant morph. These proportions
are remarkably consistent across genetic architectures. Most simulations
with supergenes resulted in very similar final morph frequencies
(`r round(mean(final_freqs$CP[grep("supergene.*highDiversity",final_freqs$file)])*100,2)`%
$\pm$
`r round(sem(final_freqs$CP[grep("supergene.*highDiversity",final_freqs$file)])*100,2)`%),
and only occasionally resulted in either three morphs being retained or
the non-courter/non-parent morph replacing the non-courter parent. The
size of the supergene did not substantially impact the final morph
frequencies, nor did the total number of QTLs underlying each locus.

```{r ArchitecturesProps, fig.cap=""}

plot_dat<-final_freqs[grep("highDiversity.*monogamy_nm.*q8", final_freqs$file),]
plot_dat$params<-gsub("../fixedART-results/(\\w+).*(_c\\d+).*$","\\1",plot_dat$file)
plot_dat$params[grep("supergene",plot_dat$params)]<-paste0(plot_dat$params[grep("supergene",plot_dat$params)],
                                                           "_", 
                                                           gsub("../fixedART-results/(\\w+).*(prop0\\.\\d).*(_c\\d+).*$","\\2",
                                                                plot_dat$file[grep("supergene",plot_dat$params)]))
plot_dat$params<-paste0(plot_dat$params, 
                        gsub("../fixedART-results/(\\w+).*(_c\\d+).*$","\\2",plot_dat$file))

par(mfrow=c(4,3),oma=c(3,5,4,1),mar=c(3,1,1,1),xpd=TRUE)
varDat<-by(plot_dat,plot_dat$params,function(dat){

  
  bp<-barplot(t(as.matrix(dat[,c("NON", "C", "P", "CP")])),
        col=cols2[c("NCNP","CNP","NCP","CP")],
        names.arg =gsub("^.*(\\d)_pop.*$","\\1",dat[,"rep"]),
        las=2,
        border=NA,
      #  main=paste(gsub("^.*c(\\d+).*$","\\1",dat[1,"params"]), "chroms"),
        beside = FALSE
       )
  #text(x=bp,y=0.5,srt=90,labels = dat["CP"])
  if(length(grep("qtl",dat[,"params"]))>0){
    mtext(paste(gsub("^.*c(\\d+).*$","\\1",dat[1,"params"]), "chroms"),3)
    if(length(grep("c2",dat[,"params"]))>0){
      mtext("QTLs",2, line=3)
    }
  } else {
    if(length(grep("c2",dat[,"params"]))>0){
      mtext(paste0("Supergene\nProp ",gsub("^.*prop(\\d\\.\\d).*$","\\1",dat[1,"params"])),2,line=3)
    }
  }
  return(dat)
})

par(fig=c(0, 1, 0, 1), oma=c(0, 0, 0, 0), mar=c(0, 0, 0, 0), new=TRUE)
plot(0,0, type='n', bty='n', xaxt='n', yaxt='n')
legend("top",bty='n',legend = c("Courter/Parent","Courter/Non-parent","Non-courter/Parent","Non-courter/Non-parent"),
       col=cols2[c("CP","CNP","NCP","NCNP")],pch=15,xpd = TRUE,ncol=2)
```

```{r seldat}
seldat<-readRDS("selection_gradients.RDS")
seldat<-seldat[grep("monogamy_nm",names(seldat))]
```

Our initial expectation was that supergenes would follow a pattern more
similar to the single locus results, because empirical examples have
demonstrated Mendelian inheritance patterns approximating single-locus
models (e.g., [@hendrickxMasculinizingSupergeneUnderlies2022,
@kupperSupergeneDeterminesHighly2016,,@lamichhaneyStructuralGenomicChanges2016]).
To better understand the reasons for the similarity between the the
genome-wide QTLs simulations versus the supergene simulations, we
visualised the QTL haplotypes in supergenes (see Fig. \@ref(fig:example)
for a representative image). These visualisations suggest that because
of the additive, polygenic nature of our traits, individuals have
multiple pathways to creating the morphs in terms of genotypes at each
QTL, even when QTLs are segregating together in a supergene.

### Selection on courter and parent traits

Selection consistently favoured parent traits, whereas courter traits
could experience negative selection in high diversity parameter
scenarios (Fig. \@ref(fig:selectionOverTime)). Selection was strong in
the first generation, but by generation 12000 selection was
substantially relaxed on one or the other trait, in most cases. A number
of the low diversity replkicates resulted in highly correlated selection
on courters and parents, with a number of points falling on the positive
1:1 line (Fig. \@ref(fig:selectionOverTime)).

```{r selectionGradients, fig.height=4.5, fig.width=7, fig.cap="Selection gradients on parent and courter traits differ across the scenarios of our simulation model. In the first generation, selection always favours the parent trait, but selection gradients on the courter trait can be negative under parameter settings that favoured high diversity. By the end of the simulations (generation 12000), selection has been relaxed, so that selection on the parent trait is often near zero, as is selection on the courter trait. In some cases, the selection gradients on courter and parent triats are correlated at this final generation, especially in low diversity scenarios with QTLs as the genetic architecture. Solid grey lines show the x- and y-axes, with dotted grey lines showing the one-to-one correlation lines."}
par(mfrow=c(1,2))
plot(-0.25:0.25, -0.25:0.25, 
     type='n',
     xlim=c(-.25,0.25),
     ylim=c(-0.01,0.25),
     xlab="Initial generation",
     ylab="Selection gradient on parent trait",
     main="",
     bty='n', 
     axes=FALSE)
abline(v=0, col="grey")
abline(a=0,b=1,col="grey",lwd=2,lty=2)
abline(a=0,b=-1,col="grey",lwd=2,lty=2)
mtext("Selection gradient on courter trait", side=1,outer=TRUE, line=-3)
axis(1, col="grey",pos=0,col.axis="grey",at = seq(-0.25,0.25,by=0.25))
axis(2, col="grey", pos=-0.25,col.axis="grey",lwd=0,at = seq(0,0.25,by=0.125))

tmp<-lapply(seldat, function(dat){
  dat<-dat[dat[,"gen"]==0,]
  

  points(dat[,"court_cat_b1"],
         dat[,"parent_cat_b1"],
         pch=dat[,"supergene"]+15,
         col=scales::alpha(divcols[dat[,"highDiv"]+1],0.5),
           cex=1.5
  )
  arrows(
    x0 = dat[,"court_cat_b1"] - dat[,"court_cat_b1_se"],
    x1 = dat[,"court_cat_b1"] + dat[,"court_cat_b1_se"],
    y0 = dat[,"parent_cat_b1"],
    y1 = dat[,"parent_cat_b1"],
    code = 0,
    col=scales::alpha(divcols[dat[,"highDiv"]+1],0.5)
      )
    arrows(
    x0 = dat[,"court_cat_b1"] ,
    x1 = dat[,"court_cat_b1"] ,
    y0 = dat[,"parent_cat_b1"] - dat[,"parent_cat_b1_se"],
    y1 = dat[,"parent_cat_b1"] + dat[,"parent_cat_b1_se"],
    code = 0,
    col=scales::alpha(divcols[dat[,"highDiv"]+1],0.5)
      )

})

plot(-0.25:0.25, -0.25:0.25, 
     type='n',
     xlim=c(-.25,0.25),
     ylim=c(-0.01,0.25),
     xlab="Final Generation",
     ylab="",
     main="",
     bty='n', 
     axes=FALSE)
abline(v=0, col="grey")
abline(a=0,b=1,col="grey",lwd=2,lty=2)
abline(a=0,b=-1,col="grey",lwd=2,lty=2)
axis(1, col="grey",pos=0,col.axis="grey",at = seq(-0.25,0.25,by=0.25))
axis(2, col="grey", pos=-0.25,col.axis="grey",lwd=0,at = seq(0,0.25,by=0.125))

tmp<-lapply(seldat, function(dat){
  dat<-dat[dat[,"gen"]==12000,]
  if(nrow(dat)>0){
    
    
    points(dat[,"court_cat_b1"],
           dat[,"parent_cat_b1"],
           pch=dat[,"supergene"]+15,
           col=scales::alpha(divcols[dat[,"highDiv"]+1],0.5),
           cex=1.5
    )
    arrows(
      x0 = dat[,"court_cat_b1"] - dat[,"court_cat_b1_se"],
      x1 = dat[,"court_cat_b1"] + dat[,"court_cat_b1_se"],
      y0 = dat[,"parent_cat_b1"],
      y1 = dat[,"parent_cat_b1"],
      code = 0,
      col=scales::alpha(divcols[dat[,"highDiv"]+1],0.5)
        )
      arrows(
      x0 = dat[,"court_cat_b1"] ,
      x1 = dat[,"court_cat_b1"] ,
      y0 = dat[,"parent_cat_b1"] - dat[,"parent_cat_b1_se"],
      y1 = dat[,"parent_cat_b1"] + dat[,"parent_cat_b1_se"],
      code = 0,
      col=scales::alpha(divcols[dat[,"highDiv"]+1],0.5)
        )
  }
})
outer_legend("top",
       bty='n',
       pch=c(15, 16, 0,0),
       lty=c(0, 0, 1,1),
       lwd=c(2,2,1,1),
       col=c("grey", "grey",divcols[2],divcols[1]),
       legend=c("Supergenes", "QTLs", "High Diversity", "Low Diversity"),
       ncol=4)
```

### Genome-wide genetic diversity

In the final generation of the simulations, the average genome-wide
heterozygosity was not substantially different in simulations with
genome-wide QTLs versus QTLs in supergenes (t = -0.31, df = 326.75, p =
-0.31; genome-wide mean = 0.15, supergene mean = 0.16; Fig.
\@ref(fig:genomewideDiv) A, B). In most runs, the heterozygosity at QTLs
was did not significantly differ from heterozygosity at neutral markers
(mean p-value from t-tests is 0.5 Â± 0.01). Similarly, Tajima's D did not
deviate substantially between genetic architectures and did not relate
to final morph frequencies (Fig. \@ref(fig:genomewideDiv) C, D). Gst
calculated between male morphs was also not predicted by the frequency
of Courter/Parent morphs (Fig. \@ref(fig:genomewideDiv) E). Significant
peaks in genome-wide Gst estimates were identified, and we evaluated
whether QTLs were located near to those Gst peaks (i.e., we identified
if Gst outliers were associated with true QTLs). From this, we then
calculated the proportion of true QTLs that were located in these
significant Gst peaks, and found that on average 0.71 (Â± 0), and this
proportion was not related to the Courter/Parent frequency (Fig.
\@ref(fig:genomewideDiv) F).

```{r}
welch_correctDF<-function(var1,var2,n1,n2){
  num<-((var1/n1)+(var2/n2))^2
  den<-(((var1/n1)^2)/(n1-1))+(((var2/n2)^2)/(n2-1))
  df<-num/den
  return(df)
}

```

```{r readGenomewide}

qtl_outliers<-read.csv("qtl_outliers.csv")
qtl_outliers$Architecture<-"qtls"
supergene_outliers<-read.csv("supergene_outliers.csv")
supergene_outliers$Architecture<-"supergene"
popgen_data<-dplyr::bind_rows(qtl_outliers,supergene_outliers)

sediff<-sqrt((popgen_data$gHtVar/(popgen_data$nChrom*1000)) + 
               (popgen_data$qHtVar/(popgen_data$nQTLS)))
HetT<-(popgen_data$gHtMean - popgen_data$qHtMean)/sediff
HetP<-pt(HetT,
         df = welch_correctDF(popgen_data$gHtVar,
                              popgen_data$qHtVar,
                              (popgen_data$nChrom*1000),
                              popgen_data$nQTLS), 
         lower.tail = (HetT >=0))

popgen_data$HetP<-HetP
popgen_data$Diversity<-gsub("^.*/(\\w+)Diversity.*$","\\1",popgen_data$sim)



final_freqs<-read.csv("../results/morph_freqs_summary.csv")
final_freqs$rep  <-gsub("../fixedART-results/.*\\/(.*\\w+_monogamy_nm_.*pop\\d).*","\\1",final_freqs$file)

td_summary<-read.csv("../results/tajimaD_summary.csv", row.names = 1)
td_summary<-td_summary[grep("monogamy_nm",rownames(td_summary)),]
td_summary$rep<-gsub("../fixedART-results/.*\\/(.*\\w+_monogamy_nm_.*pop)_(\\d).*","\\1\\2",rownames(td_summary))

td_freqs<-merge(td_summary, final_freqs,by="rep")

td_freqs$architecture<-gsub("../fixedART-results/(.*)\\/\\/(\\w+)Diversity_.*_monogamy_nm_.*pop_(\\d).*","\\1",rownames(td_summary))
td_freqs$diversity<-gsub("../fixedART-results/(.*)\\/\\/(\\w+)Diversity_.*_monogamy_nm_.*pop_(\\d).*","\\2",rownames(td_summary))

```

**Note to self: something seems off about this graph, compared to what
it was previously**

```{r genomewideDiv, fig.cap="Genome-wide genetic diversity in the simulation results. Genome-wide observed heterozygosity (A), Tajimaâs D (C), and male-male Fst (E) were not influenced by the frequency of the morphs in the final generation, nor by the genetic architecture. The final morph frequencies also did not impact the significance of comparisons of QTLs versus neutral loci for heterozygosity (B) or Tajimaâs D (D), or the proportion of QTLs found in the Fst peaks (F). Note that heterozygosity and Gst results are only shown for the runs with high diversity parameter sets"}

par(mfrow=c(3,2))

plot(popgen_data$propCP,
     popgen_data$gHtMean,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(as.factor(popgen_data[,"Diversity"]))+1],
       0.5),
     cex=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Average Obs. Het.",
     bty='l')
text(0.3,0.25,"A",cex=3)

plot(popgen_data$propCP,
     popgen_data$HetP,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(as.factor(popgen_data[,"Diversity"]))+1],
       0.5),
     cex=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Obs. Het P-values",
     bty='l')
text(0.3,0.85,"B",cex=3)

# Tajima's D
plot(td_freqs$CP,
     td_freqs$avgTD,
     pch=as.numeric(as.factor(td_freqs$architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(as.factor(td_freqs$diversity))],
       0.5),
     cex=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Average Tajima's D",
     bty='l')
text(0.3,0.25,"C",cex=3)

plot(td_freqs$CP,
     td_freqs$p,
     pch=as.numeric(as.factor(td_freqs$architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(as.factor(td_freqs$diversity))],
       0.5),
     cex=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Tajima's D P-values",
     bty='l')
text(0.3,0.25,"A",cex=3)


# Fst
plot(popgen_data$propCP,
     popgen_data$gGpMeanMM,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(as.factor(popgen_data[,"Diversity"]))+1],
       0.5),
     cex=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Mean male-male Gst",
     bty='l')
text(0.02,0.8,"E",cex=3)

plot(popgen_data$propCP,
     popgen_data$propQTLsInPeaksMM,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(as.factor(popgen_data[,"Diversity"]))+1],
       0.5),
     cex=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Prop. QTL in Gst peak",
     bty='l')
text(0.3,0.8,"F",cex=3)

outer_legend("top",
       bty='n',
       pch=c(15, 16, 0,0),
       lty=c(0, 0, 1,1),
       lwd=c(2,2,1,1),
       col=c("grey", "grey",divcols[2],divcols[1]),
       legend=c("Supergenes", "QTLs", "High Diversity", "Low Diversity"),
       ncol=4)
```

### Will GWAS detect the actual QTLs?

As expected, the detection of true QTLs in a GWAS was more likely when
fewer QTLs underpin the trait, especially if those QTLs are aggregated
on fewer chromosomes (Fig. \@ref(fig:GWASoutcomes)). Somewhat
surprisingly, the QTLs for parent traits were more reliably detected by
GWAS than the courter QTLs. In the majority of simulations, the final
population only contained parental males, but both courters and
non-courters persisted in the population (Fig.
\@ref(fig:MatingTypesFig)). The finding the parent QTLs are more likely
to be detected is likely because that trait experienced stronger
selection on it (i.e., being a non-parent was more costly than being a
non-courter).

We expected the supergene scenarios to have a higher detection rate, but
this was not the case. The primary difference was that the variance in
false positive rates was increased with supergenes, but the true
positive rates were unaffected by the localization of QTLs.

```{r getGWASinfo}
gwas_output<-read.csv("GWAS_summary.csv", header = FALSE)
colnames(gwas_output)<-c("Marker",
                          "Chrom",
                          "Position",
                          "Model",
                          "R2",
                          "pval",
                          "NearQTL",
                          "Traits",
                          "File")

gwas_summary<-aggregate(NearQTL~Traits + File,sum,data=gwas_output)

gwas_summary$numQTLonChrom<-as.numeric(gsub("^.*q(\\d+)_c(\\d+)_\\d.*vcf","\\1",gwas_summary$File))
gwas_summary$numChrom<-as.numeric(gsub("^.*q(\\d+)_c(\\d+)_\\d.*vcf","\\2",gwas_summary$File))
gwas_summary$numQTLtotal<-gwas_summary$numQTLonChrom*gwas_summary$numChrom
gwas_summary$Architecture<-gsub("^.*results\\/(.*)\\/\\w+.*.vcf","\\1",gwas_summary$File)
gwas_summary$Architecture<-gsub("\\/","",gwas_summary$Architecture) # remove any extra / 
gwas_summary$CQ<-as.factor(paste(gwas_summary$numChrom,gwas_summary$numQTLtotal))

# calculate the proportion of QTLs identified
gwas_summary$PropQTL<-gwas_summary$NearQTL/gwas_summary$numQTLtotal

fdr<-aggregate(NearQTL~Traits + File,function(cnts){
  props<-table(cnts)/sum(table(cnts))
  return(props[1])
},data=gwas_output)
colnames(fdr)[3]<-"PropWrongQTL"

gwas_summary<-merge(gwas_summary,fdr,by=c("Traits","File"))

final_freqs<-read.csv("../results/morph_freqs_summary.csv")
final_freqs$rep<-gsub("^.*Diversity_(.*)_pheno.csv","\\1",final_freqs$file)

gwas_summary$rep<-gsub("^.*Diversity_(.*)_pop_(\\d+).vcf","\\1_pop\\2",gwas_summary$File)

overall_info<-merge(final_freqs,
                    gwas_summary,
                    by="rep")

overall_info$diversity<-gsub("^.*\\/(\\w+)Diversity.*$","\\1",overall_info$file)
```

```{r propQTLplotFXN}

propQTL_plot<-function(gsummary,propCol,cols,ylabel="Prop. QTL detected",add_legend=FALSE){
  nchroms<-length(unique(gsummary$numChrom[gsummary$Traits=="Courter"]))
  nqtls<-length(unique(gsummary$numQTLonChrom[gsummary$Traits=="Courter"]))
  
  vioplot::vioplot(gsummary[gsummary$Traits=="Courter",propCol]~gsummary$CQ[gsummary$Traits=="Courter"],
          xlab="",
          ylab="",
          names=rep(unique(gsummary$numQTLonChrom[gsummary$Traits=="Courter"]),
                    nchroms),
          side="left",
          col=scales::alpha(cols["courter"],0.5),
          ylim=c(0,1),
          border=cols["courter"],
          lineCol=cols["courter"],
          colMed="black",
          cex.axis=2,
          cex.lab=2,
          las=2
          )
  
  vioplot::vioplot(gsummary[gsummary$Traits=="Parent",propCol]~gsummary$CQ[gsummary$Traits=="Parent"],
          side="right",
          col=scales::alpha(cols["parent"],0.5),
          add=TRUE,
          border=cols["parent"],
          lineCol=cols["parent"],
          colMed="black"
          )
  
  mtext(ylabel,2,cex=2,line=4)
  
  startpos<-0:(nchroms-1)*nqtls+1
  endpos<-startpos+nqtls-1
  labpos<-(startpos+endpos)/2
  
  text(x=labpos,
       y=rep(-0.25,nchroms),
       paste(unique(gsub("(\\d+) \\d+$","\\1",
                         levels(gsummary$CQ[gsummary$Traits=="Courter"]))),
             "Chr."),
       xpd=TRUE,
       cex=2)
  if(isTRUE(add_legend)){
      legend("top",
         pch=15,
         c("Courter","Parent"),
         col=cols[c("courter","parent")],
         bty='n',
         ncol=2,
         cex=2)
  }


}

```

```{r GWASoutcomes,fig.width=10,fig.height=9, fig.cap="Genome-wide association studies of the simulated genotypes reveals a relatively low detection rate of QTLs and a generally high false positive rate, both influenced by the genetic architecture of the traits. The top row show the results for genome-wide QTLs and the bottom row shows the results for supergenes."}
par(mfrow=c(2,2),mar=c(5,6,2,2))
propQTL_plot(gwas_summary[gwas_summary$Architecture=="qtls",],"PropQTL",cols)
propQTL_plot(gwas_summary[gwas_summary$Architecture=="qtls",],"PropWrongQTL",cols,"Falsely detected QTLs")


propQTL_plot(gwas_summary[gwas_summary$Architecture=="supergene",],"PropQTL",cols)
propQTL_plot(gwas_summary[gwas_summary$Architecture=="supergene",],"PropWrongQTL",cols,"Falsely detected QTLs")
```

```{r freqsDetection, fig.height=9, fig.width=9, fig.cap="The proportion of non-dominant morphs in the population did not impact the ability to detect QTLs through genome-wide associations. The top row shows multilocus QTL genetic architectures and the ability to detect traits considering either the parent/non-courters proportion or the non-courters/non-parent proportions. The bottom row shows the same results for supergenes."}
par(mfrow=c(2,2),mar=c(5,5,2,2))
plot(overall_info$P[overall_info$Traits=="Parent" & overall_info$Architecture=="qtls"],
     overall_info$PropQTL[overall_info$Traits=="Parent" & overall_info$Architecture=="qtls"],
     col=cols["parent"],
     pch=as.numeric(as.factor(overall_info$diversity[overall_info$Traits=="Parent" & overall_info$Architecture=="qtls"])),
     xlab="Parent/Non-courters Freq.",
     ylab="Prop. QTLs detected",
     cex.lab=2,
     cex=1.5,
     cex.axis=2,
     ylim=c(0,1),
     xlim=c(0,1),
     bty='l'
     )

plot(overall_info$NON[overall_info$Traits=="Courter" & overall_info$Architecture=="qtls"],
     overall_info$PropQTL[overall_info$Traits=="Courter" & overall_info$Architecture=="qtls"],
     col=cols["courter"],
     pch=as.numeric(as.factor(overall_info$diversity[overall_info$Traits=="Courter" & overall_info$Architecture=="qtls"])),
     xlab="Non-courter/Non-parent Freq.",
     ylab="Prop. QTLs detected",
     cex.lab=2,
     cex=1.5,
     cex.axis=2,
     ylim=c(0,1),
     #xlim=c(0,1),
     bty='l'
     )

# supergenes
plot(overall_info$P[overall_info$Traits=="Parent" & overall_info$Architecture=="supergene"],
     overall_info$PropQTL[overall_info$Traits=="Parent" & overall_info$Architecture=="supergene"],
     col=cols["parent"],
     pch=as.numeric(as.factor(overall_info$diversity[overall_info$Traits=="Parent" & overall_info$Architecture=="supergene"])),
     xlab="Parent/Non-courters Freq.",
     ylab="Prop. QTLs detected",
     cex.lab=2,
     cex=1.5,
     cex.axis=2,
     ylim=c(0,1),
     xlim=c(0,1),
     bty='l'
     )

plot(overall_info$NON[overall_info$Traits=="Courter" & overall_info$Architecture=="supergene"],
     overall_info$PropQTL[overall_info$Traits=="Courter" & overall_info$Architecture=="supergene"],
     col=cols["courter"],
     pch=as.numeric(as.factor(overall_info$diversity[overall_info$Traits=="Courter" & overall_info$Architecture=="supergene"])),
     xlab="Non-courter/Non-parent Freq.",
     ylab="Prop. QTLs detected",
     cex.lab=2,
     cex=1.5,
     cex.axis=2,
     ylim=c(0,1),
     #xlim=c(0,1),
     bty='l'
     )
```

# Discussion

Summary of findings

-   genetic architecture matters20

-   maybe counter intuitively, architecture reduces stochasticity,
    likely through genetic correlationsand/or linkage disequilibrium?

-   do supergenes not differ from QTLs because they are still multilocus
    (not quite operating asdiscrete haplotypes?)

-   GWAS is only going to pick up â¤ 50% of true QTLs, and will provide
    spurious results

-   a mention of intralocus conflict?

Why might cause the deviation from the single-locus expectation?

-   LD,

-   genetic correlations - compare to previous models and speculate on
    how it might impact theirresults - discuss real-world examples

-   QTLs vs supergenes don't make much of a difference, it seems:
    multilocus additive architecturesare still additive and multivariate

-   other patterns of inheritance (e.g., dominance) might be important
    (and would match patternsin empirical examples)

Inability to detect QTLs -- why would that happen?

-   Relaxed selection

-   extreme LD within supergenes makes associations difficult to
    identify (e.g., Kim et al zebrafinchpaper)

-   this is consistent with my SCA paper - what does this mean for our
    ability to study real worldpops?

-   One way that researchers have observed a role for supergenes is
    through genomic analysis. Putative regions are usually first
    identified through estimating allele frequency differencesbetween
    morphs, testing for associations with the traits of interest (e.g.,
    [22]).

-   Recent mathematical and simulation models have helped clarify the
    allelic variation that isexpected to be maintained under
    sex-specific selection pressures, and importantly have suggestedthat
    sex-specific selection must be incredibly strong [50].

Maintenance of variation (esp. morphs) and genetic architectures

-   genetic architecture is critically important

-   this doesn't include other elements of sex-specific or
    morph-specific architectures such assex-specific dominance of
    sex-biased gene expression

Overall conclusions

-   importance of supergenes? are they important?

# Acknowledgments

Thanks to the Flanagan lab for their feedback on early drafts of this
work. We acknowledge NgÄi TÅ«Ähuriri, upon whose lands the analyses and
writing was conducted. This work was conducted in part while SPF was a
postdoctoral fellowship at the National Institute for Mathematical and
Biological Synthesis, sponsored by the National Science Foundation
through NSF Award #DBI-1300426, with additional support from The
University of Tennessee, Knoxville. SPF was partially funded by the
Marsden fund grant number UOC1904 while working on this research.

# Data Accessibility

All code has been archived on Zenodo (DOI: ) and is maintained on github
(). Simulation results are archived on dryad ().

# Author Contributions

# References
