---
title: "Genetic architecture of alternative reproductive tactics impacts which morphs are maintained under balancing selection"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: no
    includes:
      in_header: preamble.sty
bibliography: references.bib
capsize: normalsize
documentclass: article
editor_options:
  chunk_output_type: console
  markdown:
    wrap: 72
fontsize: 11pt
abstract: "Alternative reproductive tactics are an intriguing example of variation maintained within populations and have evolved independently in a wide variety of animals. When these traits are fixed within the population, and individuals cannot change their tactic throughout their life, genetic variants are also expected to persist within the population. Empirical studies have found that in some cases these variants are clustered into supergenes, whereas in other cases causal genetic variants are found throughout the genome. An unknown question is to what extent the genetic architecture of loci contributing to these reproductively important traits impacts the maintenance of variation. Here, we present the a mathematical and a simulation model of a two-trait male morph, in which males can court females and provide parental care, do one or the other or neither. We use the mathematical model to identify regions of balancing selection parameter space in which multiple male morphs are retained, and used these parameters in simulation models with either a single genetic locus underpinning the traits, many QTLs distributed throughout the genome, and QTLs co-located within a low-recombination supergene. We found that including an explicit genetic architecture did not change whether multiple morphs are maintained, but multilocus genotypes resulted in less stochasticity in which morphs were retained under a high-diversity parameter setting. Regardless of genetic architecture, a surprising amount of genetic variation was retained genome-wide and in QTLs, and genome-wide associations were unreliable at correctly identifying the true QTLs. Our results contribute to the understanding of the impact of genetic architectures on evolutionary outcomes, maintenance of polymoprhisms, and will provide baseline expectations for empiricists seeking to understand the genetic architecture of alternative reproductive tactics.  \\par \\textbf{Keywords:} alternative mating tactics, frequency dependent selection, supergene, quantitative trait locus, genome-wide genetic diversity"
preprint: no
mainfont: roboto
geometry: margin=2.5cm
papersize: a4
---

```{r, include=FALSE}
remove_author <- function(x) {
  # identify empty author line
  i <- grep("^\\\\author\\{\\}$", x)
  # be sure it is the one pandoc inserts
  if(length(i) != 0 && grepl('^\\\\date\\{', x[i+1])) x <- x[-i]
  x
}
options(bookdown.post.latex = remove_author)
```

```{r setup, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",dpi=200,fig.height = 7,fig.width = 7)
knitr::opts_knit$set(root.dir='../results/') #change

```

```{r opts, echo = FALSE, warning=FALSE}
library(knitr)
knitr::opts_chunk$set(
  fig.path = "../figs/"
)
```

```{r librarysetup, echo=FALSE,echo=FALSE,message=FALSE, warning=FALSE}
library(RColorBrewer)
library(scales)
library(sm)
library(kableExtra)
library(plotly)
library(dplyr)
library(vegan)
library(tidyr)
tmp<-sapply(list.files(path="../../spfTools/R/",full.names = TRUE),source)
source("../R/002_freq_functions.R")

source("../morph_predictions/check_freqs.R")
source("../morph_predictions/morph_gens_ns.R")

source("../R/formatting.R")
```


# Introduction

Both genetic and phenotypic variation are required for selection to
occur, and as such the maintenance of variation in populations has long
been of interest to evolutionary biologists and conservation
geneticists. In general, evolution by natural selection is predicted to
decrease variation in populations, while only a few specific forms of
selection can maintain variation in populations, such as overdominance,
sexually antagonistic selection
[@connallonEvolutionaryConsequencesSexSpecific2018;
@connallonResolutionSexualAntagonism2011], and frequency dependent
selection [@alonzoMateChoiceGames2001]. Yet, substantial phenotypic
variation is observed, with ubiquitous examples such as the prevalence
of sexual dimorphism. Across a wide range of taxa, discrete alternative
reproductive tactics are frequently maintained, in which individuals of
the same species and sex display different morphs to maximize
reproductive fitness in unique ways
[@oliveiraAlternativeReproductiveTactics2008]. These stable
polymorphisms are typically maintained in populations through negative
frequency-dependent selection (e.g., female morphs in damselflies
[@iserbytNegativeFrequencydependentSelection2013]; Gouldian finches
[@prykeFrequencydependentPhysiologicalTradeoffs2007]), the interaction
of frequency-dependent and density-dependent factors (e.g.,
side-blotched lizards [@alonzoMateChoiceGames2001]), and fluctuating
selection pressures enabling the partitioning of reproductive success
over time (e.g., the feminized dwarf spider morph mates earlier, whereas
the alternative morph is able to mate with previously-mated females
later in the season [@hendrickxMasculinizingSupergeneUnderlies2022]).
While some of these alternative reproductive tactics are likely
determined by external cues and are therefore plastic (e.g., scarab
beetles, ocellated wrasse, wild turkeys), others are known to be fixed
at birth, and therefore are likely to be underpinned by genetic variants
(e.g., dwarf spiders, side-blotched lizards, ruffs; examples reviewed in
[@oliveiraAlternativeReproductiveTactics2008;
@mankSexspecificMorphsGenetics2022]).

Several authors have elegantly combined empirical data on fitness and
morph frequencies with theoretical models to uncover the primary
mechanisms driving the observed evolutionary dynamics in natural
populations [@sinervoRunawaySocialGames2001;
@sinervoRockPaperScissors1996;
@takahashiNegativeFrequencyDependentSelection2010;
@svenssonFemalePolymorphismFrequency2005;
@sinervoDevelopmentalPhysiologicalNeural2006;
@alonzoMateChoiceGames2001]. In general, the frequency-dependent fitness
surfaces prevent the loss of any morphs, and will result in either
oscillations of a predictable period or an evolutionarily stable
equilibrium point. Many of these models rely on game theory approaches
-- such as the dynamics of the rock-paper-scissors game
[@sinervoRockPaperScissors1996] -- and all of them either implicitly or
explicitly assume relatively simple genetic mechanisms underpinning the
discrete traits. The models with explicit population genetic components
to them include either one or two genetic loci with two to three alleles
[@takahashiNegativeFrequencyDependentSelection2010;
@sinervoRunawaySocialGames2001;
@svenssonFemalePolymorphismFrequency2005;
@morrisIntralocusTacticalConflict2013], including a single locus linked
to sex chromosomes [@ryanGeneticPolymorphismSwordtail1992]. Given that
morphs are suites of traits inherited together, some authors have
explored in verbal models the importance of correlational selection and
its associated formation of genetic correlations and linkage
disequilibrium across loci
[@sinervoCorrelationalSelectionEvolution2002]. While these relatively
simple genetic models do capture the evolutionary dynamics of the
specific examples they are paired with [@sinervoRunawaySocialGames2001],
genomic data has accumulated for a number of species suggesting that
categorical traits might in fact be underpinned by numerous quantitative
genetic loci [@kimGeneticsEvidenceBalancing2019;
@slyMolecularParallelismSignaling2022]. Furthermore, morphs are
comprised of multiple traits, and thus variation at multiple genetic
loci is likely to contribute to the evolutionary dynamics of morph
frequencies. As such, we might be missing out on important factors
constraining or facilitating evolution.

When traits are encoded by multiple loci, evolutionary constraints
arising from pleiotropy, genetic correlations, and physical linkage
could impact shifts in morph frequencies within and among populations.
As such, the reduction or removal of these constraints can facilitate
sequence evolution by relaxing positive selection on genes
[@harrisonSexualSelectionDrives2015; @dapperRelaxedSelectionRapid2020].
When polymorphic traits create conflict between sexes or morphs, these
genetic constraints can prevent the resolution of conflict and therefore
result in the maintenance of diversity
([@pennellContrastingEffectsIntralocus2016]). Often, these constraints
are overcome through mechanisms such as gene duplications (e.g.,
[@desmaraisEscapeAdaptiveConflict2008]), differential gene expression
[@mankPopulationGeneticsSexual2017], or even the evolution of asexuality
[@brownSocialDilemmasSupergenes2011]. Most work on such intralocus
conflict is related to sexual antagonism and sexual conflict, but
intralocus conflict can emerge among morphs of the same sex, as they
also share the majority of their genome
[@morrisIntralocusTacticalConflict2013]. Intriguingly, in ruffs, which
have three male alternative tactics, the evolution of the inversion
polymorphism underpinning the male morphs appears to have resolved
conflict among the male morphs, while simultaneously exacerbating
intralocus sexual conflict
[@giraldo-deckIntralocusConflictsAssociated2022].

The genetic architectures of alternative reproductive tactics have been
characterized in a few examples, and the diversity of outcomes highlight
the need for improved predictions. The majority of these examples first
generated a high quality reference genome and subsequently performed
whole genome resequencing of individuals of each morph and sex to
identify variants within populations that are associated with the
phenotypic tactics [@lamichhaneyStructuralGenomicChanges2016;
@maneySupergeneBirdFour2020; @kupperSupergeneDeterminesHighly2016;
@merrittSupergenelinkedEstrogenReceptor2020;
@kimSexlinkedSupergeneControls2017; @kimGeneticsEvidenceBalancing2019;
@thomasChromosomalPolymorphismLinked2008;
@hendrickxMasculinizingSupergeneUnderlies2022], although some
researchers have instead used linkage mapping
[@lepaisGeneticArchitectureThreshold2017] or investigation of
differential gene expression between morphs
[@nugentNeuroendocrineProfilesAssociated2016;
@stuglikAlternativeReproductiveTactics2014]. Consistent with other work
suggesting that traits critical to reproductive fitness are determined
by the additive contributions of many loci throughout the genome
[@slyMolecularParallelismSignaling2022;
@flanaganGenomewideSelectionComponents2017], several alternative
reproductive tactics are associated with many genetic variants
throughout the genome which contribute primarily additively to the
phenotypes [@kimGeneticsEvidenceBalancing2019] -- implying that although
the tactics might be discrete and inherited in an apparently Mendelian
fashion, their genetic basis is much more complex and multivariate. A
particularly intriguing result is that structural variants have
aggregated multiple variants contributing to the tactics into a
'supergene' in at least three species with alternative reproductive
tactics: the ruff [@lamichhaneyStructuralGenomicChanges2016;
@kupperSupergeneDeterminesHighly2016], white-throated sparrow
[@thomasChromosomalPolymorphismLinked2008], and dwarf spiders
[@hendrickxMasculinizingSupergeneUnderlies2022]. Supergenes have also
been linked to sex-specific traits important to reproduction such as
sperm morphology [@kimSexlinkedSupergeneControls2017;
@kniefSexchromosomeInversionCauses2017], head colour
[@gazdaGeneticMechanismSexual2020], and sex-specific migratory
behaviours [@pearseSexdependentDominanceMaintains2019].

The potential for supergenes to underpin sexually dimorphic polyphenisms
is intriguing, as it could explain the apparent mis-match between simple
single-locus population genetic models and the fact that these
reproductive tactics are suites of complex traits
[@sinervoRunawaySocialGames2001]. Supergenes, with their reduced
recombination between variants, also provide a mechanism for individuals
with largely similar genomic information to harbour fixed genetic
variation without requiring differential gene expression -- although in
some cases, differential expression of single genes could explain
complex and stable polymorphisms. For example, in canaries, sexual
dichromatism appears to be mediated by differential expression of a
single gene involved in carotenoid breakdown
[@gazdaGeneticMechanismSexual2020]. Androgens are a likely mechanism for
modulating these gene expression differences, for example in primates
male-biased genes are more likely to be proximal to androgen response
elements [@andersonRelationshipSexualDimorphism2022]. Unlike the
canaries, though, examples where supergenes apparently underpin sexually
dimorphic phenotypes tend to involve multiple morphs of one sex, usually
males. All known examples of alternative morphs determined by supergenes
include a female-mimic morph, including ruffs
[@kupperSupergeneDeterminesHighly2016;
@lamichhaneyStructuralGenomicChanges2016], dwarf spiders
[@hendrickxMasculinizingSupergeneUnderlies2022], and white-throated
sparrows [@thomasChromosomalPolymorphismLinked2008]. In the ruffs, a
third male morph exists, with an intermediate phenotype between the
dominant or 'classic' male morph and the female-mimic
[@kupperSupergeneDeterminesHighly2016;
@lamichhaneyStructuralGenomicChanges2016]. In some cases, a heterozygous
genotype is lethal (e.g., ruffs [@kupperSupergeneDeterminesHighly2016;
@lamichhaneyStructuralGenomicChanges2016]), but not in others (e.g.,
dwarf spiders [@hendrickxMasculinizingSupergeneUnderlies2022]). In the
white-throated sparrow and the ruff, these structural variants are
successful in coding for these variable phenotypes because they include
androgen receptors [@maneySupergenesSteroids2022]-- the white-throated
sparrow supergene contains the estrogen receptor gene which directly
influences the aggressive aspects of the dominant male phenotype
[@merrittSupergenelinkedEstrogenReceptor2020].

An open question is: to what extent do supergenes facilitate the
maintenance of multiple morphs, versus whether the evolution of multiple
morphs might facilitate the evolution of supergenes? This question is
difficult to address empirically, and is likely to have consequences for
the maintenance of genetic variation at loci contributing to traits. For
example, if a supergene exists and a mutation or arises within it that
impacts the morph traits, the genetic variation within the supergene
itself could have eroded due to relaxation of selection and reduced
recombination. As such, only the mutation would be likely to show
significant genetic variation -- this could make a supergene with
functional effects difficult to detect empirically. This is especially
true in cases such as the white-throated sparrow, in which the gene with
a major effect is an estrogen receptor
[@merrittSupergenelinkedEstrogenReceptor2020], which likely impacts the
expression of many genes. Furthermore, it is currently unknown to what
extent genetic architectures -- and the constraints inherit to them --
impact the dynamics of the frequency- and density-dependent mechanisms
that maintain simple cases of genetic variation in populations.

To tackle these problems, we use models to test how different genetic
architectures impact the maintenance of multiple male reproductive
morphs. We focus on first identifying key reproductive parameters that
influence the maintenance or loss of morphs using a mathematical model.
Then, we test whether different types of multilocus genetic
architectures change the frequencies of morphs that are maintained using
a simulation model. Our simulations of genetic architectures allow us to
output genotypes and apply genetic analyses that are commonly used in
empirical studies, including genome-wide association tests.

# Methods

Our model considers a biological scenario in which males display to
attract females, and with obligate male parental care. In our model,
males possess two independent traits, courtship and parental care, both
of which are expressed only in males (but are inherited by females).
Courting males display within a territory, and females choose a male to
mate with based on these displays. Once a female chooses a male, the
survival of the clutch is dependent on whether the chosen male is also a
parenting male. We assume that all females in the population always
prefer males with the courting trait, but females do not know when
choosing whether males will be parents. Males who do not court can only
reproduce by sneaking, and males who display but do not provide any
parental care will not produce viable offspring. Thus, our model has
four morphs: courter/parents (CP); courter/non-parents (CN);
non-courters/parents (NP); or non-courters/non-parents (NN), although
the courter/non-parent morph will never produce offspring. Our model is
diploid, and assumes non-overlapping generations. While this model does
not directly apply to all examples of alternative reproductive tactics,
it is biologically reasonable and common even if not universally true.
We expect that this model can be generalized to species in which one sex
(but not the other) has two fixed traits that impact mating success and
the survival of their offspring.

Each generation has three distinct phases: (1) mate choice and
fertilization; (2) parental care; and (3) viability selection. In the
mate choice phase of the model, we assume females always prefer to mate
with males displaying the courtship trait. If a female decides to mate
with a male, he receives all of her eggs to care for. The male is not
guaranteed paternity of his mates' entire brood, however, because
non-courting males are able to fertilize a proportion of the eggs. The
chosen parent does receive first-male priority. The consequences of this
setup are that the only mechanism by which non-courters can produce
offspring is through these sneak fertilizations. In the parental care
phase of the model, males with the parental care trait have 100%
offspring survival while males without the parental care trait have 0%
survival, independent of the male's courtship trait. Consequently,
non-parents will only be sustained in the population by sneaking
fertilizations in clutches cared for by parental males. Finally,
viability selection is imposed on offspring as they mature to adulthood,
such that males carrying the courtship trait and/or the parental care
trait have a lower probability of survival than females and males
without either courtship or parenting. Males who are both courters and
parents have a multiplicative effect of viability selection.

We implemented our model first as a mathematical model and then as an
individual-based simulation model, which for simplicity we will refer to
as the 'simulation model' throughout the text. The purpose of the
mathematical model was to generate baseline predictions for when
multiple morphs are maintained.

## Mathematical model {#math-model-methods}

In the mathematical model, we model the proportion of the population
that belong to morphs at each stage of the lifecycle (Fig.
\@ref(fig:modelDiagram)). In the first stage, we consider the proportion of
the population's clutches that will be allocated to each morph. The
maximum number of clutches that could exist is proportional to the
number of males in the population ($N_m$), if each male has the same
likelihood of attracting a female to deposit eggs with him. However, in
our model, males are not equally likely to be chosen by females, and
therefore the maximum number of clutches is scaled by the probability of
females mating with each type of male morph (denoted in a generic sense
with the subscript $M$, where $M$ contains ${CP, CN, NP, NN}$), denoted
by a sexual selection parameter $w_s$. In general for this model, the
probability of NP and NN males being chosen should equal $1-w_s$, where
$w_s$ is the probability of courting males (CP and CN) being chosen. The
general calculation for the number of clutches is:

$$
n=w_s(N_m f_{CP} + N_m f_{CN}) + (1-w_s)(N_m f_{NP} + N_m f_{NN}).
$$

We have constrained females to only mate with courting males, so
$w_{s_M}$ is equal to 1 for CP and CN males and 0 for NP and NN males.

```{r modelDiagram, fig.cap="Diagram of the two models describing the inheritance of alternative male reproductive morphs given courtship and parenting traits.", out.width='100%'}
knitr::include_graphics("../figs/Combined_model_diagrams.png", dpi=150)
```

The total reproductive effort in the population per morph is determined
by the number of males belonging to that morph and the morph's specified
reproductive allocation, which determines the maximum number of eggs an
individual male can fertilize:

$$
s_M = f_M N_m r_M.
$$

Chosen males have the first priority to fertilize the eggs in their
clutch, so the number of eggs fertilized within a single clutch per
morph is determined by the reproductive effort and the probability of
being chosen: $e_{o_M} = s_M w_{s_M}$. However, non-courting males also
sneak fertilizations, which impacts the overall proportion of
fertilizations per morph. The maximum number of possible fertilizations
arising from sneaking for a given non-parental male is
$N_s = n \times c \times n_{sneakers} \times r_{NN}$. This value
reflects that for every clutch ($n$), each sneaker can contribute up to
their individual reproductive allocation (where $r_{NN} = r_{NP}$)
offset by the sperm competition coefficient ($c$), but that this
potential allocation is proportional to the number of sneakers
($n_{sneakers}$) that are allowed to fertilize each clutch. If the
number of fertilizations arising from sneaking is smaller than the
morph's potential reproductive effort (i.e, if $N_s < s_M$), the number
of eggs fertilized through sneaking for that morph is the difference
between $N_s$ and $s_M$, truncated at zero. Otherwise, the number of
eggs fertilized through sneaking is $s_M$.

The survival of offspring can similarly be divided into offspring within
a male's own clutch versus in others' clutches. The offspring surviving
in any male's clutch is dependent on whether he is a parent or not,
which determines the selection coefficient arising from clutch survival
($w_{n_M}$), which for our purposes equals 1 for CP and NP males and
equals 0 for CN and NN males: $o_{o_M} = e_{o_M} w_{n_M}$. Offspring in
other males' clutches (i.e., those arising from sneaking) also depend on
whether the other males' clutch survives. Thus, we calculate the number
of offspring from sneaking that survive as:

$$
o_{s_M} = e_{s_M} \frac{\Sigma{o_{o_M}}}{\Sigma{e_{o_M}}}.
$$

Because we have entirely obligate parental care, the fraction will
either be 1 or 0, depending on whether the morph holding the clutch has
the parental care trait. Following the determination of clutch survival,
we then impose viability selection on surviving juveniles by multiplying
the viability selection coefficient (${w_v}_M$) by the surviving number
of offspring per morph ($o_M$): $j_M = o_M w_{v_M}$. Finally, the
frequency of each morph in the next generation was calculated by
dividing the surviving number of juveniles ($j_M$) by the sum of all
surviving juveniles: $f_M = \frac{j_M}{\Sigma_{i=1}^4{j_i}}$.

To evaluate the outcomes of this mathematical model, we iterated forward
in time for 100 generations from 1659 combinations of initial morph
frequencies and repeated those iterations across 5 values for the sperm
competition coefficient ($c$ = {0..1}), 21 values for the relative
reproductive coefficient ($r_M$ = {0...2}), and 3 values of the number
of sneakers ($n_{sneakers}$ ={1..3}). The iterations were implemented in
R v. 4.1.1 [@rcoreteamLanguageEnvironmentStatistical2021] using some
features of tidyR [@wickhamTidyrTidyMessy2021]. We also ran a subset of
parameter combinations for 1,000 generations for better comparisons with
the simulation model (see \@ref(sim-model-methods)).

We were interested in identifying the regions of parameter space that
resulted in the largest diversity of male morphs. To ameliorate the
difficulty of visualising and summarizing four frequencies, we leveraged
existing ecological mathematical tools to summarize diversity of a
system into a single metric. Specifically, we calculated the Shannon's
diversity index. We calculated this index in the 100th generation using
the R package vegan [@oksanenVeganCommunityEcology2015] and subsequently
visualised the diversity of the population using contour plots produced
by plotly [@sievertInteractiveWebbasedData2020] in a shiny app that
relies on R packages shiny [@changShinyWebApplication2021],
shinydashboard [@changShinydashboardCreateDashboards2018], rsconnect
[@mcphersonRsconnectDeploymentInterface2021], dplyr
[@wickhamDplyrGrammarData2021], and rmarkdown
[@allaireRmarkdownDynamicDocuments2021; @xieMarkdownDefinitiveGuide2018;
@xieMarkdownCookbook2020]. The interactive app is available at
<https://spflanagan.shinyapps.io/morph_predictions/>.

Note that this model does not incorporate recombination or inheritance
via females. Nonetheless, this model provides a baseline predictions for
parameter values likely to facilitate morph diversity within a
population. These limitations are addressed in the individual-based
simulation model.

## Individual-based simulation model {#sim-model-methods}

To evaluate the role of genetic architecture in the maintenance of
alternative reproductive tactics, we wrote an individual-based
simulation model in C++. The simulation model follows a lifecycle that
is largely the same as the mathematical model (Fig.
\@ref(fig:modelDiagram)), but rather than tracking proportions of the
population belonging to particular morphs, we tracked individuals within
the population (Fig. \@ref(fig:modelDiagram)). Each individual contained
diploid genotypes, which included both non-coding loci and quantitative
trait loci (QTLs). We included two types of QTLs: ones that determined
the individual's courting trait and ones that determined the
individual's parenting trait. The QTLs were each assigned a
randomly-drawn effect on the trait, and the QTLs additively determined
the trait value for each trait (i.e., the effects of all QTLs were
summed to determine the trait value). In this model, we assume no
environmental effects impacted the trait value. These trait values were
then compared to a threshold to dichotomize each trait. The threshold
for each trait was the mean trait value during the initialization of the
model (i.e., at generation 0).

We implemented three different genetic architectures for these QTLs: a
single gene per trait without any additional genetic information,
multiple QTLs randomly placed throughout chromosomes, and supergenes
located on one of multiple chromosomes. The single gene scenario is most
analogous to the mathematical model described above. When the traits
have multiple QTLs determining their trait values, each QTL is assigned
a location on a chromosome and the QTL's effect on the trait is
associated with a single nucleotide polymorphism (SNP). These SNPs
undergo recombination and mutation during mating. The supergene scenario
forced the QTLs to cluster into a small proportion of one of the
chromosomes. In all cases, we assume all QTLs are autosomal, and we do
not model sex chromosomes.

When we ran the model, we initialized four replicate populations with
the identical starting conditions to evaluate the impact of
stochasticity on the model. These identical conditions included
randomly-generated values for the allelic effects of QTLs, exact number
of individuals of each sex and morph, and thresholds for determining
morph traits (i.e., the values at which an individual is deemed a
courter vs a non-courter and a parent vs a non-parent). For each set of
four populations, we ran 10,000 burn-in generations followed by 2,000
'experimental' generations. Doing so ensured that the population reached
a steady state in which we could examine any stable polymorphisms or
cyclical fluctuations in morph frequencies.

Within each of those generations, the population underwent four main
phases: (1) mate choice and fertilization; (2) clutch survival; (3)
viability selection on juveniles; (4) maturation to adulthood. The main
difference between the individual-based simulation model and the
mathematical model were the introduction of stochasticity during mate
choice and the introduction of recombination and mutation during
reproduction. During the mate choice phase, females searched through a
set number of males and chose the first attractive male (i.e., the first
courter male she encountered). If a female did not find an acceptable
mate, she either did not mate ('no mating' or 'nm' scenarios) or she
chose a male at random with from the population and allocated all of her
offspring to him, depending on the parameter settings. Females always
preferred courting males.

Once a female chose a male, fertilization of the eggs occurred, in which
the chosen male had first male priority (i.e., the chosen male allocated
all of his reproductive effort to his own nest) and up to $n_{sneak}$
non-courting males could 'sneak' fertilizations, with their
contributions discounted by a factor of $c$. For all models analysed
here, $n_{sneak}=3$ and $c=0.5$. After the males were assigned their
paternity shares, the clutch survived if the chosen male was also a
parent. Subsequently, the total number of offspring in the population
were created based on the proportion of population reproductive success
achieved by each individual male and female (i.e., each generation, the
number of offspring created was equal to the carrying capacity).

When fertilization occurred, the parents' genotypes (both neutral
markers and QTLs) had the opportunity to recombine. Recombination
occurred independently on the chromosome inherited from the mother and
the chromosome inherited from the father. The number of recombination
events for a given chromosome was determined by drawing from a Poisson
distribution with the recombination rate as the mean, which we set to
0.2. For each recombination event, breakpoints were chosen by randomly
drawing from the total number of markers on a chromosome. These
breakpoints defined the segments of the chromosome that experienced
recombination, and each segment was randomly chosen to be assigned the
genotypes and effects from the parents' maternal or paternal chromosome.
When the genetic architecture was that of a supergene, if the
breakpoints fell within the region containing the aggregated QTLs for
the traits, the offspring was considered not viable.

After recombination occurred, if the progeny survived, the genotypes
underwent mutation, which occurred at a rate of $\mu$. For a given
individual, the number of mutations introduced was determined as
$2\mu \times n_{SNPs} \times n_{chrom}$. The location of each mutation
was randomly selected and the allele at that location was randomly
chosen from the set of existing alleles. The mutation also impacted the
phenotypic effects of QTLs by altering the original effect by a value
drawn from a normal distribution with a mean of 0 and a standard
deviation of $\sigma_\mu$. Finally, offspring were also randomly
assigned a sex.

The next stage in the lifecycle was viability selection, which impacted
only male offspring. For each male offspring, a survival probability was
calculated as $e^{\frac{-morph2}{2\omega_v}}$. Viability selection acted
multiplicatively on courter-parents (their survival probability was
$e^{\frac{-morph2}{2\omega_v}} \times e^{\frac{-morph2}{2\omega_v}}$). A
random value between 0 and 1 was drawn, and if the number was larger
than the survival probability for an individual, that individual did not
survive to adulthood. If, at this point, the number of surviving
offspring was larger than the carrying capacity, offspring were randomly
selected to become adults. When offspring matured to adulthood, the
previous generation's adults were replaced, and the lifecycle began
again.

Each set of genetic architectures and population parameters was also was
run with four types of mating systems, all of which required females to
lay their eggs with only a single male: males accept eggs from a single
female, and females deposit their eggs with a randomly-selected unmated
male in the population should they fail to find a preferred mate
('monogamy'); males accept eggs from multiple females, and females
deposit their eggs with a randomly-selected unmated male in the
population should they fail to find a preferred mate ('polygyny'); males
accept eggs from a single female, and females who fail to find a
preferred mate do not produce offspring ('monogamy no mating'); males
accept eggs from multiple females, and females who fail to find a
preferred mate do not produce offspring ('polygyny no mating').

### Model analysis

The model results were first assessed visually by graphing the
proportion of each morph that persisted to the final generation of the
model. Our first priority was to compare the outcomes of the simulation
model with a single locus for each trait to the results of our
mathematical model (from Section \@ref(math-model-methods)). Next, the
four mating systems were compared in the single-locus model. From this
comparison, we were then able to narrow the parameter space for the
comparisons of the two different multilocus genetic architectures. The
many different formats for these two genetic architectures were then
compared in terms of the morph composition in the final generation, and
the amount of variation among replicate runs.

Although selection is built into the model, if the population becomes
fixed for specific morphs -- and specific genotypes -- selection might
not be able to act due to a lack of variation. To investigate the
dynamics of selection over the course of the simulations (12,000
generations total), selection gradients on the courter trait and parent
trait (both the binary variable of courter/non-courter and
parent/non-parent) were estimated as the slope of a linear regression of
lifetime reproductive success on each trait value in generation 0 and in
generation 12,000. The patterns of selection were compared across
genetic architectures and parameter settings. Furthermore, we looked at
patterns of lifetime reproductive success across morphs within each
simulation in the final generation to evaluate whether balancing
selection contributed to morph diversity.

### Analysing multilocus genotypes

To provide a link between the predictions of our theoretical models and
empirical data, we analysed the genetic diversity of the populations at
the final generation of the individual based simulations. This is an
equivalent analysis to that of the re-sequencing that was conducted in
empirical studies [@lamichhaneyStructuralGenomicChanges2016;
@maneySupergeneBirdFour2020; @kupperSupergeneDeterminesHighly2016;
@merrittSupergenelinkedEstrogenReceptor2020;
@kimSexlinkedSupergeneControls2017; @kimGeneticsEvidenceBalancing2019;
@thomasChromosomalPolymorphismLinked2008;
@hendrickxMasculinizingSupergeneUnderlies2022]. The general idea behind
these genome-wide analyses is that QTLs will have higher genetic
diversity than background marker loci because balancing selection is
favouring multiple alleles at those loci, with genetic variation
existing between the morphs and specific variants associated with the
different phenotypes. For these analyses we only analysed multilocus
genotypes from the simulations that implemented the monogamous mating
system without random mating allowed (see Fig. S2).

To implement this analysis, we used vcfR
[@knausVcfrPackageManipulate2017] to estimate observed heterozygosity
(i.e., a measure of variation at a locus) for all markers in the genome,
including QTLs, and compared the frequencies of alleles between morphs
using $G_{ST}$, which is a measure between 0 and 1 that indicates how
different the frequencies are between groups. Additionally, an
association between male phenotypes and genetic markers was tested for
using GWASpoly [@rosyaraSoftwareGenomeWideAssociation2016]. Finally, we
used vcftools [@danecekVariantCallFormat2011] to test for signals of
balancing selection at the QTL loci compared to non-QTL loci. Because
linkage cause loci near to the actual target of selection (the QTL) to
have elevated variation as well, we identified 'peaks' of
differentiation ($G_{ST}$) and association (GWAS p-values), and
determined the proportion of these peaks that were within a 50 basepair
window of one of the true QTLs.

Linkage disequilibrium can occur among loci due to shared selection
pressures, in addition to physical linkage. Therefore, we might expect
that QTLs would be in linkage disequilibrium with each other, regardless
of genetic architecture. To investigate whether linkage disequilibrium
persisted over time within our various genetic architectures, we used
vcftools [@danecekVariantCallFormat2011] to estimate linkage
disequilibrium between all loci. We then estimated summary statistics
for each locus and compared linkage disequilibrium between QTLs to
linkage disequilibrium among non-QTLs. Because we had far fewer QTLs
than non-QTL loci in all of the iterations, we also used re-sampling of
the non-QTLs to estimate the average linkage disequilibrium, using 999
resampled estimates with the actual number of QTLs in each replicate.

# Results

## Mathematical model

The purpose of exploring the mathematical model was to identify the
parameter space in which, in a non-genetic model, more than one male
type is predicted to persist, thereby creating a baseline for comparison
to the simulation results. The courter/non-parent morph was never
maintained in the population. This result makes intuitive sense, because
if a female chooses to nest with a courter/non-parent, the nest will
always die (given our parameter settings) (Fig. \@ref(fig:baselineResults))

```{r baselineResults, fig.cap="Ternary plots of the baseline model (left) and the simulation model without genetic architectures included."}
knitr::include_graphics("../figs/baselinesNoStrictPercentages-1.png") # generated by testing_ternary_plots.Rmd
```


For identifying regions of
parameter space that resulted in multiple morphs, we focus on parameters
yielding Shannon-Wiener diversity indices $\ge 1$ , rather than attempt
to inspect plots of three or four frequencies.

```{r getMathResults}
results<-readRDS("../morph_predictions/morph_results_10000_equalStart.RDS")
results$diversity<-vegan::diversity(round(results[,c("CP","CN","NP","NN")],4))
```

```{r calcProps}
divRes<-results[results$diversity>0,]
ndiv<-nrow(divRes)

pcourt<-nrow(divRes[divRes$CP>=0.5,])/ndiv
pnp<-nrow(divRes[divRes$NP>=0.5,])/ndiv
pnn<-nrow(divRes[divRes$NN>=0.5,])/ndiv


```

The regions of the parameter space that have non-zero diversity values
correspond to scenarios where three of the four morphs were retained
(courter/parents, non-courter/parents, and non-courter/non-parents). The
dominant morph differed with various parameter combinations. Assuming
equal starting frequencies, after 10,000 generations, `r pcourt*100`% of
parameter combinations that retained polymorphism resulted in the
courter-parent morph comprising $\le$ 50% of the population. The
remaining `r (1-pcourt)*100`% of parameter combinations resulting in the
maintenance of alternative morphs had more equal representation of all
three morphs, with the average proportion of the population being
`r round(mean(divRes[divRes$CP<=0.5,"CP"])*100,2)` % ($\pm$
`r round(sem(divRes[divRes$CP<=0.5,"CP"])*100,2)` %) courter/parents,
`r round(mean(divRes[divRes$CP<=0.5,"NP"])*100,2)` % ($\pm$
`r round(sem(divRes[divRes$CP<=0.5,"NP"])*100,2)` %)
non-courter/parents, and
`r round(mean(divRes[divRes$CP<=0.5,"NN"])*100,2)` % ($\pm$
`r round(sem(divRes[divRes$CP<=0.5,"NN"])*100,2)` %)
non-courter/non-parents.

Overall, the parameters impacting whether diversity was maintained after
both 100 and 1000 generations were the initial frequencies of the
morphs, sperm competition ($c$), relative reproductive allocation ($r$),
and the number of sneakers allowed ($n_{sneak}$, which impacted the
proportion of offspring a courting male could sire in his own clutch).
Some initial starting conditions took more than 100 generations to reach
a consistent (i.e., equilibrial) state, but the contributions of the
initial frequencies and these three parameters remained qualitatively
consistent across 100 and 1000 generations.

As expected, larger sperm competition coefficients ($c$) resulted in a
higher likelihood of non-courters being retained in the population,
which makes sense given that $c$ impacts the proportion of eggs within a
clutch that a non-courting male can successfully fertilize. Sperm
competition interacted with the reproductive allocation ratio, $r$, such
that as long as $c>0.2$, if $0.5 \le r \le 1$, the maximum diversity was
achieved. As an example of why this makes sense, that if $r = 0.5$,
$c=0.2$, and the non-courting male's number of eggs he can fertilize is
8, the non-courting male would fertilize
$\frac{8 \times 2}{(8 \times 0.5)+(2 \times 0.2 \times 8)}$ =
`r round((8*0.2)/(4+(2*0.2*8)),3)` of the offspring. The impacts of
sperm competition on morph diversity were further influenced by
$n_{sneak}$, with larger contributions from an individual non-courting
male (e.g., larger $c$) being required with fewer sneakers (Fig.
S1). This result makes sense because the number of
sneakers is multiplied by the sperm competition coefficient to determine
the proportion of offspring fertilized by the non-courters for a given
clutch.

## Individual-based simulation model

The results from the mathematical model were used to select parameter
settings for the individual-based simulation model, with one set
expected to produce low-diversity populations (i.e., those where only
one morph were retained) and another set expected to produce
high-diversity populations (i.e., where 2 or more morphs were retained).
The chosen parameter settings are shown in Table
\@ref(tab:SimParamsTable).

```{r SimParamsTable}
SimParams<-data.frame(
  parameter = c("r (courter:non-courter ratio)",
                "Max number of offspring CP males can produce",
                "Max number of offspring CN males can produce",
                "Max number of offspring NP males can produce",
                "Max number of offspring NN males can produce",
                "c (sperm competition coefficient)",
                "number of sneakers"),
  LowDiv = c( 2, 
             8,
             8,
             4,
             4,
             0.5,
             2),
   HighDiv = c( 0.7, 
             6,
             6,
             8,
             8,
             0.75,
             2)
)

kable(SimParams, "latex",booktab=TRUE,
      col.names = c("Parameter", "Low Diversity Value","High Diversity Value"),
      caption="Table of parameters selected to create a high-diversity population. All morphs started at equal frequencies.")  %>% 
  kableExtra::kable_styling(latex_options = "HOLD_position")

```

Describe (Fig. \@ref(fig:baselineResults)).

### The role of polygenic inheritance of traits

```{r}
final_freqs<-read.csv("../results/morph_freqs_summary.csv")
final_freqs$rep<-gsub("^.*Diversity_(.*)_pheno.csv","\\1",final_freqs$file)

```

We incorporated polygenic inheritance of the male traits in two ways,
both of which reduced the effect of stochasticity on the final morph
frequencies. In the low diversity parameter space, the outcomes remained
consistent with the single-locus model (and the predictions of the
mathematical model): courter/parents dominated the populations, with
complete fixation of the morph in the majority of runs, and any
variation existing in the form of 1-2 individuals (courter/parent
frequencies were $\ge$ 99% in all cases). These results were consistent
across all genetic architectures, including QTLs distributed across 2,
4, and 8 chromosomes and QTLs clustered in supergenes spanning from 5%
to 50% of a single chromosome. In the high diversity parameter space,
polygenic loci distributed throughout the genome resulted in the
maintenance of the courter/parent trait at a frequency of
`r round(mean(final_freqs$CP[grep("qtls.*highDiversity",final_freqs$file)])*100,2)`%
($\pm$
`r round(sem(final_freqs$CP[grep("qtls.*highDiversity",final_freqs$file)])*100,2)`%),
with the non-courter/parent being the dominant morph. These proportions
are remarkably consistent across genetic architectures (Fig. \@ref(fig:architectureTernary)). Most simulations
with supergenes resulted in very similar final morph frequencies
(`r round(mean(final_freqs$CP[grep("supergene.*highDiversity",final_freqs$file)])*100,2)`%
$\pm$
`r round(sem(final_freqs$CP[grep("supergene.*highDiversity",final_freqs$file)])*100,2)`%),
and only occasionally resulted in either three morphs being retained or
the non-courter/non-parent morph replacing the non-courter/parent. The
size of the supergene did not substantially impact the final morph
frequencies, nor did the total number of QTLs underlying each locus.

(ref:archTernCap) Caption for the ternary diagram.

```{r architectureTernary,  fig.cap='(ref:archTernCap)'}
knitr::include_graphics("../figs/architectureTernaryLowHigh-1.png")
```



Our initial expectation was that supergenes would follow a pattern more
similar to the single locus results, because empirical examples have
demonstrated Mendelian inheritance patterns approximating single-locus
models (e.g., [@hendrickxMasculinizingSupergeneUnderlies2022,
@kupperSupergeneDeterminesHighly2016,,@lamichhaneyStructuralGenomicChanges2016]).
To better understand the reasons for the similarity between the the
genome-wide QTLs simulations versus the supergene simulations, we
visualised the QTL haplotypes in supergenes (see Fig.
\@ref(fig:ExampleGenomewide) for a representative image). These
visualisations suggest that because of the additive, polygenic nature of
our traits, individuals have multiple pathways to creating the morphs in
terms of genotypes at each QTL, even when QTLs are segregating together
in a supergene.

```{r ExampleGenomewide,echo=FALSE,out.width="49%", fig.cap="Plots of representative simulations using QTLs (left) versus supergenes (right), both with 8 QTLs per trait and 2 chromosomes with 1000 marker loci each. The plots show genome-wide estimates of commonly-estimated population genetic parameters. From top to bottom: p-values from a genome-wide association test for the courter trait; p-values from a genome-wide association test for the parent trait; Gst between male morphs; smoothed Tajima's D estimated in 10bp segments; expected heterozygosity; and average pairwise linkage disequilibrium. QTLs for courter and parent traits are circled in blue and green, respectively.",fig.show='hold',fig.align='center'}
knitr::include_graphics(c("../figs/QTLFig-1.png","../figs/supergeneFig-1.png"))
```



### Genome-wide genetic diversity

Supergenes enable specific combinations of alleles at multiple loci to
be inherited together as a single haplotype; as such, we expected that
the simulations with supergenes would evolve to have lower within-morph
genetic diversity at QTLs (i.e., low observed heterozygosity) than the
genome-wide QTLs scenarios. Surprisingly, we found that the average
genome-wide expected heterozygosity was not substantially different in
simulations with genome-wide QTLs versus QTLs in supergenes in the final
generation of the simulations (t = -0.31, df = 326.75, p = -0.31;
genome-wide mean = 0.15, supergene mean = 0.16; Fig.
\@ref(fig:genomewideDiv) A, B). In most runs, the heterozygosity at QTLs
was did not significantly differ from heterozygosity at neutral markers
(mean p-value from t-tests is 0.5 ± 0.01). Tajima's D, which tests for
directional or balancing selection, also reflects genetic variation in a
sequence of DNA, and did not deviate substantially between genetic
architectures and did not relate to final morph frequencies (Fig.
\@ref(fig:genomewideDiv) C, D).

Similarly, we expected supergenes to show more differentiation between
morphs, since their QTLs segregated together. However, $G_{ST}$ between
male morphs was also not predicted by the frequency of courter/parent
morphs (Fig. \@ref(fig:genomewideDiv) E). Significant peaks in
genome-wide $G_{ST}$ estimates were identified, and we evaluated whether
QTLs were located near to those $G_{ST}$ peaks (i.e., we identified if
$G_{ST}$ outliers were associated with true QTLs). From this, we then
calculated the proportion of true QTLs that were located in these
significant $G_{ST}$ peaks, and found that on average 0.71 (± 0), and
this proportion was not related to the courter/parent frequency (Fig.
\@ref(fig:genomewideDiv) F).

```{r}
welch_correctDF<-function(var1,var2,n1,n2){
  num<-((var1/n1)+(var2/n2))^2
  den<-(((var1/n1)^2)/(n1-1))+(((var2/n2)^2)/(n2-1))
  df<-num/den
  return(df)
}

```

```{r readGenomewide}

qtl_outliers<-read.csv("qtl_outliers.csv")
qtl_outliers$Architecture<-"qtls"
supergene_outliers<-read.csv("supergene_outliers.csv")
supergene_outliers$Architecture<-"supergene"
popgen_data<-dplyr::bind_rows(qtl_outliers,supergene_outliers)

# add low diversity ones
qtl_outliers<-read.csv("qtl_outliers_lowDiv.csv")
qtl_outliers$Architecture<-"qtls"
popgen_data<-dplyr::bind_rows(popgen_data,qtl_outliers)
supergene_outliers<-read.csv("supergene_outliers_lowDiv.csv")
supergene_outliers$Architecture<-"supergene"
popgen_data<-dplyr::bind_rows(popgen_data,supergene_outliers)

sediff<-sqrt((popgen_data$gHtVar/(popgen_data$nChrom*1000)) + 
               (popgen_data$qHtVar/(popgen_data$nQTLS)))
HetT<-(popgen_data$gHtMean - popgen_data$qHtMean)/sediff
HetP<-pt(HetT,
         df = welch_correctDF(popgen_data$gHtVar,
                              popgen_data$qHtVar,
                              (popgen_data$nChrom*1000),
                              popgen_data$nQTLS), 
         lower.tail = (HetT >=0))

popgen_data$HetP<-HetP
popgen_data$Diversity<-factor(gsub("^.*/(\\w+)Diversity.*$","\\1",popgen_data$sim),
                              levels=c("low","high"))



final_freqs<-read.csv("../results/morph_freqs_summary.csv")
final_freqs$rep  <-gsub("../fixedART-results/.*\\/(.*\\w+_monogamy_nm_.*pop\\d).*","\\1",final_freqs$file)

td_summary<-read.csv("../results/tajimaD_summary.csv", row.names = 1)
td_summary<-td_summary[grep("monogamy_nm",rownames(td_summary)),]
td_summary$rep<-gsub("../fixedART-results/.*\\/(.*\\w+_monogamy_nm_.*pop)_(\\d).*","\\1\\2",rownames(td_summary))

td_summary$architecture<-gsub("../fixedART-results/(.*)\\/\\/(\\w+)Diversity_.*_monogamy_nm_.*pop_(\\d).*","\\1",rownames(td_summary))
td_summary$diversity<-factor(gsub("../fixedART-results/(.*)\\/\\/(\\w+)Diversity_.*_monogamy_nm_.*pop_(\\d).*","\\2",rownames(td_summary)),
                           levels=c("low","high"))


td_freqs<-merge(td_summary, final_freqs,by="rep")



```

```{r genomewideDiv, fig.cap="Genome-wide genetic diversity in the simulation results. Genome-wide observed heterozygosity (A), Tajima’s D (C), and male-male Fst (E) were not influenced by the frequency of the morphs in the final generation, nor by the genetic architecture. The final morph frequencies also did not impact the significance of comparisons of QTLs versus neutral loci for heterozygosity (B) or Tajima’s D (D), or the proportion of QTLs found in the Fst peaks (F). Note that heterozygosity and Gst results are only shown for the runs with high diversity parameter sets"}

par(mfrow=c(3,2),xpd=TRUE, mar=c(5,5,1.5,1), oma=c(2,2,3,1))

plot(popgen_data$propCP,
     popgen_data$gHtMean,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(popgen_data[,"Diversity"])],
       0.5),
     cex=1.5,
     cex.lab=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Mean Obs. Het.",
     bty='l')
text(0.28,0.42,"A",cex=2.5)

plot(popgen_data$propCP,
     popgen_data$HetP,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(popgen_data[,"Diversity"])],
       0.5),
     cex=1.5,
     cex.lab=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Obs. Het P-values",
     bty='l')
text(0.28,0.9,"B",cex=2.5)

# Tajima's D
plot(td_freqs$CP,
     td_freqs$avgTD,
     pch=as.numeric(as.factor(td_freqs$architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(td_freqs$diversity)],
       0.5),
     cex=1.5,
     cex.lab=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Mean Tajima's D",
     bty='l')
text(0.28,4.4,"C",cex=2.5)

plot(td_freqs$CP,
     td_freqs$p,
     pch=as.numeric(as.factor(td_freqs$architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(td_freqs$diversity)],
       0.5),
     cex=1.5,
     cex.lab=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Tajima's D P-values",
     bty='l')
text(0.28,0.9,"D",cex=2.5)


# Fst
plot(popgen_data$propCP,
     popgen_data$gGpMeanMM,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(popgen_data[,"Diversity"])],
       0.5),
     cex=1.5,
     cex.lab=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Mean male-male Gst",
     bty='l')
text(0.28,0.6,"E",cex=2.5)

plot(popgen_data$propCP,
     popgen_data$propQTLsInPeaksMM,
     pch=as.numeric(as.factor(popgen_data$Architecture))+15,
     col=scales::alpha(
       divcols[as.numeric(popgen_data[,"Diversity"])],
       0.5),
     cex=1.5,
     cex.lab=1.5,
     xlab="Courter/Parent Frequency",
     ylab="Prop. QTL in Gst peak",
     bty='l')
text(0.28,0.9,"F",cex=2.5)

outer_legend("top",
       bty='n',
       pch=c(17, 16, 15,15),
       lty=c(0, 0, 1,1),
       col=c("grey", "grey",divcols[2],divcols[1]),
       legend=c("Supergenes", "QTLs", "High Diversity", "Low Diversity"),
       ncol=4,
       cex=1.5)
```

### Will GWAS detect the actual QTLs?

Genome-wide association studies are commonly used to test for
associations between alleles and traits within populations, and we
assessed how well this technique detects true QTLs from our simulations.
As expected, the detection of true QTLs in a GWAS was more likely when
fewer QTLs underpin the trait, especially if those QTLs are aggregated
on fewer chromosomes (Fig. \@ref(fig:GWASoutcomes)). Somewhat
surprisingly, the QTLs for parent traits were more reliably detected by
GWAS than the courter QTLs. In the majority of simulations, the final
population only contained parental males, but both courters and
non-courters persisted in the population (Fig.
\@ref(fig:MatingTypesFig)). The finding the parent QTLs are more likely
to be detected is likely because that trait experienced stronger
selection on it (i.e., being a non-parent was more costly than being a
non-courter).

We expected the supergene scenarios to have a higher detection rate, but
this was not the case. The primary difference was that the variance in
false positive rates was increased with supergenes, but the true
positive rates were unaffected by the localization of QTLs.

```{r getGWASinfo}
gwas_output<-read.csv("GWAS_summary.csv", header = FALSE)
colnames(gwas_output)<-c("Marker",
                          "Chrom",
                          "Position",
                          "Model",
                          "R2",
                          "pval",
                          "NearQTL",
                          "Traits",
                          "File")

gwas_summary<-aggregate(NearQTL~Traits + File,sum,data=gwas_output)

gwas_summary$numQTLonChrom<-as.numeric(gsub("^.*q(\\d+)_c(\\d+)_\\d.*vcf","\\1",gwas_summary$File))
gwas_summary$numChrom<-as.numeric(gsub("^.*q(\\d+)_c(\\d+)_\\d.*vcf","\\2",gwas_summary$File))
gwas_summary$numQTLtotal<-gwas_summary$numQTLonChrom*gwas_summary$numChrom
gwas_summary$Architecture<-gsub("^.*results\\/(.*)\\/\\w+.*.vcf","\\1",gwas_summary$File)
gwas_summary$Architecture<-gsub("\\/","",gwas_summary$Architecture) # remove any extra / 
gwas_summary$CQ<-as.factor(paste(gwas_summary$numChrom,gwas_summary$numQTLtotal))

# calculate the proportion of QTLs identified
gwas_summary$PropQTL<-gwas_summary$NearQTL/gwas_summary$numQTLtotal

fdr<-aggregate(NearQTL~Traits + File,function(cnts){
  props<-table(cnts)/sum(table(cnts))
  return(props[1])
},data=gwas_output)
colnames(fdr)[3]<-"PropWrongQTL"

gwas_summary<-merge(gwas_summary,fdr,by=c("Traits","File"))

final_freqs<-read.csv("../results/morph_freqs_summary.csv")
final_freqs$rep<-gsub("^.*Diversity_(.*)_pheno.csv","\\1",final_freqs$file)

gwas_summary$rep<-gsub("^.*Diversity_(.*)_pop_(\\d+).vcf","\\1_pop\\2",gwas_summary$File)

overall_info<-merge(final_freqs,
                    gwas_summary,
                    by="rep")

overall_info$diversity<-gsub("^.*\\/(\\w+)Diversity.*$","\\1",overall_info$file)
```

```{r propQTLplotFXN}

propQTL_plot<-function(gsummary,propCol,cols,ylabel="Prop. QTL detected",add_legend=FALSE){
  nchroms<-length(unique(gsummary$numChrom[gsummary$Traits=="Courter"]))
  nqtls<-length(unique(gsummary$numQTLonChrom[gsummary$Traits=="Courter"]))
  
  vioplot::vioplot(gsummary[gsummary$Traits=="Courter",propCol]~gsummary$CQ[gsummary$Traits=="Courter"],
          xlab="",
          ylab="",
          names=rep(unique(gsummary$numQTLonChrom[gsummary$Traits=="Courter"]),
                    nchroms),
          side="left",
          col=scales::alpha(cols["courter"],0.5),
          ylim=c(0,1),
          border=cols["courter"],
          lineCol=cols["courter"],
          colMed=cols["courter"],
          cex.axis=2,
          cex.lab=2,
          las=2
          )
  
  vioplot::vioplot(gsummary[gsummary$Traits=="Parent",propCol]~gsummary$CQ[gsummary$Traits=="Parent"],
          side="right",
          col=scales::alpha(cols["parent"],0.5),
          add=TRUE,
          border=cols["parent"],
          lineCol=cols["parent"],
          colMed=cols["parent"]
          )
  
  mtext(ylabel,2,cex=2,line=4)
  
  startpos<-0:(nchroms-1)*nqtls+1
  endpos<-startpos+nqtls-1
  labpos<-(startpos+endpos)/2
  
  text(x=labpos,
       y=rep(-0.25,nchroms),
       paste(unique(gsub("(\\d+) \\d+$","\\1",
                         levels(gsummary$CQ[gsummary$Traits=="Courter"]))),
             "Chr."),
       xpd=TRUE,
       cex=2)
  if(isTRUE(add_legend)){
      legend("top",
         pch=15,
         c("Courter","Parent"),
         col=cols[c("courter","parent")],
         bty='n',
         ncol=2,
         cex=2)
  }


}

```

```{r GWASoutcomes,fig.width=10,fig.height=9, fig.cap="Genome-wide association studies of the simulated genotypes reveals a relatively low detection rate of QTLs and a generally high false positive rate, both influenced by the genetic architecture of the traits. The top row show the results for genome-wide QTLs and the bottom row shows the results for supergenes."}
par(mfrow=c(2,2),mar=c(5,6,4,2))
gwas_summary<-gwas_summary[grep("monogamy_nm",gwas_summary$File),]
gwas_summary$CQ<-factor(gwas_summary$CQ)
propQTL_plot(gwas_summary[gwas_summary$Architecture=="qtls",],"PropQTL",cols)
propQTL_plot(gwas_summary[gwas_summary$Architecture=="qtls",],"PropWrongQTL",cols,"Falsely detected QTLs")


propQTL_plot(gwas_summary[gwas_summary$Architecture=="supergene",],"PropQTL",cols)
propQTL_plot(gwas_summary[gwas_summary$Architecture=="supergene",],"PropWrongQTL",cols,"Falsely detected QTLs")

mtext("QTLs",3,cex=2,line=-2,outer=TRUE)
mtext("Supergenes",3,cex=2,line=-30,outer=TRUE)

```

# Discussion

Understanding how the genetic architecture of stable polymorphisms
facilitates or constrains their evolution has been an open question.
Using a simulation model, we show that genetic architecture might not
shift the parameter space in which polymorphisms are stable, but can
determine which morphs are likely to be retained. Complex genetic
architectures reduced the stochasticity associated with
architecture-free models, likely due to genetic correlations between
traits and loci. Surprisingly, QTLs organised into supergenes, although
associated with high linkage disequilibrium, did not have different
outcomes than QTLs distributed genome-wide. In addition, unexpectedly
high levels of genetic diversity were retained within morphs in the
supergene scenarios.

We expected the supergenes to evolve to contain single haplotypes for
each morph, as has been reported in empirical examples of supergenes
controlling complex phenotypes (e.g.,
@lamichhaneyStructuralGenomicChanges2016;
@kupperSupergeneDeterminesHighly2016;
@kimSexlinkedSupergeneControls2017;
@hendrickxMasculinizingSupergeneUnderlies2022). A striking difference in
our results from those of the empirical examples is that we do not see
fixed alleles exclusively in one morph of the other. Instead,
quantitative variation is retained within the supergenes, and we do not
observe discrete haplotypes underlying the categorical morphs. If we
compare our supergene results to our QTL results, as expected, linkage
disequilibrium was elevated within the supergene region in our model,
and similarly high levels of LD were not observed in the genome-wide QTL
models (Fig. \@ref(fig:ExampleGenomewide)). However, measures of genetic
variation and between-morph differentiation were similar in the two
types of models (Fig. \@ref(fig:genomewideDiv)). To some extent, this is
likely an artefact of our model set-up. In empirical examples, a key
gene substantially contributes to the phenotypic morph, for example the
steroid hormone receptor genes in white-throated sparrows and ruffs
supergenes [@maneySupergenesSteroids2022]. In our model, we assumed all
QTLs had roughly similar but small contributions to the phenotype.
Furthermore, many polymorphisms could rely on dominance hierarchies
(e.g., as modelled by [@svenssonFemalePolymorphismFrequency2005]), and
could even be underpinned by dominance reversals between morphs
[@pearseSexdependentDominanceMaintains2019]or sexes
[@grieshopDominanceReversalsAntagonistic2021]. Regardless, we provide
the important revelation that discrete morphs can be maintained within
populations even without discrete genetic architectures with a
threshold-based switch, contrary to predictions from some authors
[@boltonColourPolymorphismLikely2016]. This means that for such
quantitative threshold traits, they could occur in a supergene and not
result in the single- or two-locus inheritance patterns seen in
side-blotched lizards [@sinervoRunawaySocialGames2001] and damselflies
[@svenssonFemalePolymorphismFrequency2005].

Our choice to compare QTLs co-located within non-recombining supergenes
to a genome-wide distribution was motivated by evidence of fixed
alternative reproductive morphs having both of these genetic
architectures [@lamichhaneyStructuralGenomicChanges2016;
@kupperSupergeneDeterminesHighly2016;
@kimSexlinkedSupergeneControls2017;
@hendrickxMasculinizingSupergeneUnderlies2022;
@kimGeneticsEvidenceBalancing2019; @gazdaGeneticMechanismSexual2020].
One of the primary ways that empirical researchers characterise the
genetic architectures underlying traits such as discrete morphs is
through genome-wide association studies and other genome-wide scans of
genetic diversity and differentiation. Putative regions are usually
first identified through estimating allele frequency differences between
morphs, testing for associations with the traits of interest (e.g.,
[@lagunas-roblesLinkedSupergenesUnderlie2021;
@kupperSupergeneDeterminesHighly2016]). We performed these analyses on
simulated genome-wide data, and found that genome-wide association
studies are likely to only identify $\le$ 50% of true QTLs, and identify
many false positives. Low true positives and high false positives are
consistent with other simulation studies of genome-wide studies of
selection and trait associations
[@flanaganIdentifyingSignaturesSexual2015]. In this study, these
limitations are consistent with both genetic architectures, so appears
to not only be driven by the extreme linkage disequilibrium within
supergenes that makes associations difficult to identify
[@kimSexlinkedSupergeneControls2017]. A possible explanation for the
results in this study could be the relaxed selection observed at the
point when the populations were 'genotyped' (Fig.
\@ref(fig:selectionGradients)) -- mutations could have cropped up that
were no longer being removed by selection due to relaxed selection.

Mathematical and simulation models have helped clarify the allelic
variation that is expected to be maintained under sex-specific selection
pressures [@flanaganIdentifyingSignaturesSexual2015;
@connallonGeneralPopulationGenetic2012], and importantly have suggested
that sex-specific selection must be incredibly strong
[@kasimatisLimitsGenomicDivergence2019]. When it comes to morphs, this
might not be true -- we observed both maintenance of variation in
phenotypes and in genotypes, with reasonably high levels of genetic
differentiation between male types, even after selection had been
relaxed. In other words, even the runs where the selection coefficient
was estimated to be zero had average male-male $G_{ST}$ substantially
above zero, which is counter the the predictions from single-locus
models [@kasimatisLimitsGenomicDivergence2019]. As such, our model
highlights how demographic factors, stochasticity, and genomic effects
such as pleiotropy and linkage can impact genetic variation, and
potentially obscure putative genome-wide signals of selection.

Our model applies to fixed polymorphisms, which only describes a portion
of alternative reproductive tactics. Evidence from ruffs and
white-throated sparrows suggest that some of the single-locus dynamics
are driven by genetic variation at key hormone receptor genes
[@maneySupergenesSteroids2022], highlighting the importance of
considering hormonal cascades and gene expression networks in models
(e.g., [@draghiEvolutionaryDynamicsEvolvability2009;
@draghiRobustnessNoiseGene2015]). These genetic networks might be
especially important for plastic traits (e.g., the swordtail fish
*Xiphophorus multilineatus* [@liottaEvidenceGeneticIntegration2021]) or
those that appear to be controlled by gene expression rather than fixed
genetic variants (e.g., ocellated wrasse,
[@nugentNeuroendocrineProfilesAssociated2016]). Future models could
investigate the evolutionary dynamics of morphs that are controlled by
gene networks.

Here, we show that the genetic architecture of discrete morphs impacts
the frequency of morphs maintained under balancing selection, but
genetic architecture does not impact the parameter space in which
multiple morphs are maintained. Our models highlight the limitations of
using genome-wide scans to identify the architecture of discrete traits,
as we show that most QTLs are not detected even when the overall number
of QTLs is low and balancing selection is ongoing. These simulation
results have implications for empirical studies aiming to identify the
genetic basis of discrete traits.

# Acknowledgments

Thanks to the Flanagan lab for their feedback on early drafts of this
work. We acknowledge Ngāi Tūāhuriri, upon whose lands the analyses and
writing was conducted. This work was conducted in part while SPF was a
postdoctoral fellowship at the National Institute for Mathematical and
Biological Synthesis, sponsored by the National Science Foundation
through NSF Award #DBI-1300426, with additional support from The
University of Tennessee, Knoxville. SPF was partially funded by the
Marsden fund grant number UOC1904 while working on this research.

# Data Accessibility

All code has been archived on Zenodo (DOI: ) and is maintained on github
(). Simulation results are archived on dryad ().

# Author Contributions

# References
